// @atproto/lex-client/dist/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const lex_data_1 = require("@atproto/lex-data");
const lex_schema_1 = require("@atproto/lex-schema");
const agent_js_1 = require("./agent.js");
const index_js_1 = require("./lexicons/index.js");
const util_js_1 = require("./util.js");
const xrpc_js_1 = require("./xrpc.js");
class Client {
    static appLabelers = [];
    /**
     * Configures the Client (or its sub classes) globally.
     */
    static configure(opts) {
        if (opts.appLabelers)
            this.appLabelers = [...opts.appLabelers];
    }
    agent;
    headers;
    service;
    labelers;
    constructor(agent, options = {}) {
        this.agent = (0, agent_js_1.buildAgent)(agent);
        this.service = options.service;
        this.labelers = new Set(options.labelers);
        this.headers = new Headers(options.headers);
    }
    get did() {
        return this.agent.did;
    }
    get assertDid() {
        this.assertAuthenticated();
        return this.did;
    }
    assertAuthenticated() {
        if (!this.did)
            throw new lex_data_1.LexError('AuthenticationRequired');
    }
    setLabelers(labelers = []) {
        this.clearLabelers();
        this.addLabelers(labelers);
    }
    addLabelers(labelers) {
        for (const labeler of labelers)
            this.labelers.add(labeler);
    }
    clearLabelers() {
        this.labelers.clear();
    }
    fetchHandler(path, init) {
        const headers = (0, util_js_1.buildAtprotoHeaders)({
            headers: init.headers,
            service: this.service,
            labelers: [
                ...this.constructor.appLabelers.map((l) => `${l};redact`),
                ...this.labelers,
            ],
        });
        // Incoming headers take precedence
        for (const [key, value] of this.headers) {
            if (!headers.has(key))
                headers.set(key, value);
        }
        // @NOTE The agent here could be another Client instance.
        return this.agent.fetchHandler(path, { ...init, headers });
    }
    async xrpc(ns, options = {}) {
        return (0, xrpc_js_1.xrpc)(this, ns, options);
    }
    async xrpcSafe(ns, options = {}) {
        return (0, xrpc_js_1.xrpcSafe)(this, ns, options);
    }
    /**
     * @param rkey Leave `undefined` to have the server generate a TID.
     */
    async createRecord(record, rkey, options) {
        return this.xrpc(index_js_1.com.atproto.repo.createRecord.main, {
            ...options,
            body: {
                repo: options?.repo ?? this.assertDid,
                collection: record.$type,
                record,
                rkey,
                validate: options?.validate,
                swapCommit: options?.swapCommit,
            },
        });
    }
    async deleteRecord(collection, rkey, options) {
        return this.xrpc(index_js_1.com.atproto.repo.deleteRecord.main, {
            ...options,
            body: {
                repo: options?.repo ?? this.assertDid,
                collection,
                rkey,
                swapCommit: options?.swapCommit,
                swapRecord: options?.swapRecord,
            },
        });
    }
    async getRecord(collection, rkey, options) {
        return this.xrpc(index_js_1.com.atproto.repo.getRecord.main, {
            ...options,
            params: {
                repo: options?.repo ?? this.assertDid,
                collection,
                rkey,
            },
        });
    }
    async putRecord(record, rkey, options) {
        return this.xrpc(index_js_1.com.atproto.repo.putRecord.main, {
            ...options,
            body: {
                repo: options?.repo ?? this.assertDid,
                collection: record.$type,
                rkey,
                record,
                validate: options?.validate,
                swapCommit: options?.swapCommit,
                swapRecord: options?.swapRecord,
            },
        });
    }
    async listRecords(nsid, options) {
        return this.xrpc(index_js_1.com.atproto.repo.listRecords.main, {
            ...options,
            params: {
                repo: options?.repo ?? this.assertDid,
                collection: nsid,
                cursor: options?.cursor,
                limit: options?.limit,
                reverse: options?.reverse,
            },
        });
    }
    async uploadBlob(body, options) {
        return this.xrpc(index_js_1.com.atproto.repo.uploadBlob.main, {
            ...options,
            body,
        });
    }
    async getBlob(did, cid, options) {
        return this.xrpc(index_js_1.com.atproto.sync.getBlob.main, {
            ...options,
            params: { did, cid },
        });
    }
    async call(ns, arg, options = {}) {
        const method = (0, lex_schema_1.getMain)(ns);
        if (typeof method === 'function') {
            return method(this, arg, options);
        }
        if (method instanceof lex_schema_1.Procedure) {
            const result = await this.xrpc(method, { ...options, body: arg });
            return result.body;
        }
        else if (method instanceof lex_schema_1.Query) {
            const result = await this.xrpc(method, { ...options, params: arg });
            return result.body;
        }
        else {
            throw new TypeError('Invalid lexicon');
        }
    }
    async create(ns, input, options = {}) {
        const schema = (0, lex_schema_1.getMain)(ns);
        const record = schema.build(input);
        const rkey = options.rkey ?? getDefaultRecordKey(schema);
        if (rkey !== undefined)
            schema.keySchema.assert(rkey);
        const response = await this.createRecord(record, rkey, options);
        return response.body;
    }
    async delete(ns, options = {}) {
        const schema = (0, lex_schema_1.getMain)(ns);
        const rkey = schema.keySchema.parse(options.rkey ?? getLiteralRecordKey(schema));
        const response = await this.deleteRecord(schema.$type, rkey, options);
        return response.body;
    }
    async get(ns, options = {}) {
        const schema = (0, lex_schema_1.getMain)(ns);
        const rkey = schema.keySchema.parse(options.rkey ?? getLiteralRecordKey(schema));
        const response = await this.getRecord(schema.$type, rkey, options);
        const value = schema.validate(response.body.value);
        return { ...response.body, value };
    }
    async put(ns, input, options = {}) {
        const schema = (0, lex_schema_1.getMain)(ns);
        const record = schema.build(input);
        const rkey = options.rkey ?? getLiteralRecordKey(schema);
        const response = await this.putRecord(record, rkey, options);
        return response.body;
    }
    async list(ns, options) {
        const schema = (0, lex_schema_1.getMain)(ns);
        const { body } = await this.listRecords(schema.$type, options);
        const records = [];
        const invalid = [];
        for (const record of body.records) {
            const parsed = schema.safeValidate(record.value);
            if (parsed.success) {
                records.push({ ...record, value: parsed.value });
            }
            else {
                invalid.push(record.value);
            }
        }
        return { ...body, records, invalid };
    }
}
exports.Client = Client;
function getDefaultRecordKey(schema) {
    // Let the server generate the TID
    if (schema.key === 'tid')
        return undefined;
    if (schema.key === 'any')
        return undefined;
    return getLiteralRecordKey(schema);
}
function getLiteralRecordKey(schema) {
    if (schema.key.startsWith('literal:')) {
        return schema.key.slice(8);
    }
    throw new TypeError(`An "rkey" must be provided for record key type "${schema.key}" (${schema.$type})`);
}
