// @atproto/lex-client/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrpcInternalError = exports.XrpcInvalidResponseError = exports.XrpcUpstreamError = exports.XrpcAuthenticationError = exports.XrpcResponseError = exports.XrpcError = exports.LexError = exports.RETRYABLE_HTTP_STATUS_CODES = void 0;
exports.isXrpcErrorPayload = isXrpcErrorPayload;
exports.asXrpcFailure = asXrpcFailure;
const lex_data_1 = require("@atproto/lex-data");
Object.defineProperty(exports, "LexError", { enumerable: true, get: function () { return lex_data_1.LexError; } });
const lex_schema_1 = require("@atproto/lex-schema");
const www_authenticate_js_1 = require("./www-authenticate.js");
/**
 * HTTP status codes that indicate a transient error that may succeed on retry.
 *
 * Includes:
 * - 408 Request Timeout
 * - 425 Too Early
 * - 429 Too Many Requests (rate limited)
 * - 500 Internal Server Error
 * - 502 Bad Gateway
 * - 503 Service Unavailable
 * - 504 Gateway Timeout
 * - 522 Connection Timed Out (Cloudflare)
 * - 524 A Timeout Occurred (Cloudflare)
 */
exports.RETRYABLE_HTTP_STATUS_CODES = new Set([
    408, 425, 429, 500, 502, 503, 504, 522, 524,
]);
/**
 * All unsuccessful responses should follow a standard error response
 * schema. The Content-Type should be application/json, and the payload
 * should be a JSON object with the following fields:
 *
 * - `error` (string, required): type name of the error (generic ASCII
 *   constant, no whitespace)
 * - `message` (string, optional): description of the error, appropriate for
 *   display to humans
 *
 * This function checks whether a given payload matches this schema.
 */
function isXrpcErrorPayload(payload) {
    return (payload != null &&
        payload.encoding === 'application/json' &&
        lex_schema_1.lexErrorDataSchema.matches(payload.body));
}
/**
 * Abstract base class for all XRPC errors.
 *
 * Extends {@link LexError} and implements {@link ResultFailure} for use with
 * safe/result-based error handling patterns.
 *
 * @typeParam M - The XRPC method type (Procedure or Query)
 * @typeParam N - The error code type
 * @typeParam TReason - The reason type for ResultFailure
 *
 * @see {@link XrpcResponseError} - For valid XRPC error responses
 * @see {@link XrpcUpstreamError} - For invalid/unexpected responses
 * @see {@link XrpcInternalError} - For network/internal errors
 */
class XrpcError extends lex_data_1.LexError {
    method;
    name = 'XrpcError';
    constructor(method, error, message = `${error} Lexicon RPC error`, options) {
        super(error, message, options);
        this.method = method;
    }
    /**
     * @see {@link ResultFailure.success}
     */
    success = false;
    matchesSchemaErrors() {
        return this.method.errors?.includes(this.error) ?? false;
    }
}
exports.XrpcError = XrpcError;
/**
 * Error class for valid XRPC error responses from the server.
 *
 * This represents a properly formatted XRPC error where the server returned
 * a non-2xx status with a valid JSON error payload containing `error` and
 * optional `message` fields.
 *
 * Use {@link matchesSchemaErrors} to check if the error matches the method's declared
 * error types for type-safe error handling.
 *
 * @typeParam M - The XRPC method type
 * @typeParam N - The error code type (inferred from method or generic)
 *
 * @example Handling specific errors
 * ```typescript
 * try {
 *   await client.xrpc(someMethod, options)
 * } catch (err) {
 *   if (err instanceof XrpcResponseError && err.error === 'RecordNotFound') {
 *     // Handle not found case
 *   }
 * }
 * ```
 */
class XrpcResponseError extends XrpcError {
    response;
    payload;
    name = 'XrpcResponseError';
    constructor(method, response, payload, options) {
        const { error, message } = payload.body;
        super(method, error, message, options);
        this.response = response;
        this.payload = payload;
    }
    get reason() {
        return this;
    }
    shouldRetry() {
        return exports.RETRYABLE_HTTP_STATUS_CODES.has(this.response.status);
    }
    toJSON() {
        return this.payload.body;
    }
    toDownstreamError() {
        // If the upstream server returned a 5xx error, we want to return a 502 Bad
        // Gateway to downstream clients, as the issue is with the upstream server,
        // not us. We still return the original error code and message in the body
        // for transparency, but we do not want to expose internal server errors
        // from the upstream server as-is to downstream clients.
        return {
            status: this.response.status === 500 ? 502 : this.status,
            headers: stripHopByHopHeaders(this.headers),
            body: this.toJSON(),
        };
    }
    get status() {
        return this.response.status;
    }
    get headers() {
        return this.response.headers;
    }
    get body() {
        return this.payload.body;
    }
}
exports.XrpcResponseError = XrpcResponseError;
/**
 * Error class for 401 Unauthorized XRPC responses.
 *
 * Extends {@link XrpcResponseError} with access to parsed WWW-Authenticate header
 * information, useful for implementing authentication flows.
 *
 * Authentication errors are never retryable as they require user intervention
 * (e.g., re-authentication, token refresh).
 *
 * @typeParam M - The XRPC method type
 * @typeParam N - The error code type
 *
 * @example Handling authentication errors
 * ```typescript
 * try {
 *   await client.xrpc(someMethod, options)
 * } catch (err) {
 *   if (err instanceof XrpcAuthenticationError) {
 *     const { DPoP } = err.wwwAuthenticate
 *     if (DPoP?.error === 'use_dpop_nonce') {
 *       // Handle DPoP nonce requirement
 *     }
 *   }
 * }
 * ```
 */
class XrpcAuthenticationError extends XrpcResponseError {
    name = 'XrpcAuthenticationError';
    shouldRetry() {
        return false;
    }
    #wwwAuthenticateCached;
    /**
     * Parsed WWW-Authenticate header from the response.
     * Contains authentication scheme parameters (e.g., Bearer realm, DPoP nonce).
     */
    get wwwAuthenticate() {
        return (this.#wwwAuthenticateCached ??=
            (0, www_authenticate_js_1.parseWWWAuthenticateHeader)(this.response.headers.get('www-authenticate')) ?? {});
    }
    toDownstreamError() {
        return {
            status: 401,
            headers: stripHopByHopHeaders(this.headers),
            body: this.toJSON(),
        };
    }
}
exports.XrpcAuthenticationError = XrpcAuthenticationError;
/**
 * Error class for invalid or unprocessable XRPC responses from upstream servers.
 *
 * This occurs when the server returns a response that doesn't conform to the
 * XRPC protocol, such as:
 * - Missing or invalid Content-Type header
 * - Response body that doesn't match the method's output schema
 * - Non-JSON error responses
 * - Responses from non-XRPC endpoints
 *
 * The error code is always 'UpstreamFailure' and maps to HTTP 502 Bad Gateway
 * when converted to a response.
 *
 * @typeParam M - The XRPC method type
 */
class XrpcUpstreamError extends XrpcError {
    response;
    payload;
    name = 'XrpcUpstreamError';
    constructor(method, response, payload = null, message = `Unexpected upstream XRPC response`, options) {
        super(method, 'UpstreamFailure', message, options);
        this.response = response;
        this.payload = payload;
    }
    get reason() {
        return this;
    }
    shouldRetry() {
        return exports.RETRYABLE_HTTP_STATUS_CODES.has(this.response.status);
    }
    toDownstreamError() {
        return { status: 502, body: this.toJSON() };
    }
}
exports.XrpcUpstreamError = XrpcUpstreamError;
/**
 * Error class for invalid XRPC responses that fail schema validation.
 *
 * This is a specific type of {@link XrpcUpstreamError} that indicates the
 * upstream server returned a response that was structurally valid but did not
 * conform to the expected schema for the method. This likely indicates a
 * mismatch between client and server versions or an issue with the server's
 * XRPC implementation.
 *
 * @typeParam M - The XRPC method type
 */
class XrpcInvalidResponseError extends XrpcUpstreamError {
    cause;
    name = 'XrpcInvalidResponseError';
    constructor(method, response, payload, cause) {
        super(method, response, payload, `Invalid response: ${cause.message}`, {
            cause,
        });
        this.cause = cause;
    }
    toDownstreamError() {
        // @NOTE This could be reflected as both a 500 ("we" are at fault) and 502
        // ("they" are at fault). We are using 502 here to allow downstream clients
        // to determine that the issue lies at the interface between us and the
        // upstream server, rather than an issue with our internal processing.
        return { status: 502, body: this.toJSON() };
    }
}
exports.XrpcInvalidResponseError = XrpcInvalidResponseError;
/**
 * Error class for internal/client-side errors during XRPC requests.
 *
 * This represents errors that occur before or during the request that are not
 * server responses, such as:
 * - Network errors (connection refused, DNS failure)
 * - Request timeouts
 * - Request aborted via AbortSignal
 * - Invalid request construction
 *
 * The error code is always 'InternalServerError' and these errors are
 * optimistically considered retryable.
 *
 * @typeParam M - The XRPC method type
 */
class XrpcInternalError extends XrpcError {
    name = 'XrpcInternalError';
    constructor(method, message, options) {
        super(method, 'InternalServerError', message ?? 'Unable to fulfill XRPC request', options);
    }
    get reason() {
        return this;
    }
    shouldRetry() {
        // Ideally, we would inspect the reason to determine if it's retryable
        // (by detecting network errors, timeouts, etc.). Since these cases are
        // highly platform-dependent, we optimistically assume all internal
        // errors are retryable.
        return true;
    }
    toJSON() {
        // @NOTE Do not expose internal error details to downstream clients
        return { error: this.error, message: 'Internal Server Error' };
    }
    toDownstreamError() {
        return { status: 500, body: this.toJSON() };
    }
}
exports.XrpcInternalError = XrpcInternalError;
/**
 * Converts an unknown error into an appropriate {@link XrpcFailure} type.
 *
 * If the error is already an XrpcFailure for the given method, returns it as-is.
 * Otherwise, wraps it in an {@link XrpcInternalError}.
 *
 * @param method - The XRPC method that was called
 * @param cause - The error to convert
 * @returns An XrpcFailure instance
 *
 * @example
 * ```typescript
 * try {
 *   const response = await fetch(...)
 *   // ... process response
 * } catch (err) {
 *   return asXrpcFailure(method, err)
 * }
 * ```
 */
function asXrpcFailure(method, cause) {
    if (cause instanceof XrpcResponseError ||
        cause instanceof XrpcUpstreamError ||
        cause instanceof XrpcInternalError) {
        if (cause.method === method)
            return cause;
    }
    return new XrpcInternalError(method, undefined, { cause });
}
const HOP_BY_HOP_HEADERS = new Set([
    'connection',
    'keep-alive',
    'proxy-authenticate',
    'proxy-authorization',
    'te',
    'trailer',
    'transfer-encoding',
    'upgrade',
]);
function stripHopByHopHeaders(headers) {
    const result = new Headers(headers);
    // Remove statically known hop-by-hop headers
    for (const name of HOP_BY_HOP_HEADERS) {
        result.delete(name);
    }
    // Remove headers listed in the "Connection" header
    const connection = headers.get('connection');
    if (connection) {
        for (const name of connection.split(',')) {
            result.delete(name.trim());
        }
    }
    // These are not actually hop-by-hop headers, but we remove them because the
    // upstream payload gets parsed and re-serialized, so content length and
    // encoding may no longer be accurate.
    result.delete('content-length');
    result.delete('content-encoding');
    return result;
}
