// @atproto/lex-client/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrpcUnexpectedError = exports.XrpcUpstreamError = exports.XrpcResponseError = exports.XrpcError = exports.LexError = void 0;
exports.isXrpcErrorPayload = isXrpcErrorPayload;
const lex_data_1 = require("@atproto/lex-data");
Object.defineProperty(exports, "LexError", { enumerable: true, get: function () { return lex_data_1.LexError; } });
const lex_schema_1 = require("@atproto/lex-schema");
class XrpcError extends lex_data_1.LexError {
    name = 'XrpcError';
    constructor(error, message = `${error} Lexicon RPC error`, options) {
        super(error, message, options);
    }
}
exports.XrpcError = XrpcError;
/**
 * All unsuccessful responses should follow a standard error response
 * schema. The Content-Type should be application/json, and the payload
 * should be a JSON object with the following fields:
 *
 * - `error` (string, required): type name of the error (generic ASCII
 *   constant, no whitespace)
 * - `message` (string, optional): description of the error, appropriate for
 *   display to humans
 *
 * This function checks whether a given payload matches this schema.
 */
function isXrpcErrorPayload(payload) {
    return (payload !== null &&
        payload.encoding === 'application/json' &&
        lex_schema_1.l.lexErrorData.matches(payload.body));
}
/**
 * Class used to represent an HTTP request that resulted in an XRPC method error
 * That is, a non-2xx response with a valid XRPC error payload.
 */
class XrpcResponseError extends XrpcError {
    method;
    status;
    headers;
    payload;
    name = 'XrpcResponseError';
    constructor(method, status, headers, payload, options) {
        const { error, message } = payload.body;
        super(error, message, options);
        this.method = method;
        this.status = status;
        this.headers = headers;
        this.payload = payload;
    }
    success = false;
    get reason() {
        return this;
    }
    get body() {
        return this.payload.body;
    }
    matchesSchema() {
        return this.method.errors?.includes(this.error) ?? false;
    }
    shouldRetry() {
        // Do not retry client errors
        if (this.status < 500)
            return false;
        return true;
    }
    toJSON() {
        return this.payload.body;
    }
    toResponse() {
        const { status, headers } = this;
        return Response.json(this.toJSON(), { status, headers });
    }
}
exports.XrpcResponseError = XrpcResponseError;
/**
 * This class represents an invalid XRPC response from the server.
 */
class XrpcUpstreamError extends XrpcError {
    name = 'XrpcUpstreamError';
    // For debugging purposes, we keep the response details here
    response;
    constructor(error, message, response, payload, options) {
        super(error, message, { cause: options?.cause });
        this.response = {
            status: response.status,
            headers: response.headers,
            payload,
        };
    }
    success = false;
    get reason() {
        return this;
    }
    matchesSchema() {
        return false;
    }
    shouldRetry() {
        // Do not retry client errors
        return this.response.status >= 500;
    }
    toResponse() {
        return Response.json(this.toJSON(), { status: 502 });
    }
}
exports.XrpcUpstreamError = XrpcUpstreamError;
class XrpcUnexpectedError extends XrpcError {
    name = 'XrpcUnexpectedError';
    constructor(message, options) {
        super('InternalServerError', message, options);
    }
    success = false;
    get reason() {
        return this.cause;
    }
    matchesSchema() {
        return false;
    }
    shouldRetry() {
        return true;
    }
    toResponse() {
        return Response.json(this.toJSON(), { status: 500 });
    }
    static from(cause, message = cause instanceof lex_data_1.LexError
        ? cause.message
        : 'XRPC request failed') {
        if (cause instanceof XrpcUnexpectedError)
            return cause;
        return new XrpcUnexpectedError(message, { cause });
    }
}
exports.XrpcUnexpectedError = XrpcUnexpectedError;
