// @atproto/lex-schema/dist/schema/array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.array = exports.ArraySchema = void 0;
const core_js_1 = require("../core.js");
const memoize_js_1 = require("../util/memoize.js");
class ArraySchema extends core_js_1.Schema {
    validator;
    options;
    constructor(validator, options = {}) {
        super();
        this.validator = validator;
        this.options = options;
    }
    validateInContext(input, ctx) {
        if (!Array.isArray(input)) {
            return ctx.issueInvalidType(input, 'array');
        }
        const { minLength, maxLength } = this.options;
        if (minLength != null && input.length < minLength) {
            return ctx.issueTooSmall(input, 'array', minLength, input.length);
        }
        if (maxLength != null && input.length > maxLength) {
            return ctx.issueTooBig(input, 'array', maxLength, input.length);
        }
        let copy;
        for (let i = 0; i < input.length; i++) {
            const result = ctx.validateChild(input, i, this.validator);
            if (!result.success)
                return result;
            if (result.value !== input[i]) {
                if (ctx.options.mode === 'validate') {
                    // In "validate" mode, we can't modify the input, so we issue an error
                    return ctx.issueInvalidPropertyValue(input, i, [result.value]);
                }
                // Copy on write (but only if we did not already make a copy)
                copy ??= Array.from(input);
                copy[i] = result.value;
            }
        }
        return ctx.success(copy ?? input);
    }
}
exports.ArraySchema = ArraySchema;
function arraySchema(items, options) {
    return new ArraySchema(items, options);
}
exports.array = (0, memoize_js_1.memoizedTransformer)(arraySchema);
