// @atproto/lex-schema/dist/schema/discriminated-union.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscriminatedUnionSchema = void 0;
exports.discriminatedUnion = discriminatedUnion;
const lex_data_1 = require("@atproto/lex-data");
const core_js_1 = require("../core.js");
const enum_js_1 = require("./enum.js");
const literal_js_1 = require("./literal.js");
/**
 * @note There is no discriminated union in Lexicon schemas. This is a custom
 * extension to allow optimized validation of union of objects when using the
 * lex library programmatically (i.e. not code generated from a lexicon schema).
 */
class DiscriminatedUnionSchema extends core_js_1.Schema {
    discriminator;
    variants;
    variantsMap;
    constructor(discriminator, variants) {
        super();
        this.discriminator = discriminator;
        this.variants = variants;
        // Although we usually try to avoid initialization work in constructors,
        // here it is necessary to ensure that invalid discriminated throw from the
        // place of construction, rather than later during validation.
        this.variantsMap = buildVariantsMap(discriminator, variants);
    }
    validateInContext(input, ctx) {
        if (!(0, lex_data_1.isPlainObject)(input)) {
            return ctx.issueInvalidType(input, 'object');
        }
        const { discriminator } = this;
        if (!Object.hasOwn(input, discriminator)) {
            return ctx.issueRequiredKey(input, discriminator);
        }
        const discriminatorValue = input[discriminator];
        const variant = this.variantsMap.get(discriminatorValue);
        if (variant) {
            return ctx.validate(input, variant);
        }
        return ctx.issueInvalidPropertyValue(input, discriminator, [
            ...this.variantsMap.keys(),
        ]);
    }
}
exports.DiscriminatedUnionSchema = DiscriminatedUnionSchema;
function buildVariantsMap(discriminator, variants) {
    const variantsMap = new Map();
    for (const variant of variants) {
        const schema = variant.shape[discriminator];
        if (schema instanceof literal_js_1.LiteralSchema) {
            if (variantsMap.has(schema.value)) {
                throw new TypeError(`Overlapping discriminator value: ${schema.value}`);
            }
            variantsMap.set(schema.value, variant);
        }
        else if (schema instanceof enum_js_1.EnumSchema) {
            for (const val of schema.values) {
                if (variantsMap.has(val)) {
                    throw new TypeError(`Overlapping discriminator value: ${val}`);
                }
                variantsMap.set(val, variant);
            }
        }
        else {
            // Only enumerable discriminator schemas are supported
            // Should never happen if types are used correctly
            throw new TypeError(`Discriminator schema must be a LiteralSchema or EnumSchema`);
        }
    }
    return variantsMap;
}
/*@__NO_SIDE_EFFECTS__*/
function discriminatedUnion(discriminator, variants) {
    return new DiscriminatedUnionSchema(discriminator, variants);
}
