// @atproto/lex-schema/dist/schema/intersection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntersectionSchema = void 0;
exports.intersection = intersection;
const core_js_1 = require("../core.js");
/**
 * Schema for combining an object schema with a dictionary schema.
 *
 * Validates that the input matches both the fixed object shape and allows
 * additional properties that match the dictionary schema. Properties defined
 * in the object schema are validated by the object, and remaining properties
 * are validated by the dictionary.
 *
 * @template Left - The ObjectSchema type for fixed properties
 * @template Right - The DictSchema type for additional properties
 *
 * @example
 * ```ts
 * const schema = new IntersectionSchema(
 *   l.object({ name: l.string() }),
 *   l.dict(l.string(), l.integer())
 * )
 * // Validates: { name: 'test', score: 100, count: 5 }
 * ```
 */
class IntersectionSchema extends core_js_1.Schema {
    left;
    right;
    type = 'intersection';
    constructor(left, right) {
        super();
        this.left = left;
        this.right = right;
    }
    validateInContext(input, ctx) {
        const leftResult = ctx.validate(input, this.left);
        if (!leftResult.success)
            return leftResult;
        return this.right.validateInContext(leftResult.value, ctx, {
            ignoredKeys: this.left.validatorsMap,
        });
    }
}
exports.IntersectionSchema = IntersectionSchema;
/**
 * Creates an intersection schema combining fixed object properties with dynamic dictionary properties.
 *
 * Useful for objects that have a known set of properties plus additional
 * arbitrary properties that follow a pattern.
 *
 * @param left - Object schema defining the fixed, known properties
 * @param right - Dictionary schema for validating additional properties
 * @returns A new {@link IntersectionSchema} instance
 *
 * @example
 * ```ts
 * // Object with fixed and dynamic properties
 * const configSchema = l.intersection(
 *   l.object({
 *     version: l.integer(),
 *     name: l.string(),
 *   }),
 *   l.dict(l.string(), l.string()) // Additional string properties
 * )
 *
 * configSchema.parse({
 *   version: 1,
 *   name: 'my-config',
 *   customField: 'value',
 *   anotherField: 'another',
 * })
 * ```
 */
/*@__NO_SIDE_EFFECTS__*/
function intersection(left, right) {
    return new IntersectionSchema(left, right);
}
