// @atproto/lex-schema/dist/schema/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.params = exports.ParamsSchema = exports.paramsSchema = exports.paramSchema = void 0;
const lex_data_1 = require("@atproto/lex-data");
const core_js_1 = require("../core.js");
const lazy_property_js_1 = require("../util/lazy-property.js");
const memoize_js_1 = require("../util/memoize.js");
const array_js_1 = require("./array.js");
const boolean_js_1 = require("./boolean.js");
const dict_js_1 = require("./dict.js");
const integer_js_1 = require("./integer.js");
const optional_js_1 = require("./optional.js");
const string_js_1 = require("./string.js");
const union_js_1 = require("./union.js");
const paramScalarSchema = (0, union_js_1.union)([(0, boolean_js_1.boolean)(), (0, integer_js_1.integer)(), (0, string_js_1.string)()]);
exports.paramSchema = (0, union_js_1.union)([paramScalarSchema, (0, array_js_1.array)(paramScalarSchema)]);
exports.paramsSchema = (0, dict_js_1.dict)((0, string_js_1.string)(), (0, optional_js_1.optional)(exports.paramSchema));
class ParamsSchema extends core_js_1.Schema {
    shape;
    constructor(shape) {
        super();
        this.shape = shape;
    }
    get shapeValidators() {
        const map = new Map(Object.entries(this.shape));
        return (0, lazy_property_js_1.lazyProperty)(this, 'shapeValidators', map);
    }
    validateInContext(input, ctx) {
        // @TODO BETTER SUPPORT Input/Output
        if (!(0, lex_data_1.isPlainObject)(input)) {
            return ctx.issueInvalidType(input, 'object');
        }
        // Lazily copy value
        let copy;
        // Ensure that non-specified params conform to param schema
        for (const key in input) {
            if (this.shapeValidators.has(key))
                continue;
            const result = ctx.validateChild(input, key, exports.paramSchema);
            if (!result.success)
                return result;
            if (result.value !== input[key]) {
                if (ctx.options.mode === 'validate') {
                    // In "validate" mode, we can't modify the input, so we issue an error
                    return ctx.issueInvalidPropertyValue(input, key, [result.value]);
                }
                copy ??= { ...input };
                copy[key] = result.value;
            }
        }
        for (const [key, propDef] of this.shapeValidators) {
            const result = ctx.validateChild(input, key, propDef);
            if (!result.success) {
                if (!(key in input)) {
                    // Transform into "required key" issue
                    return ctx.issueRequiredKey(input, key);
                }
                return result;
            }
            // Skip copying if key is not present in input (and value is undefined)
            if (result.value === undefined && !(key in input)) {
                continue;
            }
            if (!Object.is(result.value, input[key])) {
                if (ctx.options.mode === 'validate') {
                    // In "validate" mode, we can't modify the input, so we issue an error
                    return ctx.issueInvalidPropertyValue(input, key, [result.value]);
                }
                // Copy on write
                copy ??= { ...input };
                copy[key] = result.value;
            }
        }
        return ctx.success(copy ?? input);
    }
    fromURLSearchParams(urlSearchParams) {
        const params = {};
        for (const [key, value] of urlSearchParams.entries()) {
            const validator = this.shapeValidators.get(key);
            const coerced = validator instanceof string_js_1.StringSchema
                ? value
                : value === 'true'
                    ? true
                    : value === 'false'
                        ? false
                        : /^-?\d+$/.test(value)
                            ? Number(value)
                            : value;
            if (params[key] === undefined) {
                params[key] = coerced;
            }
            else if (Array.isArray(params[key])) {
                params[key].push(coerced);
            }
            else {
                params[key] = [params[key], coerced];
            }
        }
        return this.parse(params);
    }
    toURLSearchParams(input) {
        const urlSearchParams = new URLSearchParams();
        // @NOTE We apply defaults here to ensure that server with different
        // defaults still receive all expected parameters.
        const params = this.parse(input);
        for (const [key, value] of Object.entries(params)) {
            if (Array.isArray(value)) {
                for (const v of value) {
                    urlSearchParams.append(key, String(v));
                }
            }
            else if (value !== undefined) {
                urlSearchParams.append(key, String(value));
            }
        }
        return urlSearchParams;
    }
}
exports.ParamsSchema = ParamsSchema;
exports.params = (0, memoize_js_1.memoizedOptions)(function params(properties = {}) {
    return new ParamsSchema(properties);
});
