// @atproto/lex-schema/dist/schema/payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payload = void 0;
exports.payload = payload;
exports.jsonPayload = jsonPayload;
const object_js_1 = require("./object.js");
class Payload {
    encoding;
    schema;
    constructor(encoding, schema) {
        this.encoding = encoding;
        this.schema = schema;
        if (encoding === undefined && schema !== undefined) {
            throw new TypeError('schema cannot be defined when encoding is undefined');
        }
    }
    /**
     * Checks whether the given content-type matches the expected payload schema's
     * encoding.
     */
    matchesEncoding(contentType) {
        const mime = contentType?.split(';', 1)[0].trim();
        const { encoding } = this;
        // Handle undefined cases
        if (encoding === undefined) {
            // Expecting no body
            return mime === undefined;
        }
        else if (mime === undefined) {
            // Expecting a body, but got no content-type
            return false;
        }
        if (encoding === '*/*') {
            return true;
        }
        if (encoding.endsWith('/*')) {
            return mime.startsWith(encoding.slice(0, -1));
        }
        // Invalid: Lexicon can only specify "*/*" or "type/*" wildcards
        if (encoding.includes('*')) {
            return false;
        }
        return encoding === mime;
    }
}
exports.Payload = Payload;
/*@__NO_SIDE_EFFECTS__*/
function payload(encoding = undefined, validator = undefined) {
    return new Payload(encoding, validator);
}
/*@__NO_SIDE_EFFECTS__*/
function jsonPayload(properties) {
    return payload('application/json', (0, object_js_1.object)(properties));
}
