// @atproto/lex-schema/dist/schema/procedure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Procedure = void 0;
exports.procedure = procedure;
/**
 * Represents a Lexicon procedure (HTTP POST) endpoint definition.
 *
 * Procedures are operations that may modify state on the server.
 * They have parameters, an input payload (request body), an output
 * payload (response body), and optional error types.
 *
 * @template TNsid - The NSID identifying this procedure
 * @template TParameters - The parameters schema type
 * @template TInputPayload - The request body payload type
 * @template TOutputPayload - The response body payload type
 * @template TErrors - Array of error type strings, or undefined
 *
 * @example
 * ```ts
 * const createPost = new Procedure(
 *   'app.bsky.feed.post',
 *   l.params({}),
 *   l.jsonPayload({ text: l.string() }),
 *   l.jsonPayload({ uri: l.string(), cid: l.string() }),
 *   ['InvalidRecord']
 * )
 * ```
 */
class Procedure {
    nsid;
    parameters;
    input;
    output;
    errors;
    type = 'procedure';
    constructor(nsid, parameters, input, output, errors) {
        this.nsid = nsid;
        this.parameters = parameters;
        this.input = input;
        this.output = output;
        this.errors = errors;
    }
}
exports.Procedure = Procedure;
/**
 * Creates a procedure definition for a Lexicon POST endpoint.
 *
 * Procedures can modify server state. They accept both URL parameters
 * and a request body (input payload).
 *
 * @param nsid - The NSID identifying this procedure endpoint
 * @param parameters - Schema for URL query parameters
 * @param input - Schema for request body payload
 * @param output - Schema for response body payload
 * @param errors - Optional array of error type strings
 * @returns A new {@link Procedure} instance
 *
 * @example
 * ```ts
 * // Create record procedure
 * const createRecord = l.procedure(
 *   'com.atproto.repo.createRecord',
 *   l.params({}),
 *   l.jsonPayload({
 *     repo: l.string({ format: 'at-identifier' }),
 *     collection: l.string({ format: 'nsid' }),
 *     record: l.unknown(),
 *   }),
 *   l.jsonPayload({
 *     uri: l.string({ format: 'at-uri' }),
 *     cid: l.string({ format: 'cid' }),
 *   }),
 *   ['InvalidRecord', 'RepoNotFound'],
 * )
 *
 * // Procedure with binary input
 * const uploadBlob = l.procedure(
 *   'com.atproto.repo.uploadBlob',
 *   l.params({}),
 *   l.payload('*\/*', undefined), // Accept any content type
 *   l.jsonPayload({ blob: l.blob() }),
 * )
 * ```
 */
/*@__NO_SIDE_EFFECTS__*/
function procedure(nsid, parameters, input, output, errors = undefined) {
    return new Procedure(nsid, parameters, input, output, errors);
}
