// @atproto/lex-schema/dist/schema/query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Query = void 0;
exports.query = query;
/**
 * Represents a Lexicon query (HTTP GET) endpoint definition.
 *
 * Queries are read-only operations that retrieve data from a server.
 * They have parameters (passed as URL query parameters), an output
 * payload, and optional error types.
 *
 * @template TNsid - The NSID identifying this query
 * @template TParameters - The parameters schema type
 * @template TOutputPayload - The output payload type
 * @template TErrors - Array of error type strings, or undefined
 *
 * @example
 * ```ts
 * const getPostQuery = new Query(
 *   'app.bsky.feed.getPost',
 *   l.params({ uri: l.string({ format: 'at-uri' }) }),
 *   l.payload('application/json', postSchema),
 *   ['NotFound']
 * )
 * ```
 */
class Query {
    nsid;
    parameters;
    output;
    errors;
    type = 'query';
    constructor(nsid, parameters, output, errors) {
        this.nsid = nsid;
        this.parameters = parameters;
        this.output = output;
        this.errors = errors;
    }
}
exports.Query = Query;
/**
 * Creates a query definition for a Lexicon GET endpoint.
 *
 * Queries retrieve data without side effects. Parameters are sent as
 * URL query string parameters.
 *
 * @param nsid - The NSID identifying this query endpoint
 * @param parameters - Schema for URL query parameters
 * @param output - Expected response payload schema
 * @param errors - Optional array of error type strings
 * @returns A new {@link Query} instance
 *
 * @example
 * ```ts
 * // Simple query with JSON output
 * const getProfile = l.query(
 *   'app.bsky.actor.getProfile',
 *   l.params({ actor: l.string({ format: 'at-identifier' }) }),
 *   l.jsonPayload({ displayName: l.string(), handle: l.string() }),
 * )
 *
 * // Query with pagination and errors
 * const getTimeline = l.query(
 *   'app.bsky.feed.getTimeline',
 *   l.params({
 *     limit: l.optional(l.integer({ minimum: 1, maximum: 100 })),
 *     cursor: l.optional(l.string()),
 *   }),
 *   l.jsonPayload({ feed: l.array(feedItemSchema), cursor: l.optional(l.string()) }),
 *   ['BlockedActor', 'BlockedByActor'],
 * )
 * ```
 */
/*@__NO_SIDE_EFFECTS__*/
function query(nsid, parameters, output, errors = undefined) {
    return new Query(nsid, parameters, output, errors);
}
