// @atproto/lex-schema/dist/schema/subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Subscription = void 0;
exports.subscription = subscription;
/**
 * Represents a Lexicon subscription (WebSocket) endpoint definition.
 *
 * Subscriptions are real-time event streams delivered over WebSocket.
 * They have parameters for initializing the connection and a message
 * schema for validating incoming events.
 *
 * @template TNsid - The NSID identifying this subscription
 * @template TParameters - The connection parameters schema type
 * @template TMessage - The message schema type
 * @template TErrors - Array of error type strings, or undefined
 *
 * @example
 * ```ts
 * const firehose = new Subscription(
 *   'com.atproto.sync.subscribeRepos',
 *   l.params({ cursor: l.optional(l.integer()) }),
 *   repoEventSchema,
 *   ['FutureCursor']
 * )
 * ```
 */
class Subscription {
    nsid;
    parameters;
    message;
    errors;
    type = 'subscription';
    constructor(nsid, parameters, message, errors) {
        this.nsid = nsid;
        this.parameters = parameters;
        this.message = message;
        this.errors = errors;
    }
}
exports.Subscription = Subscription;
/**
 * Creates a subscription definition for a Lexicon WebSocket endpoint.
 *
 * Subscriptions enable real-time event streaming. The connection is
 * initialized with parameters, and the server sends messages matching
 * the message schema.
 *
 * @param nsid - The NSID identifying this subscription endpoint
 * @param parameters - Schema for connection parameters
 * @param message - Schema for validating incoming messages
 * @param errors - Optional array of error type strings
 * @returns A new {@link Subscription} instance
 *
 * @example
 * ```ts
 * // Repository event stream
 * const subscribeRepos = l.subscription(
 *   'com.atproto.sync.subscribeRepos',
 *   l.params({
 *     cursor: l.optional(l.integer()),
 *   }),
 *   l.typedUnion([
 *     l.typedRef(() => commitEventSchema),
 *     l.typedRef(() => handleEventSchema),
 *     l.typedRef(() => identityEventSchema),
 *   ], false),
 *   ['FutureCursor', 'ConsumerTooSlow'],
 * )
 *
 * // Label stream
 * const subscribeLabels = l.subscription(
 *   'com.atproto.label.subscribeLabels',
 *   l.params({ cursor: l.optional(l.integer()) }),
 *   labelEventSchema,
 * )
 * ```
 */
/*@__NO_SIDE_EFFECTS__*/
function subscription(nsid, parameters, message, errors = undefined) {
    return new Subscription(nsid, parameters, message, errors);
}
