// @atproto/lex-schema/dist/schema/typed-union.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypedUnionSchema = void 0;
exports.typedUnion = typedUnion;
const lex_data_1 = require("@atproto/lex-data");
const core_js_1 = require("../core.js");
const lazy_property_js_1 = require("../util/lazy-property.js");
class TypedUnionSchema extends core_js_1.Schema {
    validators;
    closed;
    constructor(validators, closed) {
        // @NOTE In order to avoid circular dependency issues, we don't access the
        // refs's schema (or $type) here. Instead, we access them lazily when first
        // needed.
        super();
        this.validators = validators;
        this.closed = closed;
    }
    get validatorsMap() {
        const map = new Map();
        for (const ref of this.validators)
            map.set(ref.$type, ref);
        return (0, lazy_property_js_1.lazyProperty)(this, 'validatorsMap', map);
    }
    get $types() {
        return Array.from(this.validatorsMap.keys());
    }
    validateInContext(input, ctx) {
        if (!(0, lex_data_1.isPlainObject)(input) || !('$type' in input)) {
            return ctx.issueInvalidType(input, '$typed');
        }
        const { $type } = input;
        const validator = this.validatorsMap.get($type);
        if (validator) {
            return ctx.validate(input, validator);
        }
        if (this.closed) {
            return ctx.issueInvalidPropertyValue(input, '$type', this.$types);
        }
        if (typeof $type !== 'string') {
            return ctx.issueInvalidPropertyType(input, '$type', 'string');
        }
        return ctx.success(input);
    }
}
exports.TypedUnionSchema = TypedUnionSchema;
/*@__NO_SIDE_EFFECTS__*/
function typedUnion(refs, closed) {
    return new TypedUnionSchema(refs, closed);
}
