// vue-data-ui/dist/A11yDataTable-KcVAX2cP.js
import { openBlock as t, createElementBlock as a, createElementVNode as l, toDisplayString as n, Fragment as o, renderList as u, renderSlot as s, createTextVNode as y } from "vue";
import { _ as h } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const m = ["id"], $ = { scope: "row" }, f = {
  __name: "A11yDataTable",
  props: {
    uid: { String, required: !0 },
    head: { Array, default: () => [] },
    body: { Array, default: () => [] },
    caption: { String, default: "Data table" },
    notice: { String, default: "A data table is available below." }
  },
  setup(e) {
    return (i, g) => (t(), a("div", {
      id: `chart-data-table-${e.uid}`,
      class: "sr-only",
      "data-dom-to-png-ignore": ""
    }, [
      l("p", null, n(e.notice), 1),
      l("table", null, [
        l("caption", null, n(e.caption), 1),
        l("thead", null, [
          l("tr", null, [
            (t(!0), a(o, null, u(e.head, (r, d) => (t(), a("th", {
              role: "cell",
              key: `a11y-head-${d}-${e.uid}`,
              scope: "col"
            }, [
              s(i.$slots, "th", { th: r }, () => [
                y(n(r), 1)
              ], !0)
            ]))), 128))
          ])
        ]),
        l("tbody", null, [
          (t(!0), a(o, null, u(e.body, (r, d) => (t(), a("tr", {
            key: `a11y-body-${d}-${e.uid}`
          }, [
            l("th", $, n(r[0]), 1),
            (t(!0), a(o, null, u(r.slice(1), (c, b) => (t(), a("td", {
              key: `a11y-cell-${d}-${b}-${e.uid}`
            }, [
              s(i.$slots, "td", { td: c }, () => [
                y(n(c), 1)
              ], !0)
            ]))), 128))
          ]))), 128))
        ])
      ])
    ], 8, m));
  }
}, S = /* @__PURE__ */ h(f, [["__scopeId", "data-v-3422f96b"]]);
export {
  S as A
};
