// vue-data-ui/dist/lib-DYekjHyv.js
import { unref as st, isRef as ot, toRaw as pt } from "vue";
const Ft = "#COMP# dataset prop is either missing, undefined or empty.", mt = "#COMP# dataset is missing the '#ATTR#' attribute.", Et = "#COMP# dataset '#ATTR#' attribute cannot be empty.", At = "#COMP# dataset #KEY# item at index #INDX# is missing the '#ATTR#' attribute.", xt = "#COMP# : Chart could not be built. Dataset is not formatted correctly", bt = "#COMP# : A wrong value was provided to the #ATTR# attribute (#KEY# is not a valid value).", yt = {
  dataset: Ft,
  datasetAttribute: mt,
  datasetAttributeEmpty: Et,
  datasetSerieAttribute: At,
  notBuildable: xt,
  attributeWrongValue: bt
};
function Jt(t, n, e, s, r, o = 1.99999, c = 2, i = 1.45, u = 360, l = 105.25, a = 0) {
  const { series: f } = t;
  if (!f || f.length === 0)
    return [];
  const h = f.reduce((p, E) => p + E.value, 0), d = [];
  let g = 0;
  for (let p = 0; p < f.length; p++) {
    const E = f[p].value, A = f.length === 1, F = A ? 1 : h > 0 ? E / h : 0, y = F * (Math.PI * o), L = (A ? 0.5 : h > 0 ? E / 2 / h : 0.5) * (Math.PI * c), { startX: S, startY: b, endX: $, endY: M, path: D } = q(
      [n, e],
      [s, r],
      [g, y],
      l,
      u,
      c
    ), m = q(
      [n, e],
      [s - a, r - a],
      [g, y],
      l,
      u,
      c,
      !0
    ), I = q(
      [n, e],
      [s * i, r * i],
      [g, L],
      l,
      u,
      c
    );
    d.push({
      arcSlice: `${D} L ${m.startX} ${m.startY} ${m.path} L ${S} ${b}`,
      cx: x(n),
      cy: x(e),
      ...f[p],
      proportion: x(F),
      ratio: x(y),
      path: D.replaceAll("NaN", "0"),
      startX: x(S),
      startY: x(b),
      endX: x($),
      endY: x(M),
      separator: {
        x: m.startX,
        y: m.startY
      },
      firstSeparator: {
        x: Number(m.path.split(" ").at(-2)),
        y: Number(m.path.split(" ").at(-1))
      },
      center: I
    }), g += y;
  }
  return d;
}
function ct([t, n], [e, s]) {
  return [t + e, n + s];
}
function at([[t, n], [e, s]], [r, o]) {
  return [t * r + n * o, e * r + s * o];
}
function Ct(t) {
  return [
    [Math.cos(t), -Math.sin(t)],
    [Math.sin(t), Math.cos(t)]
  ];
}
function q([t, n], [e, s], [r, o], c, i = 360, u = 2, l = !1) {
  o = o % (u * Math.PI);
  const a = Ct(c), [f, h] = ct(
    at(a, [
      e * Math.cos(r),
      s * Math.sin(r)
    ]),
    [t, n]
  ), [d, g] = ct(
    at(a, [
      e * Math.cos(r + o),
      s * Math.sin(r + o)
    ]),
    [t, n]
  ), p = o > Math.PI ? 1 : 0, E = o > 0 ? l ? 0 : 1 : l ? 1 : 0;
  return {
    startX: x(l ? d : f),
    startY: x(l ? g : h),
    endX: x(l ? f : d),
    endY: x(l ? h : g),
    path: `M${x(l ? d : f)} ${x(l ? g : h)} A ${[
      x(e),
      x(s),
      x(c / (u * Math.PI) * i),
      x(p),
      x(E),
      x(l ? f : d),
      x(l ? h : g)
    ].join(" ")}`
  };
}
function Mt({ defaultConfig: t, userConfig: n }) {
  const e = { ...t };
  return Object.keys(e).forEach((s) => {
    if (Object.hasOwn(n, s)) {
      const r = n[s];
      r === null ? e[s] = null : ["boolean", "function"].includes(typeof r) ? e[s] = r : ["string", "number"].includes(typeof r) ? it(r) && (e[s] = r) : Array.isArray(e[s]) ? $t({ userConfig: n, key: s }) && (e[s] = r) : Dt({ userConfig: n, key: s }) && (e[s] = Mt({
        defaultConfig: e[s],
        userConfig: r
      }));
    }
  }), Object.keys(n).forEach((s) => {
    if (!Object.hasOwn(e, s)) {
      const r = n[s];
      e[s] = r && typeof r == "object" && !Array.isArray(r) ? { ...r } : r;
    }
  }), e;
}
function $t({ userConfig: t, key: n }) {
  return Object.hasOwn(t, n) && Array.isArray(t[n]) && t[n].length >= 0;
}
function Dt({ userConfig: t, key: n }) {
  return Object.hasOwn(t, n) && !Array.isArray(t[n]) && typeof t[n] == "object";
}
function it(t) {
  return ![null, void 0, NaN, 1 / 0, -1 / 0].includes(t);
}
function tn(t) {
  return ![void 0, NaN, 1 / 0, -1 / 0].includes(t);
}
function nn(t, n = 0) {
  return it(t) ? t : n;
}
function x(t, n = 0) {
  return isNaN(t) ? n : t;
}
const ut = [
  "#1f77b4",
  "#aec7e8",
  "#ff7f0e",
  "#ffbb78",
  "#2ca02c",
  "#98df8a",
  "#d62728",
  "#ff9896",
  "#9467bd",
  "#c5b0d5",
  "#8c564b",
  "#c49c94",
  "#e377c2",
  "#f7b6d2",
  "#7f7f7f",
  "#c7c7c7",
  "#bcbd22",
  "#dbdb8d",
  "#17becf",
  "#9edae5",
  "#393b79",
  "#5254a3",
  "#6b6ecf",
  "#9c9ede",
  "#637939",
  "#8ca252",
  "#b5cf6b",
  "#cedb9c",
  "#8c6d31",
  "#bd9e39",
  "#e7ba52",
  "#e7cb94",
  "#843c39",
  "#ad494a",
  "#d6616b",
  "#e7969c",
  "#7b4173",
  "#a55194",
  "#ce6dbd",
  "#de9ed6"
];
function en(t = "default") {
  switch (t) {
    case "hack":
      return j.hack;
    case "zen":
      return j.zen;
    case "concrete":
      return j.concrete;
    case "celebration":
      return j.celebration;
    case "celebrationNight":
      return j.celebrationNight;
    case "minimal":
      return j.minimal;
    case "minimalDark":
      return j.minimalDark;
    default:
      return j.default;
  }
}
const j = {
  default: ut,
  dark: ut,
  minimal: [
    "#2A2929",
    "#454862",
    "#65698E",
    "#8D99AE",
    "#678681",
    "#7FA09B",
    "#9CBCA8",
    "#76645D",
    "#877675",
    "#A9998C",
    "#C6B7AB",
    "#906C70",
    "#B08C91",
    "#C9ACB0",
    "#9F816B",
    "#B39783",
    "#D8C3B3",
    "#825E76",
    "#9D7D92",
    "#C2A6B9"
  ],
  minimalDark: [
    "#524f4f",
    "#454862",
    "#65698E",
    "#8D99AE",
    "#678681",
    "#7FA09B",
    "#9CBCA8",
    "#76645D",
    "#877675",
    "#A9998C",
    "#C6B7AB",
    "#906C70",
    "#B08C91",
    "#C9ACB0",
    "#9F816B",
    "#B39783",
    "#D8C3B3",
    "#825E76",
    "#9D7D92",
    "#C2A6B9"
  ],
  celebration: [
    "#D32F2F",
    "#E64A19",
    "#F57C00",
    "#FF9800",
    "#FF5722",
    "#FFC107",
    "#FFEB3B",
    "#FFD54F",
    "#FF6F00",
    "#D84315",
    "#BF360C",
    "#C62828",
    "#B71C1C",
    "#FF7043",
    "#FF8A65",
    "#FFB74D",
    "#FFA726",
    "#FFCC80",
    "#FFE082",
    "#FFECB3"
  ],
  celebrationNight: [
    "#D32F2F",
    "#E64A19",
    "#F57C00",
    "#FF9800",
    "#FF5722",
    "#FFC947",
    "#FFEB3B",
    "#FFD95B",
    "#FF8800",
    "#FF5722",
    "#DD2C00",
    "#F44336",
    "#C62828",
    "#FF6E6E",
    "#FF867C",
    "#FFB547",
    "#FFA837",
    "#FFD180",
    "#FFE57F",
    "#FFF59D"
  ],
  concrete: [
    "#4A6A75",
    "#6C94A0",
    "#7DA9B5",
    "#8EBFCA",
    "#9FD4E0",
    "#B0E9F5",
    "#C1FFFF",
    "#5C6B5B",
    "#6D7D6D",
    "#7E8F7E",
    "#8FA290",
    "#A1B5A3",
    "#B2C7B5",
    "#C3DAC8",
    "#D4ECDA",
    "#E6FFF0",
    "#8A9CA5",
    "#9AA7B0",
    "#ABB1BC",
    "#BBCBC7",
    "#CCD6D3",
    "#DEE1DE",
    "#EFECEC",
    "#404C4D",
    "#50605F",
    "#617472",
    "#718885",
    "#829C98",
    "#92B0AB",
    "#A3C4BE",
    "#B3D8D2",
    "#C4EDE5",
    "#D4F1E8",
    "#404C5A",
    "#50606C",
    "#61747E",
    "#718890",
    "#829CA2",
    "#92B0B5"
  ],
  hack: [
    "#004C00",
    "#006600",
    "#008000",
    "#009900",
    "#00B300",
    "#00CC00",
    "#00E600",
    "#00FF00",
    "#33FF33",
    "#33E633",
    "#33CC33",
    "#33B333",
    "#339933",
    "#338033",
    "#336633",
    "#334C33",
    "#333333",
    "#00AF19",
    "#19E619",
    "#19CC19",
    "#19B319",
    "#199919",
    "#198019",
    "#196619",
    "#194C19",
    "#193319",
    "#191919",
    "#66FF66",
    "#66E666",
    "#66CC66",
    "#66B366",
    "#669966",
    "#668066",
    "#666666",
    "#4CFF4C",
    "#4CE64C",
    "#4CCC4C",
    "#4CB34C"
  ],
  zen: [
    "#B9B99D",
    "#E0CFC3",
    "#DFCA99",
    "#DCB482",
    "#C09E85",
    "#8F837A",
    "#858480",
    "#B0B9A8",
    "#606C5A",
    "#5E5E5E",
    "#4F5B75",
    "#647393",
    "#818EA9",
    "#9FA9BE",
    "#BBC4D3",
    "#DCDFE7",
    "#928A98",
    "#8A9892",
    "#B1A7AD",
    "#C5B8A7",
    "#EBD6CC",
    "#D7E0D2",
    "#E0D2D7",
    "#E0DBD2",
    "#D2E0DB",
    "#DBD2E0",
    "#C1B7A5",
    "#A5AFC1",
    "#E0DBD2",
    "#D2D7E0",
    "#F7EDE2",
    "#97ACB7",
    "#C4CBBC",
    "#C3C5C5",
    "#A0AC94"
  ]
}, lt = ["00", "03", "05", "08", "0A", "0D", "0F", "12", "14", "17", "1A", "1C", "1F", "21", "24", "26", "29", "2B", "2E", "30", "33", "36", "38", "3B", "3D", "40", "42", "45", "47", "4A", "4D", "4F", "52", "54", "57", "59", "5C", "5E", "61", "63", "66", "69", "6B", "6E", "70", "73", "75", "78", "7A", "7D", "80", "82", "85", "87", "8A", "8C", "8F", "91", "94", "96", "99", "9C", "9E", "A1", "A3", "A6", "A8", "AB", "AD", "B0", "B3", "B5", "B8", "BA", "BD", "BF", "C2", "C4", "C7", "C9", "CC", "CF", "D1", "D4", "D6", "D9", "DB", "DE", "E0", "E3", "E6", "E8", "EB", "ED", "F0", "F2", "F5", "F7", "FA", "FC", "FF"];
function gt(t) {
  const n = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})?$/i, e = /^#?([a-f\d])([a-f\d])([a-f\d])([a-f\d])?$/i, s = /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*([\d.]+))?\)$/i, r = /^hsla?\((\d+),\s*([\d.]+)%,\s*([\d.]+)%(?:,\s*([\d.]+))?\)$/i, o = /^oklch\(\s*([\d.]+)%?\s*[, ]\s*([\d.]+)%?\s*[, ]\s*([\d.]+)(?:deg)?\s*(?:\/\s*([\d.]+%?)\s*)?\)$/i, c = /^lch\(/i;
  if (t == null || typeof t == "number" && isNaN(t))
    return null;
  if (t = ot?.(t) ? st(t) : t, t = jt(t), Array.isArray(t)) {
    const [u, l, a, f = 1] = t;
    t = `rgba(${u},${l},${a},${f})`;
  } else if (typeof t == "object")
    if (Number.isFinite(t.r) && Number.isFinite(t.g) && Number.isFinite(t.b)) {
      const u = Number.isFinite(t.a) ? t.a : 1;
      t = `rgba(${t.r},${t.g},${t.b},${u})`;
    } else
      return null;
  else if (typeof t == "number") {
    const u = t >>> 0, l = u.toString(16).padStart(u <= 16777215 ? 6 : 8, "0");
    return `#${l.length === 6 ? l + "ff" : l}`;
  } else if (typeof t != "string")
    return null;
  if (t = t.trim(), c.test(t))
    return console.warn(
      "[convertColorToHex] lch() colors are not supported. Use oklch() instead."
    ), null;
  if (t.toLowerCase() === "transparent")
    return "#FFFFFF00";
  t = t.replace(e, (u, l, a, f, h) => `#${l}${l}${a}${a}${f}${f}${h ? h + h : ""}`);
  let i;
  if (i = t.match(n)) {
    const [, u, l, a, f] = i, h = f ? parseInt(f, 16) / 255 : 1;
    return `#${u}${l}${a}${G(Math.round(X(h) * 255))}`;
  }
  if (i = t.match(s)) {
    const [, u, l, a, f] = i, h = f ? parseFloat(f) : 1;
    return `#${G(u)}${G(l)}${G(a)}${G(Math.round(X(h) * 255))}`;
  }
  if (i = t.match(r)) {
    const [, u, l, a, f] = i, h = f ? parseFloat(f) : 1, [d, g, p] = Ot(Number(u), Number(l), Number(a));
    return `#${G(d)}${G(g)}${G(p)}${G(Math.round(X(h) * 255))}`;
  }
  if (i = t.match(o)) {
    const [, u, l, a, f] = i, h = St(u);
    if (h === null) return null;
    const d = It(l);
    if (d === null) return null;
    const g = Lt(a);
    if (g === null) return null;
    const p = Nt(f);
    if (p === null) return null;
    const [E, A, F] = Bt(h, d, g);
    return `#${G(E)}${G(A)}${G(F)}${G(Math.round(X(p) * 255))}`;
  }
  return null;
}
function X(t) {
  return Number.isFinite(t) ? t < 0 ? 0 : t > 1 ? 1 : t : 0;
}
function J(t) {
  return !Number.isFinite(t) || t < 0 ? 0 : t > 255 ? 255 : Math.round(t);
}
function tt(t) {
  return t <= 31308e-7 ? 12.92 * t : 1.055 * Math.pow(t, 1 / 2.4) - 0.055;
}
function Bt(t, n, e) {
  const s = (e % 360 + 360) % 360 * (Math.PI / 180), r = n * Math.cos(s), o = n * Math.sin(s);
  return Rt(t, r, o);
}
function Rt(t, n, e) {
  const s = t + 0.3963377774 * n + 0.2158037573 * e, r = t - 0.1055613458 * n - 0.0638541728 * e, o = t - 0.0894841775 * n - 1.291485548 * e, c = s * s * s, i = r * r * r, u = o * o * o;
  let l = 4.0767416621 * c - 3.3077115913 * i + 0.2309699292 * u, a = -1.2684380046 * c + 2.6097574011 * i - 0.3413193965 * u, f = -0.0041960863 * c - 0.7034186147 * i + 1.707614701 * u;
  l = X(l), a = X(a), f = X(f);
  const h = tt(l) * 255, d = tt(a) * 255, g = tt(f) * 255;
  return [J(h), J(d), J(g)];
}
function Nt(t) {
  if (t === void 0) return 1;
  if (typeof t == "string" && t.endsWith("%")) {
    const e = parseFloat(t);
    return Number.isFinite(e) ? X(e / 100) : null;
  }
  const n = parseFloat(t);
  return Number.isFinite(n) ? X(n) : null;
}
function St(t) {
  let n = Number(t);
  return Number.isFinite(n) ? (n > 1 && (n = n / 100), X(n)) : null;
}
function It(t) {
  let n = Number(t);
  return Number.isFinite(n) ? (n > 1 && (n = n / 100), n < 0 ? 0 : n) : null;
}
function Lt(t) {
  const n = Number(t);
  return Number.isFinite(n) ? n : null;
}
function G(t) {
  const n = Number(t).toString(16);
  return n.length === 1 ? "0" + n : n;
}
function Ot(t, n, e, s = 1) {
  t /= 360, n /= 100, e /= 100;
  let r, o, c;
  if (n === 0)
    r = o = c = e;
  else {
    const i = (a, f, h) => (h < 0 && (h += 1), h > 1 && (h -= 1), h < 0.16666666666666666 ? a + (f - a) * 6 * h : h < 0.5 ? f : h < 0.6666666666666666 ? a + (f - a) * (0.6666666666666666 - h) * 6 : a), u = e < 0.5 ? e * (1 + n) : e + n - e * n, l = 2 * e - u;
    r = i(l, u, t + 1 / 3), o = i(l, u, t), c = i(l, u, t - 1 / 3);
  }
  return [
    Math.round(r * 255),
    Math.round(o * 255),
    Math.round(c * 255),
    s
  ];
}
function rn(t, n) {
  const e = t.length === 9 ? t.substring(0, 7) : t, s = t.length === 9 ? t.substring(7, 9) : null, r = (f) => ({
    r: parseInt(f.substring(1, 3), 16),
    g: parseInt(f.substring(3, 5), 16),
    b: parseInt(f.substring(5, 7), 16)
  }), o = ({ r: f, g: h, b: d }) => {
    f /= 255, h /= 255, d /= 255;
    const g = Math.max(f, h, d), p = Math.min(f, h, d);
    let E, A, F = (g + p) / 2;
    if (g === p)
      E = A = 0;
    else {
      const y = g - p;
      switch (A = F > 0.5 ? y / (2 - g - p) : y / (g + p), g) {
        case f:
          E = (h - d) / y + (h < d ? 6 : 0);
          break;
        case h:
          E = (d - f) / y + 2;
          break;
        case d:
          E = (f - h) / y + 4;
          break;
      }
      E /= 6;
    }
    return { h: E, s: A, l: F };
  }, c = ({ h: f, s: h, l: d }) => {
    let g, p, E;
    if (h === 0)
      g = p = E = d;
    else {
      const A = (B, L, S) => (S < 0 && (S += 1), S > 1 && (S -= 1), S < 0.16666666666666666 ? B + (L - B) * 6 * S : S < 0.5 ? L : S < 0.6666666666666666 ? B + (L - B) * (0.6666666666666666 - S) * 6 : B), F = d < 0.5 ? d * (1 + h) : d + h - d * h, y = 2 * d - F;
      g = A(y, F, f + 1 / 3), p = A(y, F, f), E = A(y, F, f - 1 / 3);
    }
    return {
      r: Math.round(g * 255),
      g: Math.round(p * 255),
      b: Math.round(E * 255)
    };
  }, i = r(e || "#000000"), u = o(i);
  u.h += n, u.h = (u.h + 1) % 1;
  const l = c(u);
  return `#${(l.r << 16 | l.g << 8 | l.b).toString(16).padStart(6, "0")}` + (s || "");
}
function wt({
  centerX: t,
  centerY: n,
  outerPoints: e,
  radius: s,
  rotation: r
}) {
  const o = Math.PI / e, c = r;
  let i = "";
  const u = [];
  for (let l = 0; l < e * 2; l += 1) {
    let a = t + Math.cos(l * o + c) * s, f = n + Math.sin(l * o + c) * s;
    i += `${a},${f} `, u.push({ x: a, y: f });
  }
  return {
    path: `M${i}Z`,
    coordinates: u
  };
}
function sn({
  plot: t,
  radius: n,
  sides: e,
  rotation: s = 0
}) {
  const r = t.x, o = t.y, c = e / 2;
  return wt({
    centerX: r,
    centerY: o,
    outerPoints: c,
    radius: n + 1,
    rotation: s
  });
}
function Pt({
  centerX: t,
  centerY: n,
  innerCirclePoints: e,
  innerRadius: s,
  outerRadius: r
}) {
  const o = Math.PI / e, c = 60, i = e * 2;
  let u = "";
  for (let l = 0; l < i; l += 1) {
    let f = l % 2 == 0 ? r : s, h = t + Math.cos(l * o + c) * f, d = n + Math.sin(l * o + c) * f;
    u += `${h},${d} `;
  }
  return u;
}
function on({
  plot: t,
  radius: n,
  apexes: e = 5
}) {
  const s = t.x, r = t.y, o = e, c = n * 3.5 / o, u = c * 2;
  return Pt({
    centerX: s,
    centerY: r,
    innerCirclePoints: o,
    innerRadius: c,
    outerRadius: u
  });
}
function cn({ series: t }) {
  if (!Array.isArray(t) || t.length === 0) return "";
  const n = Array.from(
    new Map(
      t.filter((a) => a && Number.isFinite(a.x) && Number.isFinite(a.y)).map((a) => [`${a.x},${a.y}`, { x: +a.x, y: +a.y }])
    ).values()
  );
  if (n.length === 0) return "";
  if (n.length === 1) return `${Math.round(n[0].x)},${Math.round(n[0].y)} `;
  const e = (a, f) => {
    const h = a.x - f.x, d = a.y - f.y;
    return h * h + d * d;
  }, s = (a, f, h) => (f.x - a.x) * (h.y - a.y) - (f.y - a.y) * (h.x - a.x);
  let r = n[0];
  for (const a of n)
    (a.x < r.x || a.x === r.x && a.y < r.y) && (r = a);
  const o = [r];
  let c = r;
  const i = n.length + 2;
  let u = 0;
  for (; !(++u > i); ) {
    let a = n[0] === c ? n[1] : n[0];
    for (const f of n) {
      if (f === c || f === a) continue;
      const h = s(c, a, f);
      h < 0 || (h > 0 || e(c, f) > e(c, a)) && (a = f);
    }
    if (a === r) break;
    o.push(a), c = a;
  }
  let l = "";
  for (const a of o)
    l += `${Math.round(a.x)},${Math.round(a.y)} `;
  return l;
}
function W(t) {
  return t * Math.PI / 180;
}
function v(t, n, e) {
  return Math.min(Math.max(t, n), e);
}
function Tt(t) {
  const n = t.trim().match(/^rgba?\(\s*([+\-]?\d+)\s*,\s*([+\-]?\d+)\s*,\s*([+\-]?\d+)\s*(?:,\s*([+\-]?[\d.]+)\s*)?\)$/i);
  if (!n) return null;
  const e = v(Number.parseInt(n[1], 10), 0, 255), s = v(Number.parseInt(n[2], 10), 0, 255), r = v(Number.parseInt(n[3], 10), 0, 255), o = n[4] === void 0 ? 1 : v(Number.parseFloat(n[4]), 0, 1);
  return { red: e, green: s, blue: r, alpha: o };
}
function kt(t) {
  const n = t.trim();
  if (!n.startsWith("#")) return null;
  const e = n.slice(1);
  if (e.length === 3) {
    const s = Number.parseInt(e[0] + e[0], 16), r = Number.parseInt(e[1] + e[1], 16), o = Number.parseInt(e[2] + e[2], 16);
    return { red: s, green: r, blue: o, alpha: 1 };
  }
  if (e.length === 6) {
    const s = Number.parseInt(e.slice(0, 2), 16), r = Number.parseInt(e.slice(2, 4), 16), o = Number.parseInt(e.slice(4, 6), 16);
    return { red: s, green: r, blue: o, alpha: 1 };
  }
  if (e.length === 8) {
    const s = Number.parseInt(e.slice(0, 2), 16), r = Number.parseInt(e.slice(2, 4), 16), o = Number.parseInt(e.slice(4, 6), 16), c = Number.parseInt(e.slice(6, 8), 16) / 255;
    return { red: s, green: r, blue: o, alpha: c };
  }
  return null;
}
function Yt(t) {
  const n = t.trim().match(
    /^oklch\(\s*([+\-]?[\d.]+%?)\s+([+\-]?[\d.]+)\s+([+\-]?[\d.]+)(?:\s*\/\s*([+\-]?[\d.]+%?))?\s*\)$/i
  );
  if (!n) return null;
  const e = n[1], s = n[2], r = n[3], o = n[4], c = e.endsWith("%") ? v(Number.parseFloat(e) / 100, 0, 1) : v(Number.parseFloat(e), 0, 1), i = Math.max(0, Number.parseFloat(s)), u = Number.parseFloat(r), l = o ? o.endsWith("%") ? v(Number.parseFloat(o) / 100, 0, 1) : v(Number.parseFloat(o), 0, 1) : 1, a = u * Math.PI / 180, f = i * Math.cos(a), h = i * Math.sin(a), d = c + 0.3963377774 * f + 0.2158037573 * h, g = c - 0.1055613458 * f - 0.0638541728 * h, p = c - 0.0894841775 * f - 1.291485548 * h, E = d * d * d, A = g * g * g, F = p * p * p;
  let y = 4.0767416621 * E - 3.3077115913 * A + 0.2309699292 * F, B = -1.2684380046 * E + 2.6097574011 * A - 0.3413193965 * F, L = -0.0041960863 * E - 0.7034186147 * A + 1.707614701 * F;
  y = v(y, 0, 1), B = v(B, 0, 1), L = v(L, 0, 1);
  const S = (b) => {
    const $ = b <= 31308e-7 ? 12.92 * b : 1.055 * Math.pow(b, 0.4166666666666667) - 0.055;
    return v(Math.round($ * 255), 0, 255);
  };
  return {
    red: S(y),
    green: S(B),
    blue: S(L),
    alpha: l
  };
}
function Ht(t, n, e) {
  const s = t / 255, r = n / 255, o = e / 255, c = (a) => a <= 0.03928 ? a / 12.92 : Math.pow((a + 0.055) / 1.055, 2.4), i = c(s), u = c(r), l = c(o);
  return 0.2126 * i + 0.7152 * u + 0.0722 * l;
}
function an(t) {
  if (!t) return "#000000";
  const n = Tt(t), e = kt(t), s = Yt(t), r = n ?? e ?? s;
  if (!r)
    return "#000000";
  const o = Ht(r.red, r.green, r.blue);
  return (r.alpha < 1 ? r.alpha * o + (1 - r.alpha) * 1 : o) > 0.3 ? "#000000" : "#FFFFFF";
}
function V(t) {
  return t !== null && typeof t == "object" && Object.prototype.toString.call(t) === "[object Object]" && (t.constructor === Object || t.constructor == null);
}
function nt(t) {
  return !!t && (t.__v_isRef || t.__v_isReactive || t.__v_isReadonly || t.effect || t.dep || t.deps || t.subs);
}
function ft(t) {
  return t === "" ? "#000000" : t === "transparent" ? "#FFFFFF00" : gt(t) ?? t;
}
function et(t, n = /* @__PURE__ */ new WeakSet()) {
  const e = pt(t);
  if (!V(e) || n.has(e)) return e;
  n.add(e);
  for (const s in e) {
    const r = ot(e[s]) ? st(e[s]) : e[s];
    if (s === "color" || s === "backgroundColor") {
      typeof r == "string" && (e[s] = ft(r));
      continue;
    }
    if (s === "stroke") {
      typeof r == "string" ? e[s] = ft(r) : V(r) && !nt(r) && et(r, n);
      continue;
    }
    if (Array.isArray(r)) {
      for (const o of r)
        V(o) && !nt(o) && et(o, n);
      continue;
    }
    V(r) && !nt(r) && et(r, n);
  }
  return e;
}
function un(t) {
  const n = t?.length ?? 0;
  if (n < 2) return { x1: 0, y1: 0, x2: 0, y2: 0, slope: 0, trend: 0 };
  let e = 0, s = 0, r = 0, o = 0;
  for (const { x: M, y: D } of t)
    e += M, s += D, r += M * D, o += M * M;
  const c = n * o - e * e || 1, i = (n * r - e * s) / c, u = (s - i * e) / n, l = t[0].x, a = t[n - 1].x, f = i * l + u, h = i * a + u;
  let d = 0, g = 0, p = 0, E = 0;
  for (let M = 0; M < n; M += 1)
    d += M, g += t[M].value, p += M * t[M].value, E += M * M;
  const A = n * E - d * d || 1, F = (n * p - d * g) / A, y = (g - F * d) / n, B = y, L = F * (n - 1) + y, b = Math.max(Math.abs(B), Math.abs(g / n), Math.abs(L), 1e-9), $ = (L - B) / b;
  return { x1: l, y1: f, x2: a, y2: h, slope: i, trend: $ };
}
function ln(t) {
  const n = Math.floor(t.length / 2), e = [...t].sort((s, r) => s - r);
  return t.length % 2 !== 0 ? e[n] : (e[n - 1] + e[n]) / 2;
}
function fn(t) {
  let n = [];
  for (let e = 0; e < t.length; e += 1)
    n.push(`${x(t[e].x)},${x(t[e].y)} `);
  return n.join(" ").trim();
}
function hn(t) {
  if (t.length < 2) return "0,0";
  const n = t.length - 1, e = [`${x(t[0].x)},${x(t[0].y)}`], s = [], r = [], o = [], c = [];
  for (let i = 0; i < n; i += 1)
    s[i] = t[i + 1].x - t[i].x, r[i] = t[i + 1].y - t[i].y, o[i] = r[i] / s[i];
  c[0] = o[0], c[n] = o[n - 1];
  for (let i = 1; i < n; i += 1)
    if (o[i - 1] * o[i] <= 0)
      c[i] = 0;
    else {
      const u = 2 * o[i - 1] * o[i] / (o[i - 1] + o[i]);
      c[i] = u;
    }
  for (let i = 0; i < n; i += 1) {
    const u = t[i].x, l = t[i].y, a = t[i + 1].x, f = t[i + 1].y, h = c[i], d = c[i + 1], g = u + (a - u) / 3, p = l + h * (a - u) / 3, E = a - (a - u) / 3, A = f - d * (a - u) / 3;
    e.push(`C ${x(g)},${x(p)} ${x(E)},${x(A)} ${x(a)},${x(f)}`);
  }
  return e.join(" ");
}
function dn(t, n = 0.2) {
  function e(c, i) {
    const u = i.x - c.x, l = i.y - c.y;
    return {
      length: Math.sqrt(Math.pow(u, 2) + Math.pow(l, 2)),
      angle: Math.atan2(l, u)
    };
  }
  function s(c, i, u, l) {
    const h = e(i || c, u || c), d = h.angle + (l ? Math.PI : 0), g = h.length * n, p = c.x + Math.cos(d) * g, E = c.y + Math.sin(d) * g;
    return { x: p, y: E };
  }
  function r(c, i, u) {
    const l = s(u[i - 1], u[i - 2], c), a = s(c, u[i - 1], u[i + 1], !0);
    return `C ${x(l.x)},${x(l.y)} ${x(a.x)},${x(a.y)} ${x(c.x)},${x(c.y)}`;
  }
  return t.filter((c) => !!c).reduce(
    (c, i, u, l) => u === 0 ? `${x(i.x)},${x(i.y)} ` : `${c} ${r(i, u, l)} `,
    ""
  );
}
function gn() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(t) {
    const n = Math.random() * 16 | 0;
    return (t == "x" ? n : n & 3 | 8).toString(16);
  });
}
function _(t, n = !1, e = 16, s = !1) {
  let r = 0, o = n ? e : 0, c = s ? "center" : "middle";
  return t.center.endX > t.cx ? (r = t.center.endX + e + o, c = s ? "left" : "start") : t.center.endX < t.cx ? (r = t.center.endX - e - o, c = s ? "right" : "end") : (r = t.centerX + o, c = s ? "center" : "middle"), { x: r, anchor: c };
}
function z(t, n = 16, e = 16) {
  return t.center.endY > t.cy ? t.center.endY + e : t.center.endY < t.cy ? t.center.endY - n : t.center.endY;
}
function pn({
  initX: t,
  initY: n,
  offset: e,
  centerX: s,
  centerY: r
}) {
  const o = Math.atan2(n - r, t - s);
  return {
    x: t + e * Math.cos(o),
    y: n + e * Math.sin(o)
  };
}
function Ut(t) {
  const n = document.createElementNS("http://www.w3.org/2000/svg", "path");
  n.setAttribute("d", t);
  const e = n.getTotalLength();
  let s = 0, r = e, o = e / 2;
  const c = 0.01;
  for (; r - s > c; ) {
    const l = (s + r) / 2, f = n.getPointAtLength(l).x;
    if (Math.abs(f - o) < c) {
      o = l;
      break;
    } else f < o ? s = l : r = l;
  }
  const { x: i, y: u } = n.getPointAtLength(o);
  return { x: i, y: u };
}
function Fn(t, n = !1, e = 16, s = 16, r = !1, o = !1, c = 0, i = 12, u = !1) {
  const { x: l } = Ut(t.path), a = `${_(t).x},${z(t, e, s) - 4} `;
  let f = "", h, d;
  l > t.cx ? (h = _(t).x - i, d = z(t, e, s) - 4, f = `${h},${d}`) : l < t.cx ? (h = _(t).x + i, d = z(t, e, s) - 4, f = `${h},${d}`) : (h = _(t).x + i, d = z(t, e, s) - 4, f = `${h},${d}`);
  const g = ` ${t.center.endX},${t.center.endY}`;
  return u ? o ? `M${f} Q${f} ${g}` : `M${a} Q${f} ${g}` : `M${o ? "" : a}${f}${g}`;
}
function mn(t) {
  return ![null, void 0, NaN].includes(t);
}
function En(t, n) {
  return [...t].map((e) => e[n]).reduce((e, s) => e + s, 0);
}
function An(t, n = !0, e = !1) {
  if (!t.length) return "M0,0";
  let s = "";
  return t.forEach((r) => {
    if (!r) return "";
    s += `${r.x},${r.y} `;
  }), e ? s.trim() : `M${s}${n ? "Z" : ""}`;
}
function xn({ csvContent: t, title: n = "vue-data-ui" }) {
  const e = encodeURI(t), s = document.createElement("a");
  s.setAttribute("href", e), s.setAttribute("download", `${n}.csv`), document.body.appendChild(s), s.click(), s.remove(), window.URL.revokeObjectURL(e);
}
function bn(t) {
  return `data:text/csv;charset=utf-8,${t.map((n) => n.join(",")).join(`
`)}`;
}
function yn(t, n) {
  if (!/^#([0-9A-F]{3}){1,2}([0-9A-F]{2})?$/i.test(t))
    return console.warn("lightenHexColor : Invalid hex color format"), "#000000";
  let e = t.replace("#", "");
  e.length === 3 && (e = e.split("").map((a) => a + a).join(""));
  const s = parseInt(e.substring(0, 2), 16), r = parseInt(e.substring(2, 4), 16), o = parseInt(e.substring(4, 6), 16), c = Math.min(255, s + (255 - s) * n), i = Math.min(255, r + (255 - r) * n), u = Math.min(255, o + (255 - o) * n), l = `#${Math.round(c).toString(16).padStart(2, "0")}${Math.round(i).toString(16).padStart(2, "0")}${Math.round(u).toString(16).padStart(2, "0")}`;
  if (e.length === 8) {
    const a = e.substring(6, 8);
    return l + a;
  }
  return l;
}
function Cn(t, n) {
  if (!/^#([0-9A-F]{3}){1,2}([0-9A-F]{2})?$/i.test(t))
    return console.warn("darkenHexColor: Invalid hex color format"), "#000000";
  let e = t.replace("#", "");
  e.length === 3 && (e = e.split("").map((a) => a + a).join(""));
  const s = parseInt(e.substring(0, 2), 16), r = parseInt(e.substring(2, 4), 16), o = parseInt(e.substring(4, 6), 16), c = Math.max(0, s - s * n), i = Math.max(0, r - r * n), u = Math.max(0, o - o * n), l = `#${Math.round(c).toString(16).padStart(2, "0")}${Math.round(i).toString(16).padStart(2, "0")}${Math.round(u).toString(16).padStart(2, "0")}`;
  if (e.length === 8) {
    const a = e.substring(6, 8);
    return l + a;
  }
  return l;
}
function Q(t, n) {
  const e = Math.floor(Math.log10(t)), s = t / Math.pow(10, e);
  let r;
  return n ? s < 1.5 ? r = 1 : s < 3 ? r = 2 : s < 7 ? r = 5 : r = 10 : s <= 1 ? r = 1 : s <= 2 ? r = 2 : s <= 5 ? r = 5 : r = 10, r * Math.pow(10, e);
}
function Mn(t, n, e, s = !1) {
  const r = s ? n - t : Q(n - t, !1), o = s ? r / (e - 1) : Q(r / (e - 1), !0), c = Math.floor(t / o) * o, i = Math.ceil(n / o) * o, u = [];
  for (let l = c; l <= i; l += o)
    u.push(l);
  return {
    min: c,
    max: i,
    tickSize: o,
    ticks: u
  };
}
function $n(t, n, e, s = !1) {
  const r = s ? n - t : Q(n - t, !1), o = s ? r / (e - 1) : Q(r / (e - 1), !0), c = Math.floor(t / o) * o, i = Math.ceil(n / o) * o;
  let u = [], l = c;
  for (; l <= i; )
    l >= t && l <= n && u.push(l), l += o;
  return u[0] !== t && (u[0] = t), u[u.length - 1] !== n && (u[u.length - 1] = n), {
    min: t,
    max: n,
    tickSize: o,
    ticks: u
  };
}
function Dn(t, n, e, s, r) {
  const o = (h) => {
    let d = parseInt(h.substring(1, 3), 16), g = parseInt(h.substring(3, 5), 16), p = parseInt(h.substring(5, 7), 16), E = 1;
    return h.length === 9 && (E = parseInt(h.substring(7, 9), 16) / 255), { r: d, g, b: p, a: E };
  }, c = ({ r: h, g: d, b: g, a: p }) => {
    const E = (F) => F.toString(16).padStart(2, "0"), A = `#${E(h)}${E(d)}${E(g)}`;
    if (p !== 1) {
      const F = Math.round(p * 255).toString(16).padStart(2, "0");
      return A + F;
    }
    return A;
  }, i = o(t), u = o(n);
  r = Math.min(Math.max(r, e), s);
  const l = (r - e) / (s - e), a = {
    r: Math.round(i.r + (u.r - i.r) * l),
    g: Math.round(i.g + (u.g - i.g) * l),
    b: Math.round(i.b + (u.b - i.b) * l)
  }, f = i.a + (u.a - i.a) * l;
  return c({ ...a, a: f });
}
function Bn({ p: t = "", v: n, s: e = "", r: s = 0, space: r = !1, isAnimating: o = !1, regex: c = /[^%]/g, replacement: i = "-", locale: u = null }) {
  const l = u ? Number(Number(n).toFixed(s)).toLocaleString(u) : Number(Number(n).toFixed(s)).toLocaleString(), a = l === 1 / 0 ? "∞" : l === -1 / 0 ? "-∞" : l, f = `${t ?? ""}${r ? " " : ""}${[void 0, null].includes(n) ? "-" : a}${r ? " " : ""}${e ?? ""}`;
  return o ? f.replace(c, i) : f;
}
function Rn({ source: t, length: n = 3 }) {
  if (!t && t !== 0)
    return "";
  t = String(t);
  const e = t.length > 1 ? t.split(" ") : [t];
  if (e.length === 1 && e[0].length === 1)
    return String(t).toUpperCase();
  if (e.length === 1)
    return t.slice(0, n).toUpperCase();
  {
    const s = [];
    return e.forEach((r, o) => {
      o < n && s.push(r.slice(0, 1));
    }), s.join().replaceAll(",", "").toUpperCase();
  }
}
function Nn(t) {
  return !!t && typeof t == "function";
}
function Sn(t) {
  return typeof t.apply(null, arguments) == "string";
}
function In(t) {
  return Array.isArray(t) ? t.length === 0 : t ? Object.keys(t).length === 0 : !0;
}
function Ln({ componentName: t, type: n, property: e = "", index: s = "", key: r = "", warn: o = !0, debug: c = !0 }) {
  if (!c) return;
  const i = `
> ${yt[n].replace("#COMP#", t).replace("#ATTR#", e).replace("#INDX#", s).replace("#KEY#", r)}
`;
  if (o)
    console.warn(i);
  else
    throw new Error(i);
}
function Gt({ points: t, a: n, b: e, angleStep: s, startX: r, startY: o }) {
  const c = [];
  for (let i = 0; i < t; i++) {
    const u = s * i, l = n + e * u, a = l * Math.cos(u) + r, f = l * Math.sin(u) + o;
    c.push({ x: a, y: f });
  }
  return c;
}
function vt(t) {
  let n = 1 / 0, e = 1 / 0, s = -1 / 0, r = -1 / 0;
  for (const o of t)
    o.x < n && (n = o.x), o.y < e && (e = o.y), o.x > s && (s = o.x), o.y > r && (r = o.y);
  return { minX: n, minY: e, maxX: s, maxY: r, width: s - n || 1, height: r - e || 1 };
}
function Xt(t) {
  if (!t.length) return "";
  let n = `M${t[0].x} ${t[0].y}`;
  for (let e = 1; e < t.length - 2; e += 2) {
    const s = t[e - 1], r = t[e], o = t[e + 1], c = t[e + 2], i = (s.x + r.x) / 2, u = (s.y + r.y) / 2, l = (r.x + o.x) / 2, a = (r.y + o.y) / 2, f = (o.x + c.x) / 2, h = (o.y + c.y) / 2;
    n += ` C${i} ${u}, ${l} ${a}, ${f} ${h}`;
  }
  return n;
}
function On({
  maxPoints: t,
  a: n = 6,
  b: e = 6,
  angleStep: s = 0.07,
  startX: r,
  startY: o,
  boxWidth: c,
  boxHeight: i,
  padding: u = 12
}) {
  const l = Gt({ points: t, a: n, b: e, angleStep: s, startX: 0, startY: 0 }), { minX: a, minY: f, maxX: h, maxY: d, width: g, height: p } = vt(l), E = (a + h) / 2, A = (f + d) / 2, F = Math.max(1, c - 2 * u), y = Math.max(1, i - 2 * u), B = Math.min(F / g, y / p), L = r - E * B, S = o - A * B;
  return function($) {
    const M = Math.max(2, Math.min(Math.round($), l.length)), D = l.slice(0, M).map((m) => ({ x: m.x * B + L, y: m.y * B + S }));
    return Xt(D);
  };
}
function wn({ datasetObject: t, requiredAttributes: n }) {
  let e = [];
  return n.forEach((s) => {
    Object.hasOwn(t, s) || e.push(s);
  }), e;
}
const ht = {
  ALICEBLUE: "#F0F8FF",
  ANTIQUEWHITE: "#FAEBD7",
  AQUA: "#00FFFF",
  AQUAMARINE: "#7FFFD4",
  AZURE: "#F0FFFF",
  BEIGE: "#F5F5DC",
  BISQUE: "#FFE4C4",
  BLACK: "#000000",
  BLANCHEDALMOND: "#FFEBCD",
  BLUE: "#0000FF",
  BLUEVIOLET: "#8A2BE2",
  BROWN: "#A52A2A",
  BURLYWOOD: "#DEB887",
  CADETBLUE: "#5F9EA0",
  CHARTREUSE: "#7FFF00",
  CHOCOLATE: "#D2691E",
  CORAL: "#FF7F50",
  CORNFLOWERBLUE: "#6495ED",
  CORNSILK: "#FFF8DC",
  CRIMSON: "#DC143C",
  CYAN: "#00FFFF",
  DARKBLUE: "#00008B",
  DARKCYAN: "#008B8B",
  DARKGOLDENROD: "#B8860B",
  DARKGREY: "#A9A9A9",
  DARKGREEN: "#006400",
  DARKKHAKI: "#BDB76B",
  DARKMAGENTA: "#8B008B",
  DARKOLIVEGREEN: "#556B2F",
  DARKORANGE: "#FF8C00",
  DARKORCHID: "#9932CC",
  DARKRED: "#8B0000",
  DARKSALMON: "#E9967A",
  DARKSEAGREEN: "#8FBC8F",
  DARKSLATEBLUE: "#483D8B",
  DARKSLATEGREY: "#2F4F4F",
  DARKTURQUOISE: "#00CED1",
  DARKVIOLET: "#9400D3",
  DEEPPINK: "#FF1493",
  DEEPSKYBLUE: "#00BFFF",
  DIMGRAY: "#696969",
  DODGERBLUE: "#1E90FF",
  FIREBRICK: "#B22222",
  FLORALWHITE: "#FFFAF0",
  FORESTGREEN: "#228B22",
  FUCHSIA: "#FF00FF",
  GAINSBORO: "#DCDCDC",
  GHOSTWHITE: "#F8F8FF",
  GOLD: "#FFD700",
  GOLDENROD: "#DAA520",
  GREY: "#808080",
  GREEN: "#008000",
  GREENYELLOW: "#ADFF2F",
  HONEYDEW: "#F0FFF0",
  HOTPINK: "#FF69B4",
  INDIANRED: "#CD5C5C",
  INDIGO: "#4B0082",
  IVORY: "#FFFFF0",
  KHAKI: "#F0E68C",
  LAVENDER: "#E6E6FA",
  LAVENDERBLUSH: "#FFF0F5",
  LAWNGREEN: "#7CFC00",
  LEMONCHIFFON: "#FFFACD",
  LIGHTBLUE: "#ADD8E6",
  LIGHTCORAL: "#F08080",
  LIGHTCYAN: "#E0FFFF",
  LIGHTGOLDENRODYELLOW: "#FAFAD2",
  LIGHTGREY: "#D3D3D3",
  LIGHTGREEN: "#90EE90",
  LIGHTPINK: "#FFB6C1",
  LIGHTSALMON: "#FFA07A",
  LIGHTSEAGREEN: "#20B2AA",
  LIGHTSKYBLUE: "#87CEFA",
  LIGHTSLATEGREY: "#778899",
  LIGHTSTEELBLUE: "#B0C4DE",
  LIGHTYELLOW: "#FFFFE0",
  LIME: "#00FF00",
  LIMEGREEN: "#32CD32",
  LINEN: "#FAF0E6",
  MAGENTA: "#FF00FF",
  MAROON: "#800000",
  MEDIUMAQUAMARINE: "#66CDAA",
  MEDIUMBLUE: "#0000CD",
  MEDIUMORCHID: "#BA55D3",
  MEDIUMPURPLE: "#9370D8",
  MEDIUMSEAGREEN: "#3CB371",
  MEDIUMSLATEBLUE: "#7B68EE",
  MEDIUMSPRINGGREEN: "#00FA9A",
  MEDIUMTURQUOISE: "#48D1CC",
  MEDIUMVIOLETRED: "#C71585",
  MIDNIGHTBLUE: "#191970",
  MINTCREAM: "#F5FFFA",
  MISTYROSE: "#FFE4E1",
  MOCCASIN: "#FFE4B5",
  NAVAJOWHITE: "#FFDEAD",
  NAVY: "#000080",
  OLDLACE: "#FDF5E6",
  OLIVE: "#808000",
  OLIVEDRAB: "#6B8E23",
  ORANGE: "#FFA500",
  ORANGERED: "#FF4500",
  ORCHID: "#DA70D6",
  PALEGOLDENROD: "#EEE8AA",
  PALEGREEN: "#98FB98",
  PALETURQUOISE: "#AFEEEE",
  PALEVIOLETRED: "#D87093",
  PAPAYAWHIP: "#FFEFD5",
  PEACHPUFF: "#FFDAB9",
  PERU: "#CD853F",
  PINK: "#FFC0CB",
  PLUM: "#DDA0DD",
  POWDERBLUE: "#B0E0E6",
  PURPLE: "#800080",
  RED: "#FF0000",
  ROSYBROWN: "#BC8F8F",
  ROYALBLUE: "#4169E1",
  SADDLEBROWN: "#8B4513",
  SALMON: "#FA8072",
  SANDYBROWN: "#F4A460",
  SEAGREEN: "#2E8B57",
  SEASHELL: "#FFF5EE",
  SIENNA: "#A0522D",
  SILVER: "#C0C0C0",
  SKYBLUE: "#87CEEB",
  SLATEBLUE: "#6A5ACD",
  SLATEGREY: "#708090",
  SNOW: "#FFFAFA",
  SPRINGGREEN: "#00FF7F",
  STEELBLUE: "#4682B4",
  TAN: "#D2B48C",
  TEAL: "#008080",
  THISTLE: "#D8BFD8",
  TOMATO: "#FF6347",
  TURQUOISE: "#40E0D0",
  VIOLET: "#EE82EE",
  WHEAT: "#F5DEB3",
  WHITE: "#FFFFFF",
  WHITESMOKE: "#F5F5F5",
  YELLOW: "#FFFF00",
  YELLOWGREEN: "#9ACD32",
  REBECCAPURPLE: "#663399"
};
function jt(t) {
  const n = ot?.(t) ? st(t) : t;
  if (typeof n != "string") return n;
  const e = n.trim();
  if (e === "" || e[0] === "#") return e;
  if (e.toLowerCase() === "transparent") return "#FFFFFF00";
  const s = e.toUpperCase(), r = s.replace(/GRAY/g, "GREY");
  return ht[s] || ht[r] || e;
}
const rt = "http://www.w3.org/2000/svg";
function Pn(t) {
  if (t.length < 2)
    return 0;
  let n = 0, e = 0;
  for (let r = 1; r < t.length; r++) {
    const o = t[r - 1], c = t[r];
    if ([null, void 0, 0, 1 / 0, -1 / 0].includes(o))
      continue;
    const i = (c - o) / Math.abs(o) * 100;
    n += i, e++;
  }
  return e === 0 ? 0 : n / e;
}
function Wt({
  content: t,
  fontSize: n,
  fill: e,
  x: s,
  y: r,
  translateY: o = !1
}) {
  const c = t.split(`
`), u = c.length * n, l = o ? (u - n) / 2 : 0;
  return c.map(
    (a, f) => `<tspan x="${s}" y="${r - l + f * n}" fill="${e}">${a}</tspan>`
  ).join("");
}
function Tn(t) {
  return it(t) ? (Wt({
    content: t,
    fontSize: 1,
    fill: "",
    x: 0,
    y: 0
  }).match(/<tspan\b/g) || []).length : 1;
}
function kn({ content: t, fontSize: n, fill: e, x: s, autoOffset: r = !1 }) {
  const o = t.split(`
`), c = r ? (o.length - 1) * n / 2 : 0;
  return o.map((i, u) => {
    const l = u === 0 ? -c : n;
    return `<tspan x="${s}" dy="${l}" fill="${e}">${i}</tspan>`;
  }).join("");
}
function Yn({
  content: t,
  fontSize: n,
  fill: e,
  maxWords: s,
  x: r,
  y: o
}) {
  function c(l, a) {
    const f = l.split(" "), h = [];
    for (let d = 0; d < f.length; d += a)
      h.push(f.slice(d, d + a).join(" "));
    return h;
  }
  let i = "";
  return c(t, s).forEach((l, a) => {
    const f = `<tspan x="${r}" y="${o + a * n}" fill="${e}">${l}</tspan>`;
    i += f;
  }), i;
}
function Hn(t) {
  return t.length ? t.map((n) => gt(n)) : [];
}
function Un(t, n = null) {
  const e = t.replace(/[\p{P}\p{S}]+/gu, " ").trim(), c = (/[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}\p{Script=Hangul}\p{Script=Thai}\p{Script=Lao}\p{Script=Khmer}\p{Script=Tibetan}\p{Script=Myanmar}\p{Script=Devanagari}]/u.test(t) ? [...e] : e.split(/\s+/)).filter((i) => i.trim().length > 0).reduce((i, u) => (i[u] ? i[u] += 1 : i[u] = 1, i), {});
  return Object.keys(c).map((i) => {
    let u = i;
    return typeof n == "function" && typeof n(i) == "string" && (u = n(i)), {
      name: u,
      value: c[i]
    };
  });
}
function Gn(t) {
  let n = t.reduce((i, u) => i + (u.stackRatio || 0), 0), e = t.filter((i) => i.stackRatio === void 0).length, s = 1 - n, r = e > 0 ? s / e : 0, o = t.map((i) => ({
    ...i,
    stackRatio: i.stackRatio !== void 0 ? i.stackRatio : r
  })), c = 0;
  return o = o.map((i, u) => (c += i.stackRatio, {
    ...i,
    stackIndex: u,
    cumulatedStackRatio: c
  })), o;
}
function vn(t) {
  function n(l, a, f, h) {
    const d = f - l, g = h - a;
    return Math.sqrt(d * d + g * g);
  }
  function e(l, a, f, h) {
    let g = 0, p = l.x, E = l.y;
    for (let A = 1; A <= 100; A += 1) {
      const F = A / 100, y = 1 - F, B = y * y, L = F * F, S = B * y * l.x + 3 * B * F * a.x + 3 * y * L * f.x + L * F * h.x, b = B * y * l.y + 3 * B * F * a.y + 3 * y * L * f.y + L * F * h.y;
      g += n(p, E, S, b), p = S, E = b;
    }
    return g;
  }
  const s = t.match(/[a-zA-Z][^a-zA-Z]*/g);
  let r = 0, o = 0, c = 0, i = 0, u = 0;
  return s.forEach((l) => {
    const a = l[0], f = l.slice(1).trim().split(/[\s,]+/).map(Number);
    let h = 0;
    switch (a) {
      case "M":
        for (r = f[h++], o = f[h++], c = r, i = o; h < f.length; )
          u += n(r, o, f[h], f[h + 1]), r = f[h++], o = f[h++];
        break;
      case "L":
        for (; h < f.length; )
          u += n(r, o, f[h], f[h + 1]), r = f[h++], o = f[h++];
        break;
      case "H":
        for (; h < f.length; )
          u += n(r, o, f[h], o), r = f[h++];
        break;
      case "V":
        for (; h < f.length; )
          u += n(r, o, r, f[h]), o = f[h++];
        break;
      case "C":
        for (; h < f.length; )
          u += e(
            { x: r, y: o },
            { x: f[h], y: f[h + 1] },
            { x: f[h + 2], y: f[h + 3] },
            { x: f[h + 4], y: f[h + 5] }
          ), r = f[h + 4], o = f[h + 5], h += 6;
        break;
      case "Z":
        u += n(r, o, c, i), r = c, o = i;
        break;
    }
  }), u;
}
function Xn({
  relator: t,
  adjuster: n,
  source: e,
  threshold: s = 0,
  fallback: r,
  max: o = 24
}) {
  const c = t / (n / e);
  return c > o ? o : c < s ? r : c;
}
function jn(t) {
  return t.reduce((n, e) => (e.series.forEach((s, r) => {
    ![void 0, null, 1 / 0, -1 / 0].includes(s) && !isNaN(s) && (n[r] = (n[r] || 0) + s);
  }), n), []);
}
function Kt(t, { value: n, config: e }) {
  let s = !1, r = n;
  if (typeof t == "function")
    try {
      r = t({ value: n, config: e }), ["number", "string"].includes(typeof r) ? s = !0 : r = n;
    } catch (o) {
      console.warn("Formatter could not be applied:", o), s = !1;
    }
  return {
    isValid: s,
    value: r
  };
}
function Wn(t, n, e, s) {
  const { isValid: r, value: o } = Kt(t, { value: n, config: s });
  return r ? o : e;
}
function Kn(t, n) {
  return n.split(".").every((e) => t !== null && typeof t == "object" && e in t ? (t = t[e], !0) : !1);
}
function Vn(t, n = [], e = !1) {
  function s(o) {
    return e && o === null ? null : typeof o == "string" && isNaN(Number(o)) || typeof o == "number" && isFinite(o) ? o : 0;
  }
  function r(o) {
    if (Array.isArray(o))
      return o.map((c) => r(c));
    if (typeof o == "object" && o !== null) {
      let c = { ...o };
      return n.forEach((i) => {
        c.hasOwnProperty(i) && ![
          "NAME",
          "name",
          "TITLE",
          "title",
          "DESCRIPTION",
          "description",
          "LABEL",
          "label",
          "TIME",
          "time",
          "PERIOD",
          "period",
          "MONTH",
          "month",
          "YEAR",
          "year",
          "MONTHS",
          "months",
          "YEARS",
          "years",
          "DAY",
          "day",
          "DAYS",
          "days",
          "HOUR",
          "hour",
          "HOURS",
          "hours"
        ].includes(i) && Array.isArray(c[i]) && (c[i] = r(c[i]));
      }), Object.fromEntries(
        Object.entries(c).map(([i, u]) => [i, r(u)])
      );
    } else
      return s(o);
  }
  return r(t);
}
function _n(t, n = 100) {
  return t.length === 9 ? t.substring(0, 7) + lt[n] : t + lt[n];
}
function zn({ series: t, center: n, maxRadius: e, hasGhost: s = !1 }) {
  const o = 360 / (t.length - (s ? 1 : 0));
  return t.map((i, u) => {
    const a = i * e, f = u * o, h = f + o, d = f + o / 2, g = W(f) - W(90), p = W(h) - W(90), E = W(d) - W(90), A = n.x + a * Math.cos(g), F = n.y + a * Math.sin(g), y = n.x + a * Math.cos(p), B = n.y + a * Math.sin(p), L = n.x + a * Math.cos(E), S = n.y + a * Math.sin(E);
    return {
      path: `
            M ${n.x} ${n.y} 
            L ${A} ${F} 
            A ${a} ${a} 0 0 1 ${y} ${B} 
            Z
        `.trim(),
      middlePoint: { x: L, y: S },
      radius: a
    };
  });
}
function Zn({ data: t, threshold: n }) {
  if (n >= t.length || n < 3)
    return t;
  const e = [], s = (t.length - 2) / (n - 2);
  let r = 0;
  e.push(t[r]);
  for (let o = 0; o < n - 2; o += 1) {
    const c = Math.floor((o + 1) * s) + 1, i = Math.min(Math.floor((o + 2) * s) + 1, t.length), u = t.slice(c, i);
    let l = 0, a = 0;
    for (const d of u)
      l += d.x, a += d.y;
    l /= u.length, a /= u.length;
    let f = -1, h = r;
    for (let d = c; d < i; d += 1) {
      const g = Math.abs(
        (t[r].x - l) * (t[d].y - t[r].y) - (t[r].x - t[d].x) * (a - t[r].y)
      );
      g > f && (f = g, h = d);
    }
    e.push(t[h]), r = h;
  }
  return e.push(t[t.length - 1]), e;
}
function Qn({ data: t, threshold: n }) {
  if (n >= t.length || n < 3)
    return t;
  const e = [], s = (t.length - 2) / (n - 2);
  let r = 0;
  e.push(t[r]);
  for (let o = 0; o < n - 2; o += 1) {
    const c = Math.floor((o + 1) * s) + 1, i = Math.min(Math.floor((o + 2) * s) + 1, t.length), u = t.slice(c, i), l = u.reduce((h, d) => h + d, 0) / u.length;
    let a = -1, f = r;
    for (let h = c; h < i; h += 1) {
      const d = Math.abs((t[r] - l) * (h - r));
      d > a && (a = d, f = h);
    }
    e.push(t[f]), r = f;
  }
  return e.push(t[t.length - 1]), e;
}
function qn({ data: t, threshold: n, key: e = "value" }) {
  if (n >= t.length || n < 3)
    return t;
  const s = [], r = (t.length - 2) / (n - 2);
  let o = 0;
  s.push(t[o]);
  for (let c = 0; c < n - 2; c += 1) {
    const i = Math.floor((c + 1) * r) + 1, u = Math.min(Math.floor((c + 2) * r) + 1, t.length), l = t.slice(i, u), a = l.reduce((d, g) => d + g[e], 0) / l.length;
    let f = -1, h = o;
    for (let d = i; d < u; d += 1) {
      const g = Math.abs((t[o][e] - a) * (d - o));
      g > f && (f = g, h = d);
    }
    s.push(t[h]), o = h;
  }
  return s.push(t[t.length - 1]), s;
}
function Jn({ radius: t, centerX: n, centerY: e, percentage: s }) {
  s = Math.max(0, Math.min(1, s));
  const r = s * Math.PI, o = n - t, c = e, i = n - t * Math.cos(r), u = e - t * Math.sin(r);
  return `M ${n},${e} L ${o},${c} A ${t},${t} 0 0 1 ${i},${u} Z`.trim();
}
function te({ svgElement: t, x: n, y: e, offsetY: s = 0, element: r }) {
  if (!t || !r) return { top: 0, left: 0 };
  const o = t.createSVGPoint();
  o.x = n, o.y = e;
  const c = o.matrixTransform(t.getScreenCTM()), i = t.getBoundingClientRect(), u = r.getBoundingClientRect();
  let l = 0, a = 0;
  return c.x - u.width / 2 < i.left ? l = 0 : c.x + u.width > i.right ? l = -u.width : l = -u.width / 2, c.y - s - u.height < i.top ? a = s : a = -u.height - s, {
    top: c.y + a,
    left: c.x + l
  };
}
function ne({ svgElement: t, x: n, y: e, element: s, position: r }) {
  if (!t || !s) return { top: 0, left: 0 };
  const o = t.createSVGPoint();
  o.x = n, o.y = e;
  const c = o.matrixTransform(t.getScreenCTM()), { height: i, width: u } = s.getBoundingClientRect();
  let l = r === "right" ? 0 : -u, a = -(i / 2);
  return {
    top: c.y + a,
    left: c.x + l
  };
}
function Z(t) {
  if (t === null || typeof t != "object")
    return t;
  if (t instanceof Date)
    return new Date(t.getTime());
  if (t instanceof RegExp)
    return new RegExp(t.source, t.flags);
  if (t instanceof Map) {
    const e = /* @__PURE__ */ new Map();
    for (const [s, r] of t.entries())
      e.set(s, Z(r));
    return e;
  }
  if (t instanceof Set) {
    const e = /* @__PURE__ */ new Set();
    for (const s of t.values())
      e.add(Z(s));
    return e;
  }
  if (Array.isArray(t))
    return t.map((e) => Z(e));
  const n = {};
  for (const e in t)
    Object.prototype.hasOwnProperty.call(t, e) && (n[e] = Z(t[e]));
  return n;
}
function Vt(t) {
  const n = [];
  let e = [];
  for (const s of t)
    !s || s.value == null || Number.isNaN(s.x) || Number.isNaN(s.y) ? (e.length && n.push(e), e = []) : e.push(s);
  return e.length && n.push(e), n;
}
function ee(t, n) {
  const e = t.filter((c) => !!c);
  if (!e[0]) return [-10, -10, "", -10, -10].toString();
  const s = { x: e[0].x, y: n }, r = { x: e.at(-1).x, y: n }, o = [];
  return e.forEach((c) => {
    o.push(`${c.x},${c.y} `);
  }), [s.x, s.y, ...o, r.x, r.y].toString();
}
function re(t, n) {
  if (!t[0]) return [-10, -10, "", -10, -10].toString();
  const e = Vt(t);
  return e.length ? e.map((s) => {
    const r = { x: s[0].x, y: n }, o = { x: s.at(-1).x, y: n }, c = [];
    return s.forEach((i) => {
      c.push(`${i.x},${i.y} `);
    }), [r.x, r.y, ...c, o.x, o.y].toString();
  }).join(";") : "";
}
function _t(t) {
  const n = [];
  let e = [];
  for (const s of t)
    s.value == null || Number.isNaN(s.x) || Number.isNaN(s.y) ? (e.length > 1 && n.push(e), e = []) : e.push(s);
  return e.length > 1 && n.push(e), n;
}
function se(t) {
  let n = "", e = !1;
  const s = (r) => r.value != null && Number.isFinite(r.x) && Number.isFinite(r.y);
  for (let r = 0; r < t.length; r++) {
    const o = t[r];
    if (!s(o))
      continue;
    const c = `${x(o.x)},${x(o.y)}`;
    if (!e)
      n += c, e = !0;
    else {
      const i = t[r - 1], u = s(i) ? "L" : "M";
      n += `${u}${c}`;
    }
    n += " ";
  }
  return n.trim();
}
function oe(t) {
  const n = _t(t);
  if (!n.length) return "";
  let e = "";
  for (const [s, r] of n.entries()) {
    if (r.length < 2) continue;
    const o = r.length - 1, c = [], i = [], u = [], l = [];
    for (let a = 0; a < o; a += 1)
      c[a] = r[a + 1].x - r[a].x, i[a] = r[a + 1].y - r[a].y, u[a] = i[a] / c[a];
    l[0] = u[0], l[o] = u[o - 1];
    for (let a = 1; a < o; a += 1)
      if (u[a - 1] * u[a] <= 0)
        l[a] = 0;
      else {
        const f = 2 * u[a - 1] * u[a] / (u[a - 1] + u[a]);
        l[a] = f;
      }
    e += `${s === 0 ? "" : "M"}${x(r[0].x)},${x(r[0].y)} `;
    for (let a = 0; a < o; a += 1) {
      const f = r[a].x, h = r[a].y, d = r[a + 1].x, g = r[a + 1].y, p = l[a], E = l[a + 1], A = f + (d - f) / 3, F = h + p * (d - f) / 3, y = d - (d - f) / 3, B = g - E * (d - f) / 3;
      e += `C${x(A)},${x(F)} ${x(y)},${x(B)} ${x(d)},${x(g)} `;
    }
  }
  return e.trim();
}
function ie(t, n, e = !1, s = !0) {
  function r(c) {
    const i = [];
    let u = [];
    for (const l of c)
      !l || l.value == null || Number.isNaN(l.x) || Number.isNaN(l.y) ? (u.length > 1 && i.push(u), u = []) : u.push(l);
    return u.length > 1 && i.push(u), i;
  }
  return (e ? r(t) : [t]).map((c) => {
    if (c.length < 2) return "";
    const i = c.length - 1, u = [], l = [], a = [], f = [];
    for (let d = 0; d < i; d += 1)
      u[d] = c[d + 1].x - c[d].x, l[d] = c[d + 1].y - c[d].y, a[d] = l[d] / u[d];
    f[0] = a[0], f[i] = a[i - 1];
    for (let d = 1; d < i; d += 1)
      if (a[d - 1] * a[d] <= 0)
        f[d] = 0;
      else {
        const g = 2 * a[d - 1] * a[d] / (a[d - 1] + a[d]);
        f[d] = g;
      }
    let h = `M${c[0].x},${n}`;
    h += ` L${c[0].x},${c[0].y}`;
    for (let d = 0; d < i; d += 1) {
      const g = c[d].x, p = c[d].y, E = c[d + 1].x, A = c[d + 1].y, F = f[d], y = f[d + 1], B = g + (E - g) / 3, L = p + F * (E - g) / 3, S = E - (E - g) / 3, b = A - y * (E - g) / 3;
      h += ` C${B},${L} ${S},${b} ${E},${A}`;
    }
    return h += ` L${c[i].x},${n} ${s ? "Z" : ""}`, h;
  }).filter(Boolean);
}
function ce(t) {
  return t.toString().toLowerCase().replace(/\s+/g, "-").replace(/[^\w\-]+/g, "").replace(/\-\-+/g, "-").replace(/^-+/, "").replace(/-+$/, "");
}
function zt(t) {
  return t && typeof t == "object" && !Array.isArray(t) && Object.keys(t).length === 0 ? null : t;
}
function dt(t) {
  if (Array.isArray(t))
    return t.map(dt);
  if (t && typeof t == "object" && !Array.isArray(t)) {
    const n = {};
    for (const e in t)
      Object.hasOwn(t, e) && (n[e] = dt(t[e]));
    return zt(n);
  }
  return t;
}
function ae(t) {
  return 1 - Math.pow(1 - t, 3);
}
function ue({ values: t, config: n = {} }) {
  const {
    keepInvalid: e = !0,
    convertInvalidToZero: s = !1
  } = n, r = [];
  let o = 0, c = 0;
  function i(l) {
    return typeof l != "number" || !Number.isFinite(l);
  }
  function u(l) {
    o += l, c += 1, r.push(o / c);
  }
  for (const l of t)
    i(l) ? s && e ? u(0) : !s && e && r.push(l) : u(l);
  return r;
}
function le({ values: t, config: n = {} }) {
  const {
    keepInvalid: e = !0,
    convertInvalidToZero: s = !1
  } = n, r = [], o = [];
  function c(u) {
    return typeof u != "number" || !Number.isFinite(u);
  }
  function i(u) {
    o.push(u), o.sort((f, h) => f - h);
    const l = o.length, a = l >> 1;
    l % 2 === 1 ? r.push(o[a]) : r.push((o[a - 1] + o[a]) / 2);
  }
  for (const u of t)
    c(u) ? s && e ? i(0) : !s && e && r.push(u) : i(u);
  return r;
}
function fe({
  el: t,
  bounds: n,
  currentFontSize: e,
  minFontSize: s = 6,
  attempts: r = 200,
  padding: o = 1
}) {
  if (!t || !e) return 0;
  let c = e;
  t.setAttribute("font-size", c);
  const { x: i, y: u, width: l, height: a } = n, f = i + o, h = u + o, d = i + l - o, g = u + a - o;
  let p = t.getBBox();
  if (p.x >= f + o && p.y >= h + o && p.x + p.width <= d - o && p.y + p.height <= g - o)
    return c;
  let E = r;
  for (; E-- > 0 && c > s && (c--, t.setAttribute("font-size", c), p = t.getBBox(), !(p.x >= f + o && p.y >= h + o && p.x + p.width <= d - o && p.y + p.height <= g - o)); )
    ;
  return c < s && (c = 0, t.setAttribute("font-size", c)), c;
}
function he({
  value: t,
  maxDecimals: n = 4,
  fallbackFormatter: e,
  removeTrailingZero: s = !0
}) {
  if (t === 0)
    return "0";
  const r = Math.abs(t);
  if (r >= 1 && typeof e == "function") {
    const i = e(t);
    return String(i);
  }
  let o;
  if (r < 1) {
    const i = Math.floor(Math.log10(r));
    o = Math.min(Math.max(1 - i, 1), n);
  } else
    o = n;
  let c = t.toFixed(o);
  return s && (c = c.replace(/(\.\d*?[1-9])0+$/, "$1").replace(/\.0+$/, "")), c;
}
function de(t) {
  const n = [];
  for (let e = 0; e < t; e += 1)
    n.push(e === 0 ? 0 : e === 1 ? 1 : n[e - 1] + n[e - 2]);
  return n;
}
function ge(t, n = 20) {
  t = t.replace(/[\r\n]+/g, " ");
  const e = t.split(" ");
  let s = "", r = "";
  for (let o of e)
    (s + (s ? " " : "") + o).length <= n ? s += (s ? " " : "") + o : (s && (r += (r ? `
` : "") + s), s = o);
  return s && (r += (r ? `
` : "") + s), r;
}
function pe(t) {
  const {
    lineA: n,
    lineB: e,
    colorLineA: s,
    // fill when A is above
    colorLineB: r,
    // fill when B is above
    smoothA: o = !1,
    smoothB: c = !1,
    sampleStepPx: i = 2,
    cutNullValues: u = !0,
    // break across gaps
    merge: l = !0
    // merge into large polygons
  } = t || {};
  if (!Array.isArray(n) || !Array.isArray(e) || !n.length || !e.length)
    return [];
  const a = (b) => Number.isFinite(b);
  function f(b) {
    if (!u) return [b.filter((D) => D && a(D.x) && a(D.y))];
    const $ = [];
    let M = [];
    for (const D of b)
      D && a(D.x) && a(D.y) && D.value != null ? M.push({ x: D.x, y: D.y }) : (M.length > 1 && $.push(M), M = []);
    return M.length > 1 && $.push(M), $;
  }
  function h(b) {
    const $ = b.length - 1, M = new Array($), D = new Array($), m = new Array($), I = new Array(b.length);
    for (let C = 0; C < $; C += 1)
      M[C] = b[C + 1].x - b[C].x, D[C] = b[C + 1].y - b[C].y, m[C] = D[C] / M[C];
    I[0] = m[0], I[$] = m[$ - 1];
    for (let C = 1; C < $; C += 1)
      m[C - 1] * m[C] <= 0 ? I[C] = 0 : I[C] = 2 * m[C - 1] * m[C] / (m[C - 1] + m[C]);
    return I;
  }
  function d(b, $, M, D, m) {
    const I = b.x, C = $.x, w = b.y, R = $.y, P = C - I;
    if (P === 0) return w;
    const T = (m - I) / P, O = T * T, Y = O * T, H = 2 * Y - 3 * O + 1, N = Y - 2 * O + T, k = -2 * Y + 3 * O, U = Y - O;
    return H * w + N * (M * P) + k * R + U * (D * P);
  }
  function g(b, $) {
    const M = f(b);
    if (!M.length) return [];
    let D = 1 / 0, m = -1 / 0;
    for (const R of M)
      D = Math.min(D, R[0].x), m = Math.max(m, R[R.length - 1].x);
    if (!a(D) || !a(m) || m <= D) return [];
    const I = Math.max(1, i), C = [];
    for (let R = D; R <= m; R += I) C.push(R);
    C[C.length - 1] < m && C.push(m);
    const w = [];
    for (const R of C) {
      let P = null, T = !1;
      for (const O of M) {
        const Y = O.length - 1;
        if (!(R < O[0].x - 1e-9 || R > O[Y].x + 1e-9)) {
          for (let H = 0; H < Y; H += 1) {
            const N = O[H], k = O[H + 1];
            if (!(R + 1e-9 < N.x || R - 1e-9 > k.x)) {
              if ($) {
                const U = O.__tangents || (O.__tangents = h(O));
                P = d(N, k, U[H], U[H + 1], R);
              } else {
                const U = (R - N.x) / (k.x - N.x || 1);
                P = N.y + U * (k.y - N.y);
              }
              T = !0;
              break;
            }
          }
          if (T) break;
        }
      }
      P == null ? w.push({ x: R, y: null, hole: !0 }) : w.push({ x: R, y: P, hole: !1 });
    }
    return w;
  }
  function p(b, $, M) {
    return b + M * ($ - b);
  }
  function E(b, $) {
    const M = [], D = [], m = Math.min(b.length, $.length);
    for (let I = 0; I < m - 1; I += 1) {
      const C = b[I], w = b[I + 1], R = $[I], P = $[I + 1];
      if (M.push(C), D.push(R), C.hole || w.hole || R.hole || P.hole || C.y == null || w.y == null || R.y == null || P.y == null)
        continue;
      const T = C.y - R.y, O = w.y - P.y;
      if (T > 0 && O < 0 || T < 0 && O > 0) {
        const Y = T / (T - O), H = p(C.x, w.x, Y), N = p(C.y, w.y, Y), k = { x: H, y: N, hole: !1 }, U = { x: H, y: N, hole: !1 };
        M.push(k), D.push(U);
      }
    }
    return m > 0 && (M.push(b[m - 1]), D.push($[m - 1])), { A: M, B: D };
  }
  function A(b, $) {
    const M = [], D = Math.min(b.length, $.length);
    for (let m = 0; m < D - 1; m += 1) {
      const I = b[m], C = b[m + 1], w = $[m], R = $[m + 1];
      if (I.hole || C.hole || w.hole || R.hole || I.y == null || C.y == null || w.y == null || R.y == null) continue;
      const P = I.y - w.y, T = C.y - R.y, O = P <= 0 ? I : w, Y = T <= 0 ? C : R, H = T <= 0 ? R : C, N = P <= 0 ? w : I, k = P <= 0 ? s : r, U = [
        `M${O.x},${O.y}`,
        `L${Y.x},${Y.y}`,
        `L${H.x},${H.y}`,
        `L${N.x},${N.y}`,
        "Z"
      ].join(" ");
      M.push({ d: U, color: k });
    }
    return M;
  }
  function F(b, $) {
    const M = [], D = Math.min(b.length, $.length);
    if (D < 2) return M;
    let m = 0;
    for (; m < D - 1; ) {
      for (; m < D - 1; ) {
        const N = b[m], k = $[m], U = b[m + 1], K = $[m + 1];
        if (!N.hole && !k.hole && !U.hole && !K.hole && N.y != null && k.y != null && U.y != null && K.y != null) break;
        m += 1;
      }
      if (m >= D - 1) break;
      const I = m, C = Math.sign($[m].y - b[m].y || 0) || 1;
      for (m += 1; m < D - 1; ) {
        const N = b[m], k = $[m], U = b[m + 1], K = $[m + 1];
        if (N.hole || k.hole || U.hole || K.hole || N.y == null || k.y == null || U.y == null || K.y == null || (Math.sign(k.y - N.y || 0) || 1) !== C) break;
        m += 1;
      }
      const w = m + 0, R = C >= 0 ? b : $, P = C >= 0 ? $ : b, T = C >= 0 ? s : r, O = [];
      for (let N = I; N <= w; N += 1)
        O.push(`${R[N].x},${R[N].y}`);
      const Y = [];
      for (let N = w; N >= I; N -= 1)
        Y.push(`${P[N].x},${P[N].y}`);
      const H = `M${O[0]} L${O.slice(1).join(" L")} L${Y.join(" L")} Z`;
      M.push({ d: H, color: T });
    }
    return M;
  }
  const y = g(n, o), B = g(e, c), { A: L, B: S } = E(y, B);
  return l ? F(L, S) : A(L, S);
}
function Fe(t, n, e = {}) {
  const r = { ...{ bubbles: !0, cancelable: !0, composed: !0 }, ...e }, o = /* @__PURE__ */ new Set([
    "click",
    "mousedown",
    "mouseup",
    "mousemove",
    "mouseover",
    "mouseout",
    "mouseenter",
    "mouseleave",
    "dblclick",
    "contextmenu"
  ]), c = /* @__PURE__ */ new Set([
    "keydown",
    "keyup",
    "keypress"
  ]);
  let i;
  if (o.has(n))
    i = new MouseEvent(n, r);
  else if (c.has(n))
    i = new KeyboardEvent(n, r);
  else if (n === "input")
    try {
      i = new InputEvent(n, r);
    } catch {
      i = new Event(n, r);
    }
  else n.startsWith("custom:") ? i = new CustomEvent(n, { ...r, detail: r.detail }) : i = new Event(n, r);
  return t.dispatchEvent(i), i;
}
function me(t, { delta: n = 1, delay: e = 20, disableTransitions: s = !0 } = {}) {
  if (!t || !(t instanceof HTMLElement)) return;
  const r = t.style, o = {
    width: r.width,
    height: r.height,
    transition: r.transition
  }, c = t.getBoundingClientRect(), i = c.width, u = c.height;
  s && (r.transition = "none");
  const l = (a) => /%|em|rem/.test(a);
  r.width = o.width && l(o.width) ? `calc(${o.width} + ${n}px)` : `${Math.max(0, i + n)}px`, r.height = o.height && l(o.height) ? `calc(${o.height} + ${n}px)` : `${Math.max(0, u + n)}px`, t.offsetWidth, setTimeout(() => {
    r.width = o.width, r.height = o.height, t.offsetWidth, requestAnimationFrame(() => {
      s && (r.transition = o.transition);
    });
  }, e);
}
function Zt(t) {
  let n = null, e = null;
  return (...s) => {
    const r = JSON.stringify(s);
    return r === n || (n = r, e = t(...s)), e;
  };
}
const Ee = Zt((t, n, e, s, r, o, c, i) => {
  if (t)
    return s.length <= 2 ? s.map((F, y) => ({ text: F, absoluteIndex: y })) : s.map((F, y) => ({ text: y === 0 || y === s.length - 1 || c != null && y === c ? F : "", absoluteIndex: y }));
  if (!n)
    return s.map((A, F) => ({ text: A, absoluteIndex: F }));
  const u = Math.max(1, e || 1);
  if (i <= u)
    return s.map((A, F) => ({ text: A, absoluteIndex: F }));
  const l = [];
  for (let A = 0; A < s.length; A += 1) {
    const F = s[A] ?? "";
    if (!F) continue;
    const y = o + A - 1 >= 0 ? r[o + A - 1] ?? "" : null;
    F !== y && l.push(A);
  }
  if (!l.length)
    return s.map((A, F) => ({ text: "", absoluteIndex: F }));
  const a = l.length, f = u, h = Math.max(2, Math.min(f - 3, a)), d = Math.min(a, f + 3);
  let g = Math.min(f, a), p = 1 / 0;
  for (let A = h; A <= d; A += 1) {
    const F = (a - 1) % (A - 1), y = Math.abs(A - f), B = F * 10 + y;
    B < p && (p = B, g = A);
  }
  const E = /* @__PURE__ */ new Set();
  if (g <= 1)
    E.add(l[Math.round((a - 1) / 2)]);
  else {
    const A = (a - 1) / (g - 1);
    for (let F = 0; F < g; F += 1)
      E.add(l[Math.round(F * A)]);
  }
  return s.map((A, F) => ({
    text: E.has(F) ? A : "",
    absoluteIndex: F
  }));
});
function Ae(t) {
  return String(t).replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
}
function xe(t) {
  return String(t).replaceAll("&", "&amp;").replaceAll('"', "&quot;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
}
function be(t) {
  if (!t || typeof t != "string")
    return null;
  const n = document.createElementNS(rt, "svg");
  n.setAttribute("xmlns", rt);
  const e = document.createElementNS(rt, "path");
  e.setAttribute("d", t), n.appendChild(e);
  const s = e.getTotalLength(), r = e.getPointAtLength(s / 2);
  return { x: r.x, y: r.y };
}
function ye(t, n = "(", e = ")") {
  return `${n}${t}${e}`;
}
export {
  Wt as $,
  hn as A,
  se as B,
  fn as C,
  Fe as D,
  it as E,
  te as F,
  Rn as G,
  Yn as H,
  Un as I,
  he as J,
  ue as K,
  le as L,
  en as M,
  Mn as N,
  $n as O,
  tn as P,
  nn as Q,
  Qn as R,
  Ee as S,
  Gn as T,
  x as U,
  ie as V,
  re as W,
  rt as X,
  ee as Y,
  pe as Z,
  Sn as _,
  Wn as a,
  ne as a0,
  Kn as a1,
  Xn as a2,
  Jt as a3,
  Fn as a4,
  z as a5,
  _ as a6,
  Vn as a7,
  mn as a8,
  En as a9,
  lt as aA,
  be as aB,
  dt as aa,
  On as ab,
  pn as ac,
  Jn as ad,
  Z as ae,
  An as af,
  ae as ag,
  vn as ah,
  Zn as ai,
  cn as aj,
  dn as ak,
  Pn as al,
  Dn as am,
  ln as an,
  Tn as ao,
  jn as ap,
  kn as aq,
  ce as ar,
  ge as as,
  zn as at,
  Ut as au,
  xe as av,
  Ae as aw,
  me as ax,
  qn as ay,
  de as az,
  on as b,
  gn as c,
  Bn as d,
  sn as e,
  fe as f,
  Ln as g,
  rn as h,
  j as i,
  ut as j,
  Hn as k,
  wn as l,
  gt as m,
  un as n,
  In as o,
  ye as p,
  yn as q,
  xn as r,
  _n as s,
  Mt as t,
  Cn as u,
  bn as v,
  et as w,
  an as x,
  Nn as y,
  oe as z
};
