// vue-data-ui/dist/useConfig-DO94Aj5s.js
function ct(A = {}) {
  const s = "#2D353C", R = "#A1A1A1", W = "#e1e5e8", h = "#CCCCCC", V = "#4A4A4A", l = "#FFFFFF", E = "#fafafa", c = "#1f77b4", z = "#d62728", O = "#2ca02c", N = "#42d392", Y = "#FFD055", r = A.colorBackground ?? l, e = A.colorTextPrimary ?? s, T = A.colorTextSecondary ?? R, o = A.colorGrid ?? W, k = A.colorBorder ?? W, a = {
    translations: {
      keyboardNavigation: "Use the left and right arrow keys to move between data points.",
      tableAvailable: "A data table for this chart is available below.",
      tableCaption: "Chart data table"
    }
  }, t = {
    _6: 6,
    _10: 10,
    _12: 12,
    _14: 14,
    _16: 16,
    _18: 18,
    _20: 20,
    _24: 24,
    _32: 32,
    _48: 48
  }, _ = t._6, v = {
    LINE: "line",
    SQUARE: "square",
    ROUND: "round",
    CIRCLE: "circle",
    CURVED: "curved",
    BAR: "bar",
    STAR: "star"
  }, m = {
    LEFT: "left",
    CENTER: "center",
    BOTTOM: "bottom",
    TOP: "top"
  }, ee = {
    DESC: "desc"
  }, D = {
    threshold: 1095
    // v2 = 500
  }, w = {
    showValueFirst: !0,
    usePercentageParens: !0,
    useValueParens: !1
  }, n = {
    text: "",
    color: e,
    fontSize: t._20,
    bold: !0,
    textAlign: m.CENTER,
    paddingLeft: 0,
    paddingRight: 0,
    subtitle: {
      color: T,
      text: "",
      fontSize: t._16,
      bold: !1
    }
  }, b = {
    show: !0,
    bold: !1,
    backgroundColor: r,
    color: e,
    fontSize: t._14,
    selectAllToggle: {
      show: !1,
      backgroundColor: k,
      color: e
    }
  }, g = {
    show: !0,
    color: e,
    backgroundColor: r,
    fontSize: t._14,
    customFormat: null,
    borderRadius: 4,
    borderColor: k,
    borderWidth: 1,
    backgroundOpacity: 100,
    position: m.CENTER,
    offsetY: 24,
    smooth: !0,
    // v3
    backdropFilter: !0,
    // v3
    smoothForce: 0.18,
    smoothSnapThreshold: 0.25,
    teleportTo: "body"
    // or any other css selector (used in the `to` attr of the Teleport component)
  }, C = {
    enable: !1,
    locale: "en",
    useUTC: !1,
    januaryAsYear: !1,
    options: {
      year: "yyyy",
      month: "MMM 'yy",
      day: "dd MMM",
      hour: "HH:mm",
      minute: "HH:mm:ss",
      second: "HH:mm:ss"
    }
  }, i = ({
    tooltip: L = !1,
    pdf: H = !1,
    csv: U = !1,
    img: I = !1,
    table: X = !1,
    labels: G = !1,
    fullscreen: $ = !1,
    sort: j = !1,
    stack: q = !1,
    animation: Z = !1,
    annotator: K = !1,
    svg: Q = !1,
    zoom: J = !1,
    altCopy: ft = !1
  }) => {
    const y = {
      open: "Open options",
      close: "Close options",
      tooltip: "Toggle tooltip",
      pdf: "Download PDF",
      csv: "Download CSV",
      img: "Download PNG",
      table: "Toggle table",
      labels: "Toggle labels",
      fullscreen: "Toggle fullscreen",
      sort: "Toggle sort",
      stack: "Toggle stack mode",
      animation: "Toggle animation",
      annotator: "Toggle annotator",
      svg: "Download SVG",
      zoom: "Toggle zoom lock",
      altCopy: "Copy alt text"
    };
    return L || delete y.tooltip, H || delete y.pdf, U || delete y.csv, I || delete y.img, X || delete y.table, G || delete y.labels, $ || delete y.fullscreen, j || delete y.sort, q || delete y.stack, Z || delete y.animation, K || delete y.annotator, Q || delete y.svg, J || delete y.zoom, {
      show: !0,
      showOnChartHover: !1,
      keepStateOnChartLeave: !0,
      position: "right",
      buttons: {
        tooltip: L,
        pdf: H,
        csv: U,
        img: I,
        table: X,
        labels: G,
        fullscreen: $,
        sort: j,
        stack: q,
        animation: Z,
        annotator: K,
        svg: Q,
        zoom: J,
        altCopy: ft
      },
      callbacks: {
        animation: null,
        annotator: null,
        csv: null,
        fullscreen: null,
        img: null,
        labels: null,
        pdf: null,
        sort: null,
        stack: null,
        table: null,
        tooltip: null,
        svg: null,
        zoom: null,
        altCopy: null
      },
      buttonTitles: y,
      print: {
        scale: 2,
        orientation: "auto",
        // 'auto' | 'l' | 'p'
        overflowTolerance: 0.2
      },
      useCursorPointer: !1
    };
  }, p = (L) => ({
    top: L[0],
    right: L[1],
    bottom: L[2],
    left: L[3]
  }), u = {
    backgroundColor: r,
    color: e,
    outline: "none"
  }, d = {
    backgroundColor: r,
    color: e,
    outline: "none"
  }, f = {
    show: !1,
    responsiveBreakpoint: 400
  }, P = {
    show: !0,
    color: h,
    highlightColor: V,
    fontSize: t._14,
    useResetSlot: !1,
    startIndex: null,
    endIndex: null,
    enableRangeHandles: !0,
    enableSelectionDrag: !0,
    focusOnDrag: !1,
    focusRangeRatio: 0.2,
    maxWidth: null
  }, F = {
    show: !1,
    selectedColor: c,
    selectedColorOpacity: 0.2,
    indicatorColor: s,
    verticalHandles: !1,
    compact: !0,
    frameColor: R,
    additionalHeight: 0,
    handleIconColor: null,
    handleBorderWidth: 1,
    handleBorderColor: null,
    handleFill: null,
    handleWidth: 20,
    handleType: "grab"
    // 'empty' | 'chevron' | 'grab' | 'arrow'
  }, M = {
    ...F,
    smooth: !1,
    lineColor: s,
    selectionRadius: 2,
    merged: !1
  }, te = {
    skeletonDataset: null,
    skeletonConfig: null,
    loading: !1,
    debug: !1,
    theme: "",
    responsive: !1,
    a11y: a,
    events: {
      datapointEnter: null,
      datapointLeave: null,
      datapointClick: null
    },
    customPalette: [],
    useCssAnimation: !1,
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        period: "Period",
        total: "Total"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 500,
        width: 800,
        padding: p([12, 12, 12, 12]),
        title: n,
        legend: {
          ...b,
          position: "bottom"
        },
        zoom: {
          ...P,
          minimap: F,
          preview: {
            enable: !0,
            fill: "#CCCCCC50",
            stroke: "#6A6A6A",
            strokeWidth: 2,
            strokeDasharray: 0
          },
          useDefaultFormat: !0,
          timeFormat: "yyyy-MM-dd HH:mm:ss",
          // When datetimeFormatter is enabled and useDefaultFormat is false
          customFormat: null
          // overrides all if callback => string
        },
        highlighter: {
          color: e,
          opacity: 5,
          useLine: !1,
          lineDasharray: 2,
          lineWidth: 1
        },
        tooltip: {
          ...g,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0,
          showTimeLabel: !0,
          showTotal: !1,
          totalTranslation: "Total",
          useDefaultTimeFormat: !0,
          timeFormat: "yyyy-MM-dd HH:mm:ss"
          // When datetimeFormatter is enabled and useDefaultFormat is false
        },
        grid: {
          scale: {
            ticks: 10,
            scaleMin: null,
            // Force min scale (defaults to dataset's min)
            scaleMax: null
            // Force max scale (defaults to dataset's max)
          },
          frame: {
            show: !1,
            stroke: o,
            strokeWidth: 2,
            strokeLinecap: "round",
            strokeLinejoin: "round",
            strokeDasharray: 0
          },
          x: {
            showAxis: !0,
            showHorizontalLines: !1,
            axisColor: o,
            linesColor: o,
            linesThickness: 1,
            linesStrokeDasharray: 0,
            axisThickness: 2,
            axisName: {
              show: !0,
              text: "",
              fontSize: t._14,
              color: e,
              bold: !1,
              offsetY: 0
            },
            timeLabels: {
              show: !0,
              values: [],
              datetimeFormatter: C,
              offsetY: 0,
              rotation: 0,
              autoRotate: {
                enable: !0,
                angle: -30
              },
              fontSize: t._14,
              color: e,
              bold: !1,
              showOnlyFirstAndLast: !1,
              showOnlyAtModulo: !1,
              modulo: 12
            }
          },
          y: {
            showAxis: !0,
            showVerticalLines: !1,
            linesColor: o,
            linesThickness: 1,
            linesStrokeDasharray: 0,
            axisColor: o,
            axisThickness: 2,
            axisName: {
              show: !0,
              text: "",
              fontSize: t._14,
              color: e,
              bold: !1,
              offsetX: 0
            },
            axisLabels: {
              formatter: null,
              show: !0,
              color: e,
              fontSize: t._14,
              bold: !1,
              rounding: 0
            }
          }
        },
        lines: {
          useArea: !0,
          smooth: !1,
          areaOpacity: 0.5,
          distributed: !1,
          showDistributedPercentage: !0,
          strokeWidth: 1,
          gradient: {
            show: !0,
            intensity: 40
          },
          dot: {
            hideAboveMaxSerieLength: 62,
            useSerieColor: !0,
            fill: l,
            strokeWidth: 0.5,
            radius: 4
          },
          totalValues: {
            show: !0,
            offsetY: 0,
            fontSize: t._16,
            bold: !0,
            color: e
          },
          dataLabels: {
            hideAboveMaxSerieLength: 62,
            offsetY: 0,
            show: !0,
            hideEmptyValues: !0,
            hideUnderValue: null,
            hideEmptyPercentages: !0,
            hideUnderPercentage: null,
            color: e,
            fontSize: t._14,
            bold: !1,
            rounding: 0,
            prefix: "",
            suffix: "",
            formatter: null
          }
        }
      }
    }
  }, oe = {
    skeletonDataset: null,
    skeletonConfig: null,
    loading: !1,
    // v3
    debug: !1,
    // v3
    theme: "",
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    customPalette: [],
    useCssAnimation: !1,
    // v3 (v2 = true)
    orientation: "vertical",
    // or 'horizontal'
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        period: "Period",
        total: "Total"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 500,
        width: 800,
        padding: p([12, 12, 12, 12]),
        title: n,
        legend: {
          ...b,
          position: "bottom"
        },
        zoom: {
          ...P,
          minimap: F,
          preview: {
            enable: !0,
            fill: "#CCCCCC50",
            stroke: "#6A6A6A",
            strokeWidth: 2,
            strokeDasharray: 0
          },
          useDefaultFormat: !0,
          timeFormat: "yyyy-MM-dd HH:mm:ss",
          // When datetimeFormatter is enabled and useDefaultFormat is false
          customFormat: null
          // overrides all if callback => string
        },
        tooltip: {
          ...g,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0,
          showTimeLabel: !0,
          showTotal: !1,
          totalTranslation: "Total",
          useDefaultTimeFormat: !0,
          timeFormat: "yyyy-MM-dd HH:mm:ss"
          // When datetimeFormatter is enabled and useDefaultFormat is false
        },
        highlighter: {
          color: e,
          opacity: 5
        },
        bars: {
          gapRatio: 0.5,
          distributed: !1,
          showDistributedPercentage: !0,
          borderRadius: 0,
          strokeWidth: 1,
          gradient: {
            show: !0,
            intensity: 20
          },
          totalValues: {
            show: !0,
            offsetY: 0,
            offsetX: 0,
            fontSize: t._16,
            bold: !1,
            color: e
          },
          dataLabels: {
            show: !0,
            hideEmptyValues: !1,
            hideUnderValue: null,
            hideEmptyPercentages: !1,
            hideUnderPercentage: null,
            adaptColorToBackground: !0,
            color: e,
            fontSize: t._14,
            bold: !1,
            rounding: 0,
            prefix: "",
            suffix: "",
            formatter: null
          }
        },
        grid: {
          scale: {
            ticks: 10,
            scaleMin: null,
            // Force min scale (defaults to dataset's min)
            scaleMax: null
            // Force max scale (defaults to dataset's max)
          },
          frame: {
            show: !1,
            stroke: o,
            strokeWidth: 2,
            strokeLinecap: "round",
            strokeLinejoin: "round",
            strokeDasharray: 0
          },
          x: {
            showAxis: !0,
            showHorizontalLines: !1,
            axisColor: o,
            linesColor: o,
            linesThickness: 1,
            linesStrokeDasharray: 0,
            axisThickness: 2,
            axisName: {
              show: !0,
              text: "",
              fontSize: t._14,
              color: e,
              bold: !1,
              offsetY: 0
            },
            timeLabels: {
              show: !0,
              values: [],
              datetimeFormatter: C,
              offsetY: 0,
              rotation: 0,
              autoRotate: {
                // v3
                enable: !0,
                // v3
                angle: -30
                // v3
              },
              fontSize: t._14,
              color: e,
              bold: !1,
              showOnlyFirstAndLast: !1,
              showOnlyAtModulo: !1,
              modulo: 12
            }
          },
          y: {
            showAxis: !0,
            showVerticalLines: !1,
            linesColor: o,
            linesThickness: 1,
            linesStrokeDasharray: 0,
            axisColor: o,
            axisThickness: 2,
            axisName: {
              show: !0,
              text: "",
              fontSize: t._14,
              color: e,
              bold: !1,
              offsetX: 0
            },
            axisLabels: {
              formatter: null,
              show: !0,
              color: e,
              fontSize: t._14,
              bold: !1,
              rounding: 0
            }
          }
        }
      }
    }
  }, S = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    theme: "",
    responsive: !1,
    loading: !1,
    // v3
    a11y: a,
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    responsiveProportionalSizing: !0,
    customPalette: [],
    useCssAnimation: !1,
    // v3 (v2 = true)
    downsample: D,
    chart: {
      fontFamily: "inherit",
      backgroundColor: r,
      color: e,
      height: 600,
      width: 1e3,
      // Annotations are hidden in stack mode
      annotations: [
        {
          show: !1,
          yAxis: {
            yTop: null,
            yBottom: null,
            label: {
              text: "",
              textAnchor: "start",
              // or middle or end
              position: "start",
              // or end
              offsetX: 0,
              offsetY: 0,
              padding: p([12, 12, 12, 12]),
              border: {
                stroke: l,
                strokeWidth: 1,
                rx: 0,
                ry: 0
              },
              fontSize: 14,
              color: e,
              backgroundColor: o
            },
            line: {
              stroke: s,
              strokeWidth: 1,
              strokeDasharray: 0
            },
            area: {
              fill: o,
              opacity: 30
            }
          }
        }
      ],
      zoom: {
        ...P,
        minimap: M,
        preview: {
          // v3
          enable: !0,
          fill: "#CCCCCC50",
          stroke: "#6A6A6A",
          strokeWidth: 2,
          strokeDasharray: 0
        },
        useDefaultFormat: !0,
        timeFormat: "yyyy-MM-dd HH:mm:ss",
        // When datetimeFormatter is enabled
        customFormat: null
        // overrides all if callback => string
      },
      padding: p([12, 12, 6, 6]),
      highlighter: {
        color: e,
        opacity: 5,
        useLine: !1,
        lineDasharray: 2,
        lineWidth: 1
      },
      // Highlight area was initially a single object of type VueUiXyHighlightArea.
      // It now also works with a type VueUiXyHighlightArea[]
      highlightArea: {
        show: !1,
        from: 0,
        to: 0,
        color: s,
        opacity: 20,
        caption: {
          text: "",
          fontSize: t._10,
          color: e,
          bold: !1,
          offsetY: 0,
          width: "auto",
          padding: 3,
          textAlign: m.CENTER
        }
      },
      timeTag: {
        show: !1,
        backgroundColor: o,
        color: e,
        fontSize: t._12,
        circleMarker: {
          radius: 3,
          color: e
        },
        useDefaultFormat: !0,
        timeFormat: "yyyy-MM-dd HH:mm:ss",
        // When datetimeFormatter is enabled
        customFormat: null
        // overrides all if callback => string
      },
      grid: {
        stroke: o,
        showVerticalLines: !1,
        showHorizontalLines: !1,
        position: "middle",
        // or 'start'
        frame: {
          show: !1,
          stroke: o,
          strokeWidth: 2,
          strokeLinecap: "round",
          strokeLinejoin: "round",
          strokeDasharray: 0
        },
        labels: {
          show: !0,
          color: e,
          fontSize: t._16,
          axis: {
            yLabel: "",
            yLabelOffsetX: 0,
            xLabel: "",
            xLabelOffsetY: 0,
            fontSize: t._14
          },
          zeroLine: {
            show: !0
          },
          xAxis: {
            showBaseline: !1,
            showCrosshairs: !0,
            crosshairsAlwaysAtZero: !1,
            crosshairSize: 6
          },
          yAxis: {
            position: "left",
            showBaseline: !0,
            showCrosshairs: !0,
            crosshairSize: 6,
            commonScaleSteps: 10,
            useIndividualScale: !1,
            useNiceScale: !1,
            stacked: !1,
            gap: 12,
            labelWidth: 64,
            formatter: null,
            scaleMin: null,
            // Overrides auto scaling
            scaleMax: null,
            // idem
            groupColor: null,
            // force yAxis labels color
            scaleLabelOffsetX: 0,
            scaleValueOffsetX: 0,
            rounding: 1,
            serieNameFormatter: null
            // v3, for individual scale & stacked modes
          },
          xAxisLabels: {
            color: e,
            show: !0,
            values: [],
            datetimeFormatter: C,
            fontSize: t._14,
            showOnlyFirstAndLast: !1,
            showOnlyAtModulo: !1,
            modulo: 12,
            yOffset: 24,
            rotation: 0,
            autoRotate: {
              // v3
              enable: !0,
              // v3
              angle: -30
              // v3
            }
          }
        }
      },
      comments: {
        show: !0,
        showInTooltip: !0,
        width: 200,
        offsetX: 0,
        offsetY: 0
      },
      labels: {
        fontSize: t._10,
        prefix: "",
        suffix: ""
      },
      legend: {
        color: e,
        show: !0,
        fontSize: t._14,
        position: "bottom",
        // bottom | top
        selectAllToggle: {
          show: !1,
          backgroundColor: k,
          color: e
        }
      },
      title: {
        ...n,
        show: !0
      },
      tooltip: {
        ...g,
        showTimeLabel: !0,
        showValue: !0,
        showPercentage: !0,
        roundingValue: 0,
        roundingPercentage: 0,
        useDefaultTimeFormat: !0,
        timeFormat: "yyyy-MM-dd HH:mm:ss"
        // When datetimeFormatter is used
      },
      userOptions: i({
        tooltip: !0,
        pdf: !0,
        csv: !0,
        img: !0,
        table: !0,
        labels: !0,
        fullscreen: !0,
        stack: !0,
        annotator: !0,
        svg: !0
      })
    },
    bar: {
      showTransition: !0,
      transitionDurationMs: 300,
      borderRadius: 2,
      useGradient: !0,
      periodGap: 0.1,
      innerGap: 0,
      border: {
        useSerieColor: !1,
        strokeWidth: 0,
        stroke: l
      },
      labels: {
        show: !1,
        offsetY: -6,
        rounding: 0,
        color: e,
        formatter: null
      },
      serieName: {
        show: !1,
        offsetY: -6,
        useAbbreviation: !0,
        abbreviationSize: 3,
        useSerieColor: !0,
        color: e,
        bold: !1
      }
    },
    line: {
      showTransition: !0,
      transitionDurationMs: 300,
      radius: 3,
      useGradient: !0,
      strokeWidth: 3,
      cutNullValues: !1,
      interLine: {
        pairs: [],
        colors: [],
        fillOpacity: 0.25
      },
      dot: {
        hideAboveMaxSerieLength: 62,
        useSerieColor: !0,
        fill: l,
        strokeWidth: 0.5
      },
      labels: {
        show: !1,
        offsetY: -6,
        rounding: 0,
        color: e,
        formatter: null
      },
      area: {
        useGradient: !0,
        opacity: 30
      },
      tag: {
        followValue: !0,
        formatter: null,
        fontSize: t._14
      }
    },
    plot: {
      showTransition: !0,
      transitionDurationMs: 300,
      radius: 3,
      useGradient: !0,
      dot: {
        useSerieColor: !0,
        fill: l,
        strokeWidth: 0.5
      },
      labels: {
        show: !1,
        offsetY: -6,
        rounding: 0,
        color: e,
        formatter: null
      },
      tag: {
        followValue: !0,
        formatter: null,
        fontSize: t._14
      }
    },
    table: {
      useDialog: !1,
      responsiveBreakpoint: 400,
      rounding: 0,
      sparkline: !0,
      showSum: !0,
      columnNames: {
        period: "Period",
        total: "Total"
      },
      th: u,
      td: d,
      useDefaultTimeFormat: !0,
      timeFormat: "yyyy-MM-dd HH:mm:ss"
    },
    showTable: !1
  }, x = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    type: "classic",
    loading: !1,
    // v3
    pie: !1,
    // v3
    autoSize: !0,
    // false = v2
    responsive: !1,
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v3 (v2 = true)
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    serieToggleAnimation: {
      show: !0,
      durationMs: 500
    },
    startAnimation: {
      show: !1,
      // v3 (v2 = true)
      durationMs: 1e3,
      staggerMs: 50
    },
    useBlurOnHover: !0,
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    translations: {
      total: "Total",
      average: "Average"
    },
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      },
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      }
    },
    style: {
      fontFamily: "inherit",
      chart: {
        useGradient: !0,
        gradientIntensity: 40,
        backgroundColor: r,
        color: e,
        padding: p([0, 0, 0, 0]),
        width: 512,
        height: 360,
        layout: {
          curvedMarkers: !1,
          labels: {
            dataLabels: {
              ...w,
              show: !0,
              useLabelSlots: !1,
              hideUnderValue: 3,
              smallArcClusterThreshold: 8,
              // percentage under which dataLabels are clusted on a side
              smallArcClusterFontSize: t._12,
              oneLine: !1,
              prefix: "",
              suffix: ""
            },
            value: {
              rounding: 0,
              show: !0,
              formatter: null
            },
            percentage: {
              show: !0,
              color: e,
              bold: !0,
              fontSize: t._18,
              minFontSize: _,
              // v3
              rounding: 0,
              formatter: null
            },
            name: {
              show: !0,
              color: e,
              bold: !1,
              fontSize: t._14,
              minFontSize: _
              // v3
            },
            hollow: {
              show: !0,
              total: {
                show: !0,
                bold: !1,
                fontSize: t._18,
                color: e,
                text: "Total",
                offsetY: 0,
                value: {
                  color: e,
                  fontSize: t._18,
                  bold: !0,
                  suffix: "",
                  prefix: "",
                  offsetY: 0,
                  rounding: 0,
                  formatter: null
                }
              },
              average: {
                show: !0,
                bold: !1,
                fontSize: t._18,
                color: T,
                text: "Average",
                offsetY: 0,
                value: {
                  color: e,
                  fontSize: t._18,
                  bold: !0,
                  suffix: "",
                  prefix: "",
                  offsetY: 0,
                  rounding: 0,
                  formatter: null
                }
              }
            }
          },
          donut: {
            radiusRatio: 0.3,
            // v3 (clamped between 0.1 and 0.5)
            strokeWidth: 64,
            // v3 (v2 = 55)
            borderWidth: 1,
            useShadow: !1,
            shadowColor: s,
            emptyFill: o,
            selectedColor: "#0000001A",
            borderColorAuto: !0,
            borderColor: k
          }
        },
        comments: {
          show: !0,
          showInTooltip: !0,
          width: 100,
          offsetY: 0,
          offsetX: 0
        },
        legend: {
          ...b,
          ...w,
          roundingValue: 0,
          roundingPercentage: 0,
          showPercentage: !0,
          showValue: !0,
          position: "bottom"
        },
        tooltip: {
          ...g,
          ...w,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0
        },
        title: n
      }
    }
  }, re = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 500,
        width: 800,
        padding: p([0, 0, 0, 0]),
        layout: {
          sorted: !0,
          rects: {
            stroke: o,
            strokeWidth: 1,
            borderRadius: 0,
            colorRatio: 0.3,
            gradient: {
              show: !0,
              intensity: 30
            },
            group: {
              stroke: o,
              strokeWidth: 1,
              useSeriesBackgroundColor: !1,
              backgroundLighterRatio: 0.4,
              label: {
                adaptColorToBackground: !1,
                color: e
              }
            },
            selected: {
              stroke: o,
              strokeWidth: 1,
              unselectedOpacity: 0.6
            }
          },
          labels: {
            showDefaultLabels: !0,
            fontSize: t._24,
            minFontSize: t._10,
            hideUnderProportion: 0.03,
            prefix: "Value: ",
            suffix: "",
            rounding: 0,
            formatter: null,
            name: {
              show: !0,
              bold: !0
            },
            value: {
              show: !0,
              bold: !1
            }
          }
        },
        legend: {
          ...b,
          roundingValue: 0,
          roundingPercentage: 0,
          showValue: !0,
          showPercentage: !0,
          position: "bottom"
        },
        title: n,
        tooltip: {
          ...g,
          roundingValue: 1
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, ae = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useBlurOnHover: !0,
    useCustomCells: !1,
    useAnimation: !0,
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        layout: {
          labels: {
            dataLabels: {
              ...w,
              prefix: "",
              suffix: "",
              formatter: null
            },
            captions: {
              show: !1,
              showSerieName: !1,
              serieNameAbbreviation: !0,
              serieNameMaxAbbreviationSize: 3,
              fontSize: t._12,
              showValue: !0,
              showPercentage: !0,
              roundingValue: 0,
              roundingPercentage: 0,
              offsetX: 0,
              offsetY: 0
            }
          },
          grid: {
            size: 10,
            spaceBetween: 2,
            vertical: !1
          },
          rect: {
            rounded: !0,
            rounding: 2,
            stroke: s,
            strokeWidth: 1,
            useGradient: !0,
            gradientIntensity: 40
          }
        },
        title: n,
        tooltip: {
          ...g,
          ...w,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0
        },
        legend: {
          ...b,
          ...w,
          roundingValue: 0,
          roundingPercentage: 0,
          showValue: !0,
          showPercentage: !0,
          position: "bottom"
        }
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, le = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        layout: {
          plots: {
            show: !0,
            radius: 2
          },
          outerPolygon: {
            stroke: h,
            strokeWidth: 1
          },
          dataPolygon: {
            strokeWidth: 1,
            transparent: !1,
            opacity: 20,
            useGradient: !0
          },
          grid: {
            show: !0,
            stroke: o,
            strokeWidth: 0.5,
            graduations: 5
          },
          labels: {
            dataLabels: {
              show: !0,
              fontSize: t._12,
              color: e
            }
          }
        },
        title: n,
        tooltip: {
          ...g,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0,
          animation: {
            show: !0,
            animationFrames: 60
          }
        },
        legend: {
          ...b,
          roundingPercentage: 0,
          position: "bottom"
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    translations: {
      target: "Target",
      value: "Value",
      datapoint: "Datapoint"
    }
  }, se = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    zoomAnimationFrames: 20,
    downsample: D,
    style: {
      fontFamily: "inherit",
      chart: {
        height: 512,
        width: 512,
        backgroundColor: r,
        color: e,
        layout: {
          labels: {
            quadrantLabels: {
              show: !0,
              tl: {
                text: "",
                color: e,
                fontSize: t._16,
                bold: !0
              },
              tr: {
                text: "",
                color: e,
                fontSize: t._16,
                bold: !0
              },
              br: {
                text: "",
                color: e,
                fontSize: t._16,
                bold: !0
              },
              bl: {
                text: "",
                color: e,
                fontSize: t._16,
                bold: !0
              }
            },
            plotLabels: {
              showAsTag: !1,
              show: !0,
              fontSize: t._10,
              color: e,
              offsetY: 8,
              rounding: 0,
              x: {
                formatter: null
              },
              y: {
                formatter: null
              }
            },
            axisLabels: {
              show: !0,
              fontSize: t._14,
              color: {
                positive: e,
                negative: e
              }
            }
          },
          grid: {
            stroke: o,
            strokeWidth: 1.5,
            showArrows: !0,
            graduations: {
              stroke: o,
              strokeWidth: 0.5,
              show: !0,
              steps: 5,
              fill: !0,
              color: o,
              roundingForce: 10
            },
            xAxis: {
              min: -100,
              max: 100,
              auto: !0,
              name: ""
            },
            yAxis: {
              min: -100,
              max: 100,
              auto: !0,
              name: ""
            }
          },
          plots: {
            radius: 6,
            outline: !0,
            outlineColor: l,
            outlineWidth: 1
          },
          areas: {
            show: !0,
            opacity: 40,
            useGradient: !0
          }
        },
        title: n,
        tooltip: {
          ...g,
          roundingValue: 0,
          showShape: !0
        },
        legend: {
          ...b,
          position: "bottom"
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    translations: {
      category: "Category",
      item: "Item",
      side: "Side"
    }
  }, ne = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    theme: "",
    customPalette: [],
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        animation: {
          use: !0,
          speed: 1,
          acceleration: 1
        },
        layout: {
          radiusRatio: 1,
          track: {
            size: 1,
            useGradient: !0,
            gradientIntensity: 20
          },
          markers: {
            show: !0,
            color: e,
            bold: !0,
            fontSizeRatio: 1,
            offsetY: 0,
            roundingValue: 0,
            formatter: null,
            prefix: "",
            suffix: ""
          },
          segmentSeparators: {
            show: !1,
            offsetOut: 0,
            offsetIn: 0,
            stroke: s,
            strokeWidth: 2
          },
          segmentNames: {
            show: !0,
            curved: !0,
            offsetRatio: 1.1,
            fontSize: t._16,
            minFontSize: _,
            // v3
            useSerieColor: !0,
            color: e,
            bold: !1
          },
          indicatorArc: {
            show: !1,
            radius: 123,
            fill: o
          },
          pointer: {
            show: !0,
            type: "pointy",
            size: 1,
            stroke: l,
            strokeWidth: 12,
            useRatingColor: !0,
            color: h,
            circle: {
              radius: 10,
              stroke: s,
              strokeWidth: 2,
              color: l
            }
          }
        },
        legend: {
          show: !0,
          fontSize: t._48,
          prefix: "",
          suffix: "",
          roundingValue: 1,
          showPlusSymbol: !0,
          useRatingColor: !0,
          color: e,
          formatter: null
        },
        title: n
      }
    },
    userOptions: i({
      pdf: !0,
      img: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    translations: {
      base: "Base"
    }
  }, ie = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    theme: "",
    layout: "classic",
    // or '3d'
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        animation: {
          use: !0,
          speed: 0.5,
          acceleration: 1
        },
        layout: {
          wheel: {
            radiusRatio: 1,
            tiltAngle3d: 50,
            ticks: {
              type: "classic",
              // or 'arc'
              rounded: !0,
              inactiveColor: o,
              activeColor: c,
              sizeRatio: 0.9,
              quantity: 100,
              // min: 12, max: 200
              strokeWidth: 5,
              stroke: "transparent",
              spacingRatio3d: 1,
              shadeColorRatio3d: 0.15,
              depth3d: 0,
              gradient: {
                show: !0,
                shiftHueIntensity: 100
              }
            }
          },
          innerCircle: {
            show: !0,
            stroke: o,
            strokeWidth: 1,
            radiusRatio: 1
          },
          percentage: {
            show: !0,
            fontSize: t._48,
            rounding: 1,
            bold: !0,
            formatter: null,
            offsetX: 0,
            offsetY: 0,
            stroke: "transparent",
            strokeWidth: 0
          }
        },
        title: n
      }
    },
    userOptions: i({
      pdf: !0,
      img: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    })
  }, ue = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    theme: "",
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        width: 312,
        // v3
        height: 56,
        // v3
        animation: {
          use: !0,
          speed: 0.5,
          acceleration: 1
        },
        layout: {
          display: "horizontal",
          crescendo: !1,
          curved: !1,
          curveAngleX: 10,
          curveAngleY: 10,
          activeColor: c,
          inactiveColor: o,
          ticks: {
            gradient: {
              show: !0,
              shiftHueIntensity: 100
            }
          }
        },
        percentage: {
          show: !0,
          useGradientColor: !0,
          color: c,
          fontSize: t._16,
          bold: !0,
          rounding: 1,
          verticalPosition: m.BOTTOM,
          horizontalPosition: m.LEFT,
          formatter: null
        },
        title: n
      }
    },
    userOptions: i({
      pdf: !0,
      img: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    })
  }, de = {
    debug: !1,
    // v3
    loading: !1,
    // v3
    theme: "",
    customPalette: [],
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        layout: {
          grandTotal: {
            show: !0,
            fontSize: t._20,
            bold: !0,
            prefix: "",
            suffix: "",
            roundingValue: 0,
            color: e,
            text: "Grand total",
            offsetY: 0,
            formatter: null
          },
          roots: {
            stroke: l,
            strokeWidth: 5,
            useGradient: !0,
            gradientIntensity: 20,
            underlayerColor: l,
            labels: {
              show: !0,
              fontSize: t._16,
              adaptColorToBackground: !0,
              color: l,
              bold: !0,
              roundingValue: 0,
              prefix: "",
              suffix: "",
              formatter: null,
              name: {
                color: e,
                fontSize: t._16,
                bold: !0
              }
            }
          },
          verticalSeparator: {
            stroke: l,
            strokeWidth: 5
          },
          links: {
            opacity: 10
          },
          branches: {
            stroke: l,
            strokeWidth: 0,
            borderRadius: 6,
            useGradient: !0,
            gradientIntensity: 20,
            underlayerColor: l,
            widthRatio: 1.5,
            labels: {
              show: !0,
              fontSize: t._14,
              color: e,
              bold: !0,
              dataLabels: {
                show: !0,
                hideUnderValue: 5,
                fontSize: t._14,
                roundingValue: 0,
                roundingPercentage: 0,
                prefix: "",
                suffix: "",
                formatter: null
              }
            }
          },
          nuts: {
            offsetX: 20,
            useGradient: !0,
            gradientIntensity: 30,
            selected: {
              useMotion: !0,
              useGradient: !0,
              gradientIntensity: 40,
              roundingValue: 0,
              roundingPercentage: 0,
              labels: {
                dataLabels: {
                  hideUnderValue: 5,
                  color: e,
                  fontSize: t._12,
                  bold: !0,
                  prefix: "",
                  suffix: "",
                  formatter: null
                },
                core: {
                  total: {
                    color: e,
                    fontSize: t._24,
                    bold: !1
                  },
                  value: {
                    color: e,
                    fontSize: t._24,
                    bold: !0,
                    prefix: "",
                    suffix: ""
                  }
                }
              }
            }
          },
          legend: {
            fontSize: t._16,
            color: e,
            roundingValue: 0,
            roundingPercentage: 0,
            prefix: "",
            suffix: ""
          },
          title: {
            ...n,
            offsetY: 0,
            subtitle: {
              ...n.subtitle,
              offsetY: 0
            }
          }
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      th: {
        ...u,
        translations: {
          rootName: "root name",
          rootValue: "root value",
          rootToTotal: "%/total",
          branchName: "branch name",
          branchValue: "branch value",
          branchToRoot: "%/root",
          branchToTotal: "%/total",
          nutName: "nut name",
          nutValue: "nut value",
          nutToBranch: "%/branch",
          nutToRoot: "%/root",
          nutToTotal: "%/total"
        }
      },
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    },
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    translations: {
      total: "Total",
      proportionToTree: "of grand total",
      of: "of"
    }
  }, fe = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    useStartAnimation: !0,
    useBlurOnHover: !0,
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        useGradient: !0,
        gradientIntensity: 20,
        layout: {
          maxThickness: 64,
          gutter: {
            color: o,
            width: 0.62
          },
          track: {
            width: 0.62
          },
          labels: {
            ...w,
            show: !0,
            fontSize: t._14,
            minFontSize: _,
            // v3
            color: e,
            roundingValue: 0,
            roundingPercentage: 0,
            bold: !0,
            offsetX: 0,
            offsetY: 0,
            value: {
              show: !0,
              formatter: null
            },
            percentage: {
              show: !0
            }
          }
        },
        title: n,
        legend: {
          ...b,
          ...w,
          roundingValue: 0,
          roundingPercentage: 0,
          showValue: !0,
          showPercentage: !0,
          position: "bottom"
        },
        tooltip: {
          ...g,
          ...w,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0
        }
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      },
      translations: {
        value: "Value",
        percentage: "Percentage",
        serie: "Serie"
      }
    }
  }, B = {
    // v3 renamed to _horizontal_ (yet still works)
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    autoSize: !0,
    // Legacy - removed
    responsive: !1,
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v3 (v2 = true)
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      datapointEnter: null,
      datapointLeave: null,
      datapointClick: null
    },
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        width: 512,
        height: 316,
        layout: {
          bars: {
            rowColor: null,
            rowRadius: 4,
            sort: ee.DESC,
            useStroke: !1,
            strokeWidth: 2,
            height: 32,
            // legacy - unused now
            gap: 6,
            borderRadius: 4,
            offsetX: 12,
            paddingRight: 0,
            useGradient: !0,
            gradientIntensity: 20,
            fillOpacity: 90,
            underlayerColor: l,
            dataLabels: {
              color: e,
              bold: !0,
              fontSize: t._14,
              // v3 increased
              value: {
                show: !0,
                roundingValue: 0,
                prefix: "",
                suffix: "",
                formatter: null
              },
              percentage: {
                show: !0,
                roundingPercentage: 0
              },
              offsetX: 0
            },
            nameLabels: {
              show: !0,
              color: e,
              bold: !1,
              fontSize: t._14,
              // v3 increased
              offsetX: 0
            },
            parentLabels: {
              show: !0,
              color: e,
              bold: !1,
              fontSize: t._14,
              // v3 increased
              offsetX: 0,
              paddingBottom: 0
            }
          },
          highlighter: {
            color: s,
            opacity: 5
          },
          separators: {
            show: !1,
            color: o,
            strokeWidth: 1,
            fullWidth: !0
          }
        },
        title: n,
        legend: {
          ...b,
          position: m.TOP,
          roundingValue: 0,
          roundingPercentage: 0,
          prefix: "",
          suffix: ""
        },
        tooltip: {
          ...g,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0,
          prefix: "",
          suffix: ""
        }
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      sort: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0,
        prefix: "",
        suffix: ""
      }
    },
    translations: {
      parentName: "Serie",
      childName: "Child",
      value: "value",
      percentageToTotal: "%/total",
      percentageToSerie: "%/serie"
    }
  }, ce = B, he = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      color: e,
      layout: {
        height: 300,
        // v3
        width: 1e3,
        // v3
        padding: p([0, 0, 0, 0]),
        crosshairs: {
          show: !1,
          stroke: s,
          strokeWidth: 1,
          strokeDasharray: 0
        },
        cells: {
          // height: 36, // v3 deprecated
          rowTotal: {
            value: {
              show: !1
            },
            color: {
              show: !1
            }
          },
          columnTotal: {
            value: {
              show: !1,
              rotation: 0,
              autoRotate: {
                // v3
                enable: !0,
                // v3
                angle: -30
                // v3
              },
              offsetX: 0,
              offsetY: 0
            },
            color: {
              show: !1
            }
          },
          value: {
            show: !1,
            fontSize: t._18,
            bold: !1,
            roundingValue: 0,
            color: e,
            formatter: null
          },
          colors: {
            hot: z,
            cold: c,
            underlayer: l
          },
          spacing: 2,
          selected: {
            border: 4,
            color: s
          }
        },
        dataLabels: {
          prefix: "",
          suffix: "",
          xAxis: {
            show: !0,
            values: [],
            datetimeFormatter: C,
            showOnlyAtModulo: null,
            rotation: 0,
            autoRotate: {
              // v3
              enable: !0,
              // v3
              angle: -30
              // v3
            },
            fontSize: t._10,
            color: e,
            bold: !1,
            offsetX: 0,
            offsetY: 0
          },
          yAxis: {
            show: !0,
            values: [],
            datetimeFormatter: C,
            fontSize: t._10,
            color: e,
            bold: !1,
            offsetX: 0,
            offsetY: 0
          }
        }
      },
      title: n,
      legend: {
        ...b,
        fontSize: t._12,
        roundingValue: 0,
        width: 24
        // position: POSITION.RIGHT, // v3 deprecated
        // scaleBorderRadius: 18 // v3 deprecated
      },
      tooltip: {
        ...g,
        roundingValue: 0
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      colNames: {
        xAxis: "X"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0
      }
    }
  }, ge = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    downsample: D,
    usePerformanceMode: !1,
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      color: e,
      layout: {
        height: 316,
        width: 512,
        padding: p([0, 0, 0, 0]),
        // v3 modification
        axis: {
          show: !0,
          stroke: o,
          strokeWidth: 1
        },
        marginalBars: {
          show: !1,
          size: 40,
          tranches: 20,
          opacity: 0.6,
          fill: s,
          strokeWidth: 1,
          offset: 20,
          borderRadius: 2,
          useGradient: !0,
          showLines: !1,
          linesStrokeWidth: 1,
          highlighter: {
            show: !0,
            opacity: 0.1,
            color: s,
            stroke: s,
            strokeWidth: 0.5,
            strokeDasharray: 2,
            highlightBothAxes: !1
          }
        },
        plots: {
          radius: 2,
          stroke: l,
          strokeWidth: 0.3,
          opacity: 0.6,
          significance: {
            show: !0,
            useDistanceOpacity: !1,
            deviationThreshold: 10,
            opacity: 0.3
          },
          deviation: {
            translation: "deviation",
            roundingValue: 1
          },
          giftWrap: {
            show: !1,
            strokeWidth: 1,
            strokeDasharray: 0,
            fillOpacity: 0.2
          },
          selectors: {
            show: !0,
            stroke: s,
            strokeWidth: 0.7,
            strokeDasharray: 0,
            labels: {
              fontSize: t._12,
              color: e,
              rounding: 2,
              bold: !1,
              showName: !0,
              prefix: "",
              suffix: "",
              x: {
                formatter: null
              },
              y: {
                formatter: null
              }
            },
            markers: {
              radius: 1.5,
              stroke: l,
              strokeWidth: 0.5,
              fill: s
            }
          }
        },
        correlation: {
          show: !0,
          strokeDasharray: 2,
          strokeWidth: 1,
          label: {
            show: !0,
            fontSize: t._12,
            color: e,
            bold: !0,
            roundingValue: 2,
            useSerieColor: !0
          }
        },
        dataLabels: {
          xAxis: {
            name: "",
            show: !0,
            fontSize: t._10,
            color: e,
            bold: !1,
            offsetX: 0,
            offsetY: 0,
            roundingValue: 0
          },
          yAxis: {
            name: "",
            show: !0,
            fontSize: t._10,
            color: e,
            bold: !1,
            offsetX: 0,
            offsetY: 0,
            roundingValue: 0
          }
        }
      },
      title: n,
      legend: {
        ...b,
        roundingValue: 0,
        position: "bottom"
      },
      tooltip: {
        ...g,
        roundingValue: 2,
        showShape: !0,
        prefix: "",
        suffix: ""
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 2,
        roundingAverage: 1
      },
      translations: {
        correlationCoefficient: "Correlation Coef.",
        nbrPlots: "Nbr plots",
        average: "Average",
        series: "Series"
      }
    }
  }, pe = {
    skeletonDataset: null,
    skeletonConfig: null,
    type: "candlestick",
    // or 'ohlc'
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    responsiveProportionalSizing: !0,
    a11y: a,
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    useCssAnimation: !1,
    // v3 (v2 = true)
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      color: e,
      height: 316,
      width: 512,
      layout: {
        padding: p([0, 0, 0, 0]),
        selector: {
          color: o,
          opacity: 10
        },
        grid: {
          show: !0,
          stroke: o,
          strokeWidth: 0.5,
          verticalLines: {
            show: !1,
            strokeDasharray: 0,
            strokeWidth: 0.5,
            stroke: o
          },
          horizontalLines: {
            show: !1,
            strokeDasharray: 0,
            strokeWidth: 0.5,
            stroke: o
          },
          xAxis: {
            ticks: {
              show: !0
            },
            dataLabels: {
              show: !0,
              fontSize: t._10,
              color: e,
              offsetY: 0,
              bold: !1,
              rotation: 0,
              autoRotate: {
                // v3
                enable: !0,
                // v3
                angle: -30
                // v3
              },
              datetimeFormatter: C,
              showOnlyFirstAndLast: !1,
              showOnlyAtModulo: !1,
              modulo: 12
            }
          },
          yAxis: {
            scale: {
              min: null,
              max: null
            },
            dataLabels: {
              show: !0,
              fontSize: t._12,
              color: e,
              roundingValue: 0,
              offsetX: 0,
              bold: !1,
              steps: 10,
              prefix: "",
              suffix: ""
            }
          }
        },
        wick: {
          stroke: s,
          strokeWidth: 0.5,
          extremity: {
            shape: v.LINE,
            size: "auto",
            color: s
          }
        },
        candle: {
          borderRadius: 1,
          stroke: s,
          strokeWidth: 0.5,
          colors: {
            bearish: z,
            bullish: O
          },
          gradient: {
            show: !0,
            underlayer: l
          },
          widthRatio: 0.5
        }
      },
      zoom: {
        ...P,
        minimap: F,
        preview: {
          enable: !0,
          fill: "#CCCCCC50",
          stroke: "#6A6A6A",
          strokeWidth: 2,
          strokeDasharray: 0
        },
        useDefaultFormat: !0,
        timeFormat: "yyyy-MM-dd HH:mm:ss",
        // When datetimeFormatter is enabled and useDefaultFormat is false
        customFormat: null
        // overrides all if callback => string
      },
      title: n,
      tooltip: {
        ...g,
        roundingValue: 0,
        prefix: "",
        suffix: "",
        showChart: !0,
        useDefaultTimeFormat: !0,
        timeFormat: "yyyy-MM-dd HH:mm:ss"
        // When datetimeFormatter is enabled and useDefaultFormat is false
      }
    },
    translations: {
      period: "Period",
      open: "Open",
      high: "High",
      low: "Low",
      last: "Close",
      volume: "Volume"
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      th: u,
      useDialog: !1,
      td: {
        ...d,
        roundingValue: 2,
        prefix: "",
        suffix: ""
      }
    }
  }, be = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    theme: "",
    responsive: !1,
    type: v.LINE,
    downsample: D,
    a11y: a,
    translations: {
      period: "period",
      value: "value"
    },
    // `gradientPath` creates multiple path segments to fake a gradient that follows the path
    // Only 2 colors are possible.
    // Could be considered 'experimental'
    gradientPath: {
      show: !1,
      segments: 256,
      colors: {
        high: "#34eb96",
        low: "#eb4034"
      }
    },
    // `temperatureColors` applies a color linearGradient def to the exsiting path.
    // It does not technically follow the path, but is more performant than `gradientPath`
    // and allows the usage of any number of colors in its `colors` array. 
    temperatureColors: {
      show: !1,
      colors: ["#34eb96", "#eb4034"]
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    style: {
      chartWidth: 290,
      animation: {
        show: !0,
        animationFrames: 360
      },
      padding: p([12, 12, 3, 0]),
      fontFamily: "inherit",
      backgroundColor: r,
      scaleMin: null,
      scaleMax: null,
      line: {
        color: c,
        strokeWidth: 3,
        smooth: !1,
        pulse: {
          show: !1,
          loop: !0,
          color: l,
          durationMs: 4e3,
          easing: "ease-in-out",
          radius: 2,
          cubicBezier: [0.45, 0.45, 0.55, 0.55],
          trail: {
            show: !1,
            length: 6,
            opacity: 0.75
          }
        }
      },
      bar: {
        borderRadius: 3,
        color: c
      },
      zeroLine: {
        color: s,
        strokeWidth: 1
      },
      plot: {
        show: !0,
        radius: 4,
        stroke: l,
        strokeWidth: 1
      },
      verticalIndicator: {
        show: !0,
        strokeWidth: 1.5,
        color: c,
        strokeDasharray: 3
      },
      dataLabel: {
        show: !0,
        position: m.LEFT,
        offsetX: 0,
        offsetY: 0,
        fontSize: t._20,
        bold: !0,
        color: e,
        roundingValue: 0,
        valueType: "latest",
        prefix: "",
        suffix: "",
        formatter: null,
        datetimeFormatter: C
        // v3
      },
      title: {
        show: !0,
        textAlign: m.LEFT,
        color: e,
        fontSize: t._16,
        bold: !0,
        text: ""
      },
      tooltip: {
        show: !1,
        fontSize: t._14,
        color: e,
        backgroundColor: r,
        offsetY: 0,
        borderWidth: 0,
        borderColor: k,
        borderRadius: 2,
        backgroundOpacity: 100
      },
      area: {
        show: !0,
        useGradient: !0,
        opacity: 30,
        color: c
      }
    }
  }, me = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      animation: {
        show: !0,
        animationFrames: 60
      },
      layout: {
        independant: !0,
        percentage: !0,
        target: 0,
        showTargetValue: !1,
        targetValueText: ""
      },
      gutter: {
        backgroundColor: o,
        opacity: 100
      },
      bar: {
        gradient: {
          show: !0,
          intensity: 40,
          underlayerColor: l
        }
      },
      labels: {
        fontSize: t._16,
        name: {
          position: "top-left",
          width: "100%",
          color: e,
          bold: !1
        },
        value: {
          show: !0,
          bold: !0
        }
      },
      title: {
        text: "",
        color: e,
        fontSize: t._16,
        bold: !0,
        textAlign: m.LEFT,
        margin: "0 0 6px 0",
        subtitle: {
          color: T,
          text: "",
          fontSize: t._12,
          bold: !1
        }
      },
      gap: 4
    }
  }, ke = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    useCursorPointer: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        series: "Series",
        percentage: "Percentage",
        value: "Value"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      animation: {
        show: !0,
        animationFrames: 60
      },
      bar: {
        gradient: {
          show: !0,
          intensity: 40,
          underlayerColor: l
        }
      },
      legend: {
        show: !0,
        textAlign: m.LEFT,
        fontSize: t._12,
        margin: "6px 0 0 0",
        name: {
          color: e,
          bold: !1
        },
        value: {
          show: !0,
          bold: !1,
          color: e,
          prefix: "",
          suffix: "",
          rounding: 0,
          formatter: null
        },
        percentage: {
          show: !0,
          bold: !0,
          color: e,
          rounding: 1
        },
        selectAllToggle: {
          show: !1,
          backgroundColor: k,
          color: e
        }
      },
      title: {
        ...n,
        textAlign: m.LEFT,
        margin: "0 0 6px 0"
      },
      tooltip: {
        ...g
      }
    }
  }, we = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        series: "Series",
        time: "Time",
        value: "Value",
        label: "Label"
      }
    },
    events: {
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      animation: {
        show: !0,
        speedMs: 500
      },
      layout: {
        height: 96,
        width: 640,
        padding: p([24, 0, 36, 0])
      },
      bars: {
        shape: v.SQUARE,
        strokeWidth: 0,
        colors: {
          positive: c,
          negative: z,
          gradient: {
            show: !0
          }
        },
        borderRadius: 24,
        gap: 12
      },
      labels: {
        value: {
          show: !0,
          fontSize: t._14,
          minFontSize: _,
          // v3
          color: e,
          bold: !0,
          rounding: 1,
          prefix: "",
          suffix: "",
          offsetY: 0,
          formatter: null
        },
        valueLabel: {
          show: !0,
          fontSize: t._14,
          minFontSize: _,
          // v3
          color: e,
          bold: !1,
          rounding: 0
        },
        timeLabel: {
          show: !0,
          fontSize: t._12,
          minFontSize: _,
          // v3
          color: e,
          bold: !1
        }
      },
      selector: {
        stroke: c,
        fill: "#2D353C10",
        strokeWidth: 2,
        strokeDasharray: 0,
        borderRadius: 2
      },
      title: {
        text: "",
        color: e,
        fontSize: t._16,
        bold: !0,
        textAlign: m.LEFT,
        subtitle: {
          color: T,
          text: "",
          fontSize: t._12,
          bold: !1
        },
        margin: "0 0 6px 0"
      }
    }
  }, ye = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    theme: "",
    style: {
      fontFamily: "inherit",
      background: l,
      height: 84,
      basePosition: 72,
      animation: {
        show: !0,
        speedMs: 150
      },
      title: {
        show: !0,
        fontSize: t._12,
        position: m.TOP,
        textAlign: m.CENTER,
        bold: !1,
        color: e
      },
      dataLabel: {
        fontSize: t._20,
        autoColor: !0,
        color: e,
        offsetY: 0,
        bold: !0,
        rounding: 0,
        prefix: "",
        suffix: "",
        formatter: null
      },
      colors: {
        min: z,
        max: O,
        showGradient: !0
      },
      track: {
        autoColor: !0,
        color: c,
        strokeLinecap: v.ROUND
      },
      gutter: {
        color: o,
        strokeLinecap: v.ROUND
      }
    }
  }, ve = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    theme: "",
    downsample: D,
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      height: 80,
      width: 300,
      animation: {
        show: !0,
        animationFrames: 20
      },
      line: {
        stroke: s,
        strokeWidth: 2,
        strokeLinecap: v.ROUND,
        strokeLinejoin: v.ROUND,
        smooth: !0,
        useColorTrend: !0
      },
      area: {
        show: !0,
        useGradient: !0,
        opacity: 20
      },
      dataLabel: {
        show: !0,
        useColorTrend: !0,
        color: e,
        fontSize: t._14,
        bold: !1,
        prefix: "",
        suffix: "",
        rounding: 0,
        formatter: null
      },
      trendLabel: {
        trendType: "global",
        useColorTrend: !0,
        color: e,
        fontSize: t._14,
        bold: !0,
        rounding: 0
      },
      arrow: {
        colors: {
          positive: O,
          neutral: R,
          negative: z
        }
      },
      padding: p([12, 12, 12, 82])
    }
  }, Ce = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    a11y: a,
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    responsive: !1,
    theme: "",
    axisLabelsFontSize: t._12,
    backgroundColor: r,
    barGap: 12,
    barAnimated: !0,
    barStrokeWidth: 1,
    blurOnHover: !0,
    chartIsBarUnderDatasetLength: 6,
    color: e,
    customPalette: [],
    dataLabelFontSize: t._14,
    dataLabelRoundingPercentage: 1,
    dataLabelRoundingValue: 0,
    donutHideLabelUnderPercentage: 3,
    donutCurvedMarkers: !0,
    // v3
    donutLabelMarkerStrokeWidth: 1,
    donutRadiusRatio: 0.4,
    donutShowTotal: !0,
    donutStrokeWidth: 2,
    donutStroke: r,
    donutThicknessRatio: 0.18,
    donutTotalLabelFontSize: t._24,
    donutTotalLabelOffsetY: 0,
    donutTotalLabelText: "Total",
    donutUseShadow: !1,
    donutShadowColor: s,
    fontFamily: "inherit",
    formatter: null,
    height: 338,
    legendFontSize: t._12,
    legendIcon: "circleFill",
    legendIconSize: t._12,
    legendPosition: "bottom",
    showLegendSelectAllToggle: !1,
    legendSelectAllToggleBackgroundColor: k,
    legendSelectAllToggleColor: e,
    lineAnimated: !0,
    lineSmooth: !0,
    lineStrokeWidth: 2,
    paletteStartIndex: 0,
    showDataLabels: !0,
    showLegend: !0,
    showTooltip: !0,
    showUserOptions: !0,
    userOptionsPosition: "right",
    showUserOptionsOnChartHover: !1,
    keepUserOptionsStateOnChartLeave: !0,
    useCursorPointer: !1,
    userOptionsButtons: {
      tooltip: !0,
      pdf: !0,
      img: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0,
      altCopy: !1
    },
    userOptionsButtonTitles: {
      open: "Open options",
      close: "Close options",
      tooltip: "Toggle tooltip",
      pdf: "Download PDF",
      img: "Download PNG",
      fullscreen: "Toggle fullscreen",
      annotator: "Toggle annotator",
      svg: "Download SVG",
      altCopy: "Copy alt text"
    },
    userOptionsPrint: {
      overflowTolerance: 0.2,
      orientation: "auto",
      scale: 2
    },
    userOptionsCallbacks: {
      tooltip: null,
      pdf: null,
      img: null,
      fullscreen: null,
      annotator: null,
      svg: null
    },
    title: "",
    titleBold: !0,
    titleFontSize: t._16,
    titleTextAlign: "center",
    tooltipTeleportTo: "body",
    tooltipCustomFormat: null,
    tooltipBackgroundOpacity: 100,
    tooltipBorderRadius: 4,
    tooltipBorderColor: W,
    tooltipBorderWidth: 1,
    tooltipFontSize: t._14,
    tooltipPosition: m.CENTER,
    tooltipOffsetY: 24,
    tooltipSmooth: !0,
    tooltipSmoothForce: 0.18,
    tooltipSmoothSnapThreshold: 0.25,
    tooltipBackdropFilter: !0,
    useCustomLegend: !1,
    valuePrefix: "",
    valueSuffix: "",
    width: 512,
    xAxisLabel: "",
    xyAxisStroke: o,
    xyAxisStrokeWidth: 1,
    xyGridStroke: o,
    xyGridStrokeWidth: 0.5,
    xyHighlighterColor: s,
    xyHighlighterOpacity: 0.05,
    xyLabelsXFontSize: t._10,
    xyLabelsYFontSize: t._12,
    xyPaddingBottom: 12,
    xyPaddingLeft: 12,
    xyPaddingRight: 12,
    xyPaddingTop: 12,
    xyPeriodLabelsRotation: 0,
    xyPeriodLabelsAutoRotate: {
      // v3
      enable: !0,
      // v3
      angle: -30
      // v3
    },
    xyPeriods: [],
    datetimeFormatter: C,
    xyPeriodsShowOnlyAtModulo: !1,
    xyPeriodsModulo: 12,
    xyScaleSegments: 15,
    xyShowAxis: !0,
    xyShowGrid: !0,
    xyShowScale: !0,
    yAxisLabel: "",
    zoomXy: !0,
    zoomColor: h,
    zoomHighlightColor: V,
    zoomFontSize: t._14,
    zoomUseResetSlot: !1,
    zoomMinimap: M,
    zoomStartIndex: null,
    zoomEndIndex: null,
    zoomEnableRangeHandles: !0,
    zoomEnableSelectionDrag: !0,
    zoomFocusOnDrag: !1,
    zoomFocusRangeRatio: 0.2,
    zoomMaxWidth: null
  }, xe = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      color: e,
      height: 500,
      width: 500,
      layout: {
        padding: p([12, 12, 36, 12]),
        grid: {
          show: !0,
          stroke: o,
          strokeWidth: 1
        },
        dataLabels: {
          sideTitles: {
            show: !0,
            fontSize: t._18,
            color: e,
            useSideColor: !0,
            bold: !1,
            offsetY: 0
          },
          xAxis: {
            show: !0,
            fontSize: t._12,
            color: e,
            bold: !1,
            scale: 1e3,
            translation: "in thousands",
            formatter: null,
            rotation: 0,
            // v3
            autoRotate: {
              // v3
              enable: !0,
              // v3
              angle: -30
              // v3
            }
          },
          yAxis: {
            show: !0,
            display: "age",
            fontSize: t._12,
            color: e,
            bold: !1,
            showEvery: 5,
            formatter: null
          }
        },
        centerSlit: {
          width: 20
        },
        bars: {
          gap: 2,
          borderRadius: 2,
          left: {
            color: z
          },
          right: {
            color: c
          },
          gradient: {
            show: !0,
            underlayer: l,
            intensity: 60,
            shiftHue: 0.05
          }
        }
      },
      highlighter: {
        color: s,
        opacity: 5
      },
      title: n,
      tooltip: {
        ...g,
        roundingValue: 0
      }
    },
    translations: {
      age: "age",
      male: "male",
      female: "female",
      total: "total",
      year: "year"
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: d
    }
  }, Se = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    responsiveProportionalSizing: !0,
    theme: "",
    customPalette: [],
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      color: e,
      size: 400,
      limit: 50,
      animation: {
        show: !0,
        speedMs: 300
      },
      labels: {
        color: e,
        fontSize: t._14,
        minFontSize: _
        // v3
      },
      weightLabels: {
        size: 8,
        show: !0,
        formatter: null,
        prefix: "",
        suffix: "",
        rounding: 0
      },
      links: {
        curved: !1,
        maxWidth: 3
      },
      circle: {
        radiusProportion: 0.2,
        stroke: h,
        strokeWidth: 1,
        offsetY: 0
      },
      plot: {
        radius: 3,
        useSerieColor: !0,
        color: s
      },
      title: n
    },
    userOptions: i({
      pdf: !0,
      img: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    })
  }, _e = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    theme: "",
    customPalette: [],
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 360,
        width: 256,
        // v3
        thermometer: {
          width: 48
        },
        padding: {
          // v3 left and right are deprecated
          top: 12,
          bottom: 12
        },
        graduations: {
          show: !0,
          sides: "both",
          height: 2,
          stroke: o,
          strokeWidth: 1,
          showIntermediate: !0,
          gradient: {
            show: !0,
            intensity: 20
          }
        },
        animation: {
          use: !0,
          speedMs: 1e3
        },
        label: {
          show: !0,
          // v3
          fontSize: t._20,
          minFontSize: _,
          // v3
          rounding: 1,
          bold: !0,
          color: e,
          prefix: "",
          suffix: "",
          formatter: null
        }
      },
      title: n
    },
    userOptions: i({
      pdf: !0,
      img: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    })
  }, ze = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    useBlurOnHover: !0,
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        size: 360,
        layout: {
          labels: {
            dataLabels: {
              ...w,
              prefix: "",
              suffix: "",
              formatter: null,
              show: !1,
              offsetX: 0,
              fontSize: 10,
              color: e,
              bold: !1,
              roundingValue: 0,
              roundingPercentage: 0,
              showValue: !0,
              showPercentage: !0,
              // Markers are applied when labels are enabled
              markers: {
                stroke: "#8A8A8A",
                strokeWidth: 1,
                radius: 3,
                position: "right"
              }
            }
          },
          rings: {
            strokeWidth: 2,
            stroke: o,
            gradient: {
              show: !0,
              intensity: 40,
              underlayerColor: l
            },
            useShadow: !0
          }
        },
        legend: {
          ...b,
          roundingValue: 0,
          roundingPercentage: 0,
          showValue: !0,
          showPercentage: !0,
          position: "bottom"
        },
        title: n,
        tooltip: {
          ...g,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0
        }
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0,
      labels: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, Le = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    theme: "",
    a11y: a,
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    customPalette: [],
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        zoom: P,
        donuts: {
          hover: {
            hideLabelsUnderValue: 5
          }
        },
        dialog: {
          show: !0,
          backgroundColor: r,
          color: e,
          header: {
            backgroundColor: r,
            color: e
          },
          donutChart: {
            ...x,
            responsive: !0,
            userOptions: {
              ...x.userOptions,
              buttons: {
                ...x.userOptions.buttons,
                pdf: !1
              }
            },
            style: {
              ...x.style,
              chart: {
                ...x.style.chart,
                layout: {
                  ...x.style.chart.layout,
                  donut: {
                    ...x.style.chart.layout.donut,
                    strokeWidth: 64
                  },
                  labels: {
                    ...x.style.chart.layout.labels,
                    percentage: {
                      ...x.style.chart.layout.labels.percentage,
                      fontSize: 16
                    },
                    name: {
                      ...x.style.chart.layout.labels.name,
                      fontSize: 12
                    },
                    hollow: {
                      ...x.style.chart.layout.labels.hollow,
                      average: {
                        ...x.style.chart.layout.labels.hollow.average,
                        show: !1
                      }
                    }
                  }
                }
              }
            }
          }
        },
        layout: {
          height: 316,
          width: 512,
          padding: p([5, 10, 5, 10]),
          grid: {
            show: !0,
            stroke: o,
            strokeWidth: 0.7,
            showVerticalLines: !0,
            axis: {
              yLabel: "",
              yLabelOffsetX: 0,
              xLabel: "",
              xLabelOffsetY: 0,
              fontSize: t._14,
              color: e
            },
            yAxis: {
              scaleMin: null,
              scaleMax: null,
              autoScale: !1,
              dataLabels: {
                show: !0,
                fontSize: t._10,
                color: e,
                roundingValue: 0,
                offsetX: 0,
                bold: !1,
                steps: 10
              }
            },
            xAxis: {
              dataLabels: {
                show: !0,
                values: [],
                datetimeFormatter: C,
                fontSize: t._10,
                showOnlyFirstAndLast: !1,
                color: e,
                rotation: 0,
                autoRotate: {
                  // v3
                  enable: !0,
                  // v3
                  angle: -30
                  // v3
                },
                offsetY: 0
              }
            }
          },
          line: {
            show: !0,
            stroke: h,
            strokeWidth: 4
          },
          highlighter: {
            color: s,
            opacity: 5
          },
          dataLabels: {
            show: !0,
            fontSize: t._10,
            color: e,
            bold: !1,
            rounding: 0,
            prefix: "",
            suffix: "",
            offsetY: 0,
            formatter: null
          }
        },
        title: n,
        legend: {
          ...b,
          roundingValue: 0,
          roundingPercentage: 0,
          showValue: !0,
          showPercentage: !0,
          position: "bottom"
        }
      }
    },
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        period: "Period",
        total: "Total"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, Pe = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        layout: {
          grid: {
            show: !0,
            stroke: o,
            strokeWidth: 0.5
          },
          outerPolygon: {
            stroke: o,
            strokeWidth: 1
          },
          dataPolygon: {
            color: c,
            opacity: 60,
            gradient: {
              show: !0,
              intensity: 1
            },
            stroke: c,
            strokeWidth: 0
          },
          smileys: {
            strokeWidth: 1,
            colors: {
              1: "#e20001",
              2: "#ff9f03",
              3: "#ffd004",
              4: "#9ac900",
              5: "#059f00"
            }
          },
          dataLabel: {
            color: e,
            roundingValue: 0,
            roundingPercentage: 0,
            bold: !0,
            prefix: "",
            suffix: "",
            formatter: null
          }
        },
        title: n,
        legend: {
          ...b,
          roundingValue: 0,
          roundingPercentage: 0,
          position: "bottom"
        }
      }
    },
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, Ae = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        nodes: {
          stroke: l,
          strokeHovered: s
        },
        links: {
          stroke: h
        },
        title: n,
        tooltip: g,
        zoom: {
          show: !0,
          speed: 1
        }
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0,
      zoom: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: d,
      translations: {
        nodeName: "Node name",
        details: "Details",
        ancestor: "Parent node"
      }
    }
  }, Te = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v3 (v2 = true)
    useBlurOnHover: !0,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    serieToggleAnimation: {
      show: !0,
      durationMs: 500
    },
    startAnimation: {
      show: !1,
      // v3 (v2 = false)
      durationMs: 1e3,
      staggerMs: 50
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        useGradient: !0,
        gradientIntensity: 40,
        padding: p([0, 0, 0, 0]),
        width: 512,
        height: 512,
        layout: {
          labels: {
            dataLabels: {
              ...w,
              show: !0,
              prefix: "",
              suffix: "",
              hideUnderValue: 3,
              fontSize: t._14,
              offsetX: 4,
              offsetY: 12,
              useSerieColor: !1,
              color: e,
              showValue: !0,
              showPercentage: !0,
              boldValue: !1,
              boldPercentage: !0,
              roundingValue: 0,
              roundingPercentage: 0,
              showDonutName: !0,
              boldDonutName: !0,
              curvedDonutName: !0,
              // v3
              donutNameAbbreviation: !1,
              // v2 = true
              donutNameMaxAbbreviationSize: 3,
              donutNameOffsetY: 0,
              formatter: null
            }
          },
          donut: {
            strokeWidth: 200,
            borderWidth: 2,
            spacingRatio: 0.5,
            useShadow: !1,
            shadowColor: s,
            emptyFill: o,
            selectedColor: "#0000001A",
            borderColorAuto: !0,
            borderColor: k
          }
        },
        legend: {
          ...b,
          ...w,
          roundingValue: 0,
          roundingPercentage: 0,
          showValue: !0,
          showPercentage: !0,
          position: "bottom"
        },
        title: n,
        tooltip: {
          ...g,
          ...w,
          showAllItemsAtIndex: !0,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, De = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    useBlurOnHover: !0,
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        layout: {
          arcs: {
            strokeWidth: 24,
            borderWidth: 12,
            offsetX: 0,
            offsetY: 0,
            hoverEffect: {
              show: !0,
              multiplicator: 1.1
            },
            gradient: {
              show: !0,
              intensity: 30,
              color: l
            }
          },
          labels: {
            dataLabels: {
              ...w,
              prefix: "",
              suffix: "",
              formatter: null
            }
          }
        },
        legend: {
          ...b,
          roundingValue: 0,
          roundingPercentage: 0,
          showValue: !0,
          showPercentage: !0,
          position: "bottom"
        },
        title: n,
        tooltip: {
          ...g,
          showValue: !0,
          showPercentage: !0,
          roundingValue: 0,
          roundingPercentage: 0
        }
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, Oe = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    responsiveProportionalSizing: !0,
    theme: "",
    customPalette: [],
    useCssAnimation: !0,
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 600,
        width: 600,
        // v3
        // stripWidth: 120, v3 deprecated
        padding: p([12, 12, 12, 12]),
        // v3 modified
        grid: {
          show: !0,
          stroke: h,
          strokeWidth: 1,
          scaleSteps: 10,
          horizontalGrid: {
            show: !0,
            stroke: h,
            strokeWidth: 0.5,
            strokeDasharray: 4
          },
          verticalGrid: {
            show: !0,
            stroke: h,
            strokeWidth: 0.5,
            strokeDasharray: 4
          }
        },
        plots: {
          opacity: 0.5,
          radius: 20,
          stroke: l,
          strokeWidth: 1,
          shape: v.CIRCLE,
          gradient: {
            show: !0,
            intensity: 40
          }
        },
        labels: {
          prefix: "",
          suffix: "",
          formatter: null,
          bestPlotLabel: {
            show: !0,
            showValue: !0,
            fontSize: t._14,
            color: e,
            rounding: 0,
            offsetY: 0
          },
          axis: {
            xLabel: "",
            xLabelOffsetY: 0,
            yLabel: "",
            yLabelOffsetX: 0,
            fontSize: t._14,
            color: e
          },
          xAxisLabels: {
            show: !0,
            color: e,
            fontSize: t._14,
            offsetY: 0,
            rotation: 0,
            // v3,
            autoRotate: {
              // v3
              enable: !0,
              // v3
              angle: -30
              // v3
            }
          },
          yAxisLabels: {
            show: !0,
            color: e,
            fontSize: t._14,
            rounding: 0,
            offsetX: 0
          }
        },
        title: n,
        tooltip: {
          ...g,
          roundingValue: 0
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0
      }
    }
  }, Fe = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    useAnimation: !1,
    // v2 = true
    animationSpeed: 2,
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        width: 600,
        rowHeight: 48,
        // v3 modified
        padding: p([12, 24, 12, 12]),
        // v3 modified
        plots: {
          startColor: z,
          endColor: c,
          evaluationColors: {
            // v3
            enable: !1,
            // v3
            positive: "#2ca02c",
            // v3
            negative: "#d62728",
            // v3
            neutral: "#c7c7c7"
            // v3
          },
          radius: 6,
          stroke: l,
          strokeWidth: 1,
          link: {
            strokeWidth: 2,
            type: v.CURVED
          },
          gradient: {
            show: !0,
            intensity: 40
          }
        },
        grid: {
          strokeWidth: 1,
          scaleSteps: 10,
          scaleMin: null,
          // v3
          scaleMax: null,
          // V3
          horizontalGrid: {
            show: !0,
            stroke: h,
            strokeWidth: 0.5,
            strokeDasharray: 0
          },
          verticalGrid: {
            show: !0,
            stroke: h,
            strokeWidth: 0.5,
            strokeDasharray: 0
          }
        },
        comparisonLines: {
          // v3
          show: !0,
          // v3
          strokeWidth: 1,
          // v3
          strokeDasharray: 4,
          // v3
          showRect: !0,
          // v3
          rectColor: s,
          // v3
          rectOpacity: 5,
          // v3
          showLabel: !0,
          // v3
          labelColor: s,
          // v3
          labelFontSize: t._12
          // v3
        },
        highlighter: {
          // v3
          color: s,
          // v3
          opacity: 5
          // v3
        },
        labels: {
          prefix: "",
          suffix: "",
          formatter: null,
          axis: {
            // v3
            yLabel: "",
            // v3
            yLabelOffsetX: 0,
            // v3
            xLabel: "",
            // v3
            xLabelOffsetY: 0,
            // v3
            fontSize: t._14,
            // v3
            color: e
            // v3
          },
          yAxisLabels: {
            show: !0,
            fontSize: t._14,
            color: e,
            offsetX: 0,
            bold: !0,
            showProgression: !0,
            rounding: 1,
            formatter: null
            // v3
          },
          xAxisLabels: {
            show: !0,
            fontSize: t._14,
            color: e,
            offsetY: 0,
            bold: !1,
            rounding: 0,
            rotation: 0,
            // v3
            autoRotate: {
              // v3
              enable: !0,
              // v3
              angle: -30
              // v3
            }
          },
          startLabels: {
            show: !0,
            fontSize: t._10,
            color: e,
            offsetY: 0,
            rounding: 0,
            useStartColor: !0,
            useEvaluationColor: !0
            // v3
          },
          endLabels: {
            show: !0,
            fontSize: t._10,
            color: e,
            offsetY: 0,
            rounding: 0,
            useEndColor: !0,
            useEvaluationColor: !0
            // v3
          }
        },
        legend: {
          ...b,
          labelStart: "start",
          labelEnd: "end",
          labelPositive: "positive",
          // v3
          labelNegative: "negative",
          // v3
          labelNeutral: "neutral",
          // v3
          position: "bottom"
        },
        title: n
      }
    },
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        start: "Start value",
        end: "End value",
        progression: "Progression"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, Re = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v3
    style: {
      fontFamily: "inherit",
      shape: v.BAR,
      chart: {
        animation: {
          use: !0,
          speed: 1,
          acceleration: 1
        },
        backgroundColor: r,
        color: e,
        bar: {
          color: c,
          stroke: c,
          strokeWidth: 0.7,
          shadeColor: s
        },
        box: {
          stroke: h,
          strokeWidth: 0.7,
          strokeDasharray: 2,
          dimensions: {
            width: 128,
            height: 256,
            top: 27,
            bottom: 9,
            left: 24,
            right: 24,
            perspective: 18
          }
        },
        title: n,
        legend: {
          showDefault: !0,
          fontSize: t._10,
          color: e,
          bold: !1,
          roundingValue: 0,
          roundingPercentage: 0,
          prefix: "",
          suffix: "",
          hideUnderPercentage: 3
        },
        dataLabel: {
          show: !0,
          bold: !0,
          color: c,
          fontSize: t._12,
          rounding: 1,
          formatter: null
        }
      }
    },
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        value: "Value",
        percentage: "Percentage"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      }
    }
  }, We = {
    theme: "",
    customPalette: [],
    responsiveBreakpoint: 500,
    showAverage: !0,
    showMedian: !0,
    showTotal: !0,
    roundingAverage: 0,
    roundingMedian: 0,
    roundingValues: 0,
    roundingTotal: 0,
    prefix: "",
    suffix: "",
    formatter: null,
    showSparklines: !0,
    fontFamily: "inherit",
    colNames: [],
    sortedDataColumnIndices: [],
    sortedSeriesName: !1,
    sortedSum: !1,
    sortedAverage: !1,
    sortedMedian: !1,
    resetSortOnClickOutside: !1,
    sparkline: {
      useGradient: !0,
      showArea: !0,
      strokeWidth: 3,
      type: v.LINE,
      smooth: !0,
      dimensions: {
        width: 150,
        heightRatio: 1
      },
      animation: {
        show: !0,
        animationFrames: 360
      }
    },
    translations: {
      serie: "Serie",
      total: "Total",
      average: "Average",
      median: "Median",
      chart: "Evolution"
    },
    title: {
      text: "",
      color: e,
      fontSize: t._20,
      bold: !0,
      textAlign: "center",
      backgroundColor: r,
      subtitle: {
        color: T,
        text: "",
        fontSize: t._16,
        bold: !1
      }
    },
    thead: {
      backgroundColor: r,
      color: e,
      fontSize: t._14,
      outline: "none",
      textAlign: m.LEFT,
      bold: !1
    },
    tbody: {
      showColorMarker: !0,
      backgroundColor: r,
      color: e,
      fontSize: t._14,
      outline: "none",
      textAlign: m.LEFT,
      bold: !1,
      selectedColor: {
        useSerieColor: !0,
        fallback: o
      }
    },
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      fullscreen: !0
    })
  }, Ve = {
    theme: "",
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      color: e,
      shapeSize: t._14,
      heatmapColors: {
        useIndividualScale: !1,
        min: l,
        max: c
      }
    },
    table: {
      responsiveBreakpoint: 400,
      borderWidth: 1,
      showSum: !1,
      showAverage: !1,
      showMedian: !1,
      head: {
        backgroundColor: r,
        color: e,
        values: []
      }
    },
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      fullscreen: !0
    })
  }, Ee = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    animationDelayMs: 20,
    strictPixelPadding: !0,
    // If true, strict per-pixel padding is used (dilateWordMask); if false, just rectangular bounding box (or pad).
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0,
      zoom: !0
    }),
    nodeCategories: {},
    nodeCategoryColors: {},
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 512,
        width: 512,
        zoom: {
          show: !0
        },
        controls: {
          position: "bottom",
          show: !0,
          backgroundColor: o,
          buttonColor: o,
          color: e,
          fontSize: 14,
          border: `1px solid ${k}`,
          padding: "0.5rem",
          borderRadius: "0.25rem"
        },
        words: {
          maxFontSize: 100,
          minFontSize: t._10,
          bold: !1,
          proximity: 10,
          packingWeight: 1,
          color: e,
          usePalette: !0,
          hoverOpacity: 0.5,
          selectedStroke: "transparent"
        },
        title: n,
        tooltip: {
          ...g,
          roundingValue: 0
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Word",
        value: "Value"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        prefix: "",
        suffix: ""
      }
    }
  }, Ne = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    loading: !1,
    responsive: !1,
    theme: "",
    customPalette: [],
    a11y: a,
    downsample: {
      threshold: 1e4
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      stack: !0,
      annotator: !0
    }),
    style: {
      fontFamily: "Arial",
      // A defined font must be provided as 'inherit' fails with canvas
      chart: {
        backgroundColor: r,
        color: e,
        aspectRatio: "12 / 9",
        stacked: !1,
        stackGap: 20,
        scale: {
          ticks: 10,
          min: null,
          max: null
        },
        zoom: {
          ...P,
          minimap: M,
          preview: {
            // v3
            enable: !1,
            fill: "#CCCCCC50",
            stroke: "#6A6A6A",
            strokeWidth: 2,
            strokeDasharray: 0
          },
          useDefaultFormat: !0,
          timeFormat: "yyyy-MM-dd HH:mm:ss",
          // When datetimeFormatter is enabled
          customFormat: null
          // overrides all if callback => string
        },
        selector: {
          show: !0,
          color: s,
          dashed: !0,
          showHorizontalSelector: !1
        },
        tooltip: {
          ...g,
          showTimeLabel: !0,
          useDefaultTimeFormat: !0,
          timeFormat: "yyyy-MM-dd HH:mm:ss"
          // When datetimeFormatter is enabled and useDefaultFormat is false
        },
        legend: {
          ...b,
          position: "bottom"
        },
        title: n,
        grid: {
          y: {
            showAxis: !0,
            axisColor: s,
            axisThickness: 2,
            axisName: "",
            axisLabels: {
              show: !0,
              fontSizeRatio: 1,
              color: e,
              offsetX: 0,
              rounding: 1,
              prefix: "",
              suffix: "",
              bold: !1
            },
            verticalLines: {
              show: !0,
              color: h,
              hideUnderXLength: 20,
              position: "middle"
              // 'start'
            }
          },
          x: {
            showAxis: !0,
            axisColor: s,
            axisThickness: 2,
            axisName: "",
            horizontalLines: {
              show: !0,
              color: h,
              alternate: !0,
              opacity: 20
            },
            timeLabels: {
              show: !0,
              showMarker: !0,
              fontSizeRatio: 0.8,
              values: [],
              datetimeFormatter: C,
              rotation: 0,
              offsetY: 30,
              color: e,
              modulo: 12,
              bold: !1
            }
          },
          zeroLine: {
            show: !0,
            color: s,
            dashed: !0
          }
        },
        line: {
          cutNullValues: !0,
          plots: {
            show: !0,
            radiusRatio: 1
          }
        },
        bar: {
          gradient: {
            show: !0
          }
        },
        area: {
          opacity: 60
        },
        dataLabels: {
          show: !0,
          fontSizeRatio: 1,
          useSerieColor: !0,
          color: e,
          offsetY: -12,
          formatter: null,
          bold: !0
        },
        paddingProportions: {
          top: 0.1,
          right: 0.05,
          bottom: 0.1,
          left: 0.1
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      rounding: 1,
      columnNames: {
        period: "Period",
        total: "Total"
      },
      th: u,
      td: d
    }
  }, Me = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    userOptions: i({
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      fullscreen: !0,
      annotator: !0,
      tooltip: !0,
      svg: !0
    }),
    nodeCategories: {},
    nodeCategoryColors: {},
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        width: 1e3,
        // v3
        height: 600,
        // v3
        color: e,
        padding: p([12, 12, 12, 12]),
        // v3 update
        title: n,
        tooltip: {
          ...g,
          showPercentage: !0,
          roundingPercentage: 0,
          translations: {
            from: "From:",
            to: "To:",
            percentOfTotal: "Percent of total:"
          }
        },
        legend: {
          ...b,
          position: "bottom"
        },
        nodes: {
          gap: 10,
          // minHeight: 20, // v3 deprecated
          width: 40,
          labels: {
            show: !0,
            // v3
            showValue: !0,
            fontSize: t._14,
            abbreviation: {
              use: !0,
              length: 3
            },
            prefix: "",
            suffix: "",
            rounding: 0,
            formatter: null
          },
          stroke: l,
          strokeWidth: 1,
          borderRadius: 0
        },
        links: {
          // width: 200, // v3 deprecated
          opacity: 0.8,
          stroke: l,
          strokeWidth: 1,
          smooth: !0
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        source: "Source",
        target: "Target",
        value: "Value"
      },
      th: u,
      td: d
    }
  }, Ye = {
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    responsiveProportionalSizing: !0,
    theme: "",
    useCssAnimation: !1,
    // v3
    customPalette: [],
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 600,
        width: 1e3,
        padding: p([0, 0, 0, 0]),
        comments: {
          show: !0,
          showInTooltip: !0,
          width: 200,
          offsetX: 0,
          offsetY: 0
        },
        lines: {
          smooth: !0,
          strokeWidth: 2,
          opacity: 0.8
        },
        plots: {
          show: !0,
          radius: 6,
          opacity: 0.8
        },
        yAxis: {
          scaleTicks: 10,
          stroke: s,
          strokeWidth: 1,
          labels: {
            showAxisNames: !0,
            axisNames: [],
            axisNamesRotation: 0,
            // v3
            axisNamesAutoRotate: {
              // v3
              enable: !0,
              // v3
              angle: -30
              // v3
            },
            axisNamesColor: e,
            axisNamesFontSize: t._16,
            axisNamesBold: !0,
            roundings: [],
            prefixes: [],
            suffixes: [],
            formatters: [],
            ticks: {
              show: !0,
              fontSize: t._14,
              color: e,
              bold: !1,
              offsetX: 0,
              offsetY: 0
            },
            datapoints: {
              show: !0,
              fontSize: t._14,
              useSerieColor: !0,
              color: e,
              offsetX: 0,
              offsetY: 0,
              bold: !0
            }
          }
        },
        title: n,
        legend: {
          ...b,
          position: "bottom"
        },
        tooltip: g
      }
    },
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        item: "Item"
      },
      th: u,
      td: d
    }
  }, Be = {
    type: "stopwatch",
    responsive: !1,
    responsiveProportionalSizing: !0,
    useCursorPointer: !1,
    style: {
      fontFamily: "inherit",
      backgroundColor: r,
      height: 300,
      width: 300,
      title: n
    },
    stopwatch: {
      showHours: !1,
      showHundredth: !0,
      cycleSeconds: 5,
      track: {
        radiusRatio: 1,
        stroke: o,
        fill: l,
        strokeWidth: 2
      },
      tracker: {
        radiusRatio: 1,
        stroke: l,
        strokeWidth: 1,
        fill: s,
        gradient: {
          show: !0,
          color: l
        },
        aura: {
          show: !0,
          radiusRatio: 1,
          fill: s,
          stroke: l,
          strokeWidth: 0
        }
      },
      cycleTrack: {
        show: !0,
        stroke: s,
        strokeWidth: 3
      },
      label: {
        fontSize: t._24,
        color: e,
        bold: !1
      },
      legend: {
        backgroundColor: r,
        buttons: {
          start: !0,
          pause: !0,
          reset: !0,
          restart: !0,
          lap: !0,
          iconColor: s
        },
        buttonTitles: {
          start: "Start",
          pause: "Pause",
          resume: "Resume",
          reset: "Reset",
          restart: "Restart",
          lap: "Lap"
        }
      }
    }
  }, He = {
    bubbleEffect: !0,
    bubbleEffectColor: l,
    bubbleEffectOpacity: 0.1,
    centerCircleColor: l,
    centerCircleDasharray: 0,
    centerCircleOpacity: 0,
    centerCircleRadius: 50,
    centerCircleStroke: h,
    centerCircleStrokeWidth: 0.5,
    coordinatesColor: h,
    coordinatesFontSize: t._10,
    coordinatesOffset: 0,
    crosshairDasharray: 0,
    crosshairStroke: h,
    crosshairStrokeWidth: 0.5,
    intersectCirclesFill: h,
    intersectCirclesRadius: 2,
    isLoading: !1,
    parentId: "",
    showCenterCircle: !0,
    showCoordinates: !0,
    showCrosshair: !0,
    showIntersectCircles: !0,
    useWaveOnClick: !0
  }, Ue = {
    open: !1,
    maxHeight: 2e3,
    useCursorPointer: !1,
    head: {
      useArrowSlot: !1,
      backgroundColor: r,
      color: e,
      iconColor: c,
      iconSize: 20,
      icon: "arrowRight",
      padding: "12px 6px"
    },
    body: {
      backgroundColor: r,
      color: e
    }
  }, Ie = {
    debug: !1,
    // v3
    animationFrames: 60,
    animationValueStart: 0,
    backgroundColor: r,
    fontFamily: "inherit",
    layoutClass: "",
    layoutCss: "",
    prefix: "",
    suffix: "",
    title: "",
    titleBold: !0,
    titleColor: e,
    titleClass: "",
    titleCss: "",
    titleFontSize: t._16,
    useAnimation: !0,
    valueBold: !0,
    valueColor: c,
    valueClass: "",
    valueCss: "",
    valueFontSize: t._32,
    valueRounding: 0,
    formatter: null,
    analogDigits: {
      show: !1,
      height: 40,
      color: e,
      skeletonColor: o
    }
  }, Xe = {
    type: "onion",
    onion: {
      gutterColor: h,
      gutterOpacity: 0.3,
      gutterBlur: 0,
      trackHueRotate: 20,
      trackBlur: 1,
      trackColor: N
    },
    line: {
      gutterColor: h,
      gutterOpacity: 0.3,
      gutterBlur: 0,
      trackHueRotate: 20,
      trackBlur: 1,
      trackColor: N
    },
    bar: {
      gutterColor: h,
      gutterOpacity: 0.3,
      gutterBlur: 0,
      trackHueRotate: 20,
      trackBlur: 1,
      trackColor: N
    }
  }, Ge = {
    readonly: !1,
    useCursorPointer: !1,
    style: {
      fontFamily: "inherit",
      itemSize: 32,
      backgroundColor: r,
      colors: {
        activeReadonly: [
          "#e20001",
          "#ff9f03",
          "#ffd004",
          "#61c900",
          "#059f00"
        ],
        active: [
          "#e20001",
          "#ff9f03",
          "#ffd004",
          "#61c900",
          "#059f00"
        ],
        inactive: [
          o,
          o,
          o,
          o,
          o
        ]
      },
      icons: {
        filled: !1,
        useGradient: !0
      },
      title: {
        ...n,
        textAlign: m.CENTER,
        offsetY: 6,
        subtitle: {
          ...n.subtitle,
          offsetY: 12
        }
      },
      rating: {
        show: !0,
        fontSize: t._24,
        bold: !0,
        roundingValue: 1,
        position: m.BOTTOM,
        offsetX: 0,
        offsetY: 0,
        formatter: null
      },
      tooltip: {
        show: !0,
        fontSize: t._14,
        offsetY: 0,
        color: e,
        bold: !0,
        backgroundColor: r,
        borderColor: k,
        borderRadius: 4,
        boxShadow: "0 6px 12px -6px rgba(0,0,0,0.2)",
        roundingValue: 0,
        formatter: null
      }
    }
  }, $e = {
    type: v.STAR,
    readonly: !1,
    from: 1,
    to: 5,
    useCursorPointer: !1,
    style: {
      fontFamily: "inherit",
      animated: !0,
      itemSize: 32,
      backgroundColor: r,
      star: {
        activeColor: Y,
        borderColor: Y,
        borderWidth: 3,
        apexes: 5,
        inactiveColor: o,
        useGradient: !0
      },
      image: {
        src: "",
        inactiveOpacity: 0.3,
        alt: "rating image"
      },
      title: {
        textAlign: m.CENTER,
        fontSize: t._20,
        color: e,
        bold: !0,
        text: "",
        offsetY: 6,
        subtitle: {
          fontSize: t._14,
          color: h,
          bold: !1,
          text: "",
          offsetY: 12
        }
      },
      rating: {
        show: !0,
        fontSize: t._24,
        bold: !0,
        roundingValue: 1,
        position: m.BOTTOM,
        offsetY: 0,
        offsetX: 0,
        formatter: null
      },
      tooltip: {
        show: !0,
        fontSize: t._14,
        offsetY: 0,
        color: e,
        bold: !0,
        backgroundColor: r,
        borderColor: k,
        borderRadius: 4,
        boxShadow: "0 6px 12px -6px rgba(0,0,0,0.2)",
        roundingValue: 0,
        formatter: null
      }
    }
  }, je = {
    alwaysVisible: !1,
    useCursorPointer: !1,
    style: {
      backgroundColor: r,
      color: e,
      fontFamily: "inherit",
      showPrint: !1,
      showSave: !0,
      showImage: !0,
      showTooltips: !0,
      buttons: {
        borderRadius: 6,
        controls: {
          backgroundColor: r,
          color: e,
          border: "1px solid #262626",
          selected: {
            backgroundColor: s,
            color: E,
            border: `1px solid ${k}`
          }
        },
        shapes: {
          backgroundColor: r,
          color: e,
          border: "1px solid #262626",
          selected: {
            backgroundColor: e,
            color: r,
            border: `1px solid ${k}`
          }
        }
      },
      tooltips: {
        backgroundColor: E,
        color: e,
        border: `1px solid ${k}`,
        borderRadius: 6,
        boxShadow: "0 6px 12px -6px rgba(0,0,0,0.2)"
      }
    },
    translations: {
      colorAlpha: "Color alpha",
      dashedLines: "Dashed lines",
      filled: "Filled",
      fontSize: "Font size",
      thickness: "Thickness",
      title: "Annotations",
      tooltipGroup: "Select & group",
      tooltipDelete: "Delete",
      tooltipMove: "Move",
      tooltipResize: "Resize",
      tooltipBringToFront: "Bring to front",
      tooltipBringToBack: "Bring to back",
      tooltipDuplicate: "Duplicate",
      tooltipUndo: "Undo last shape",
      tooltipRedo: "Redo last shape",
      tooltipPdf: "Save pdf",
      tooltipSave: "Save annotations",
      tooltipShapeCircle: "Draw circle",
      tooltipShapeRect: "Draw rect",
      tooltipShapeArrow: "Draw arrow",
      tooltipShapeFreehand: "Freehand line",
      tooltipShapeText: "Text mode",
      tooltipShapeTextLeft: "Align left",
      tooltipShapeTextCenter: "Align center",
      tooltipShapeTextRight: "Align right",
      tooltipShapeTextBullet: "Bullet points",
      tooltipShapeTextBold: "Bold",
      tooltipShapeTextItalic: "Italic",
      tooltipShapeTextUnderline: "Underlined",
      tooltipShapeColor: "Color",
      tooltipImage: "Download PNG"
    }
  }, qe = {
    locked: !1,
    style: {
      board: {
        backgroundColor: r,
        color: e,
        aspectRatio: "1/1.4141",
        border: "none"
      },
      item: {
        backgroundColor: r,
        borderColor: k
      },
      resizeHandles: {
        backgroundColor: e
      }
    },
    userOptions: {
      show: !0,
      useCursorPointer: !1,
      showOnChartHover: !1,
      keepStateOnChartLeave: !0,
      position: "right",
      buttons: {
        pdf: !0,
        img: !0,
        annotator: !0,
        altCopy: !1
      },
      callbacks: {
        pdf: null,
        img: null,
        annotator: null,
        altCopy: null
      },
      buttonTitles: {
        pdf: "Download PDF",
        img: "Download PNG",
        annotator: "Toggle annotator",
        altCopy: "Copy alt text"
      },
      print: {
        scale: 2,
        filename: ""
      }
    }
  }, Ze = {
    type: v.LINE,
    style: {
      backgroundColor: r,
      color: e,
      maxHeight: 500,
      animated: !0,
      chord: {
        color: o
      },
      ridgeline: {
        color: o
      },
      circlePack: {
        color: o
      },
      historyPlot: {
        color: o
      },
      bullet: {
        color: o
      },
      flow: {
        color: o
      },
      parallelCoordinatePlot: {
        color: o
      },
      treemap: {
        color: o
      },
      dumbbell: {
        color: o
      },
      stripPlot: {
        color: o
      },
      galaxy: {
        color: o
      },
      bar3d: {
        color: o
      },
      sparkHistogram: {
        color: o
      },
      sparkStackbar: {
        color: o
      },
      sparkbar: {
        color: o
      },
      thermometer: {
        color: o
      },
      relationCircle: {
        color: o
      },
      molecule: {
        color: o
      },
      tiremarks: {
        color: o
      },
      chestnut: {
        color: o
      },
      sparkline: {
        color: o,
        strokeWidth: 0.7
      },
      donutEvolution: {
        axis: {
          show: !0,
          color: o,
          strokeWidth: 0.5
        },
        donuts: {
          strokeWidth: 0.5,
          color: o
        }
      },
      line: {
        axis: {
          show: !0,
          color: o,
          strokeWidth: 0.5
        },
        path: {
          color: o,
          strokeWidth: 1,
          showPlots: !0
        }
      },
      bar: {
        axis: {
          show: !0,
          color: o,
          strokeWidth: 0.5
        },
        borderRadius: 0.5,
        color: o,
        barWidth: 9
      },
      donut: {
        color: o,
        strokeWidth: 64
      },
      onion: {
        color: o
      },
      gauge: {
        color: o
      },
      quadrant: {
        grid: {
          color: o,
          strokeWidth: 0.5
        },
        plots: {
          radius: 1.5,
          color: o
        }
      },
      radar: {
        grid: {
          color: o,
          strokeWidth: 0.5
        },
        shapes: {
          color: o
        }
      },
      waffle: {
        color: o
      },
      table: {
        th: {
          color: o
        },
        td: {
          color: o,
          strokeWidth: 0.5
        }
      },
      rating: {
        useSmiley: !1,
        color: o,
        filled: !0,
        strokeWidth: 1,
        maxWidth: 200
      },
      verticalBar: {
        axis: {
          show: !0,
          color: o,
          strokeWidth: 0.5
        },
        borderRadius: 0.5,
        color: o
      },
      heatmap: {
        cellsX: 26,
        cellsY: 7,
        color: o
      },
      candlesticks: {
        axis: {
          show: !0,
          color: o,
          strokeWidth: 0.5
        },
        candle: {
          color: o,
          strokeWidth: 1
        }
      },
      pyramid: {
        color: o
      },
      wheel: {
        color: o
      },
      rings: {
        color: o
      }
    }
  }, Ke = {
    fontFamily: "inherit",
    maxHeight: 500,
    rowsPerPage: 25,
    useCursorPointer: !1,
    style: {
      title: {
        ...n,
        backgroundColor: r
      },
      th: {
        backgroundColor: o,
        color: e,
        outline: `1px solid ${l}`,
        selected: {
          backgroundColor: c,
          color: l
        },
        buttons: {
          filter: {
            inactive: {
              backgroundColor: o,
              color: e
            },
            active: {
              backgroundColor: c,
              color: l
            }
          },
          cancel: {
            inactive: {
              backgroundColor: o,
              color: s
            },
            active: {
              backgroundColor: "#F17171",
              color: l
            }
          }
        }
      },
      rows: {
        outline: `1px solid ${r}`,
        even: {
          backgroundColor: "#f3f5f7",
          color: e,
          selectedCell: {
            backgroundColor: "#1f77b45b",
            color: e
          },
          selectedNeighbors: {
            backgroundColor: "#63dd821e",
            color: e
          }
        },
        odd: {
          backgroundColor: r,
          color: e,
          selectedCell: {
            backgroundColor: "#1f77b45b",
            color: e
          },
          selectedNeighbors: {
            backgroundColor: "#63dd821e",
            color: e
          }
        }
      },
      inputs: {
        backgroundColor: r,
        color: e,
        border: `1px solid ${k}`,
        accentColor: c
      },
      dropdowns: {
        backgroundColor: o,
        color: e,
        icons: {
          selected: {
            color: O,
            unicode: "✔"
          },
          unselected: {
            color: z,
            unicode: "✖"
          }
        }
      },
      infoBar: {
        backgroundColor: o,
        color: e
      },
      pagination: {
        buttons: {
          backgroundColor: o,
          color: e,
          opacityDisabled: 0.5
        },
        navigationIndicator: {
          backgroundColor: c
        }
      },
      exportMenu: {
        show: !0,
        backgroundColor: o,
        color: e,
        buttons: {
          backgroundColor: E,
          color: e
        },
        filename: ""
      },
      closeButtons: {
        backgroundColor: "transparent",
        color: e,
        borderRadius: "50%"
      },
      chart: {
        modal: {
          backgroundColor: o,
          color: e,
          buttons: {
            selected: {
              backgroundColor: c,
              color: l
            },
            unselected: {
              backgroundColor: r,
              color: e
            }
          }
        },
        layout: {
          backgroundColor: r,
          axis: {
            stroke: "#ccd1d4",
            strokeWidth: 2
          },
          bar: {
            fill: c,
            stroke: l
          },
          line: {
            smooth: !0,
            useArea: !1,
            stroke: c,
            strokeWidth: 4,
            plot: {
              fill: c,
              stroke: l,
              // DEPRECATED
              strokeWidth: 1,
              radius: {
                selected: 6,
                unselected: 4
              }
            },
            selector: {
              // DEPRECATED
              stroke: "#ccc",
              // DEPRECATED
              strokeWidth: 1,
              // DEPRECATED
              strokeDasharray: 5
              // DEPRECATED
            }
          },
          labels: {
            color: e
          },
          progression: {
            stroke: s,
            strokeWidth: 2,
            strokeDasharray: 4,
            arrowSize: 7
          },
          timeLabels: {
            showOnlyAtModulo: !0,
            modulo: 12
          },
          datetimeFormatter: {
            ...C,
            enable: !0
          },
          zoom: {
            show: !0
          }
        }
      }
    },
    translations: {
      average: "Average",
      by: "by",
      chooseCategoryColumn: "Choose category column",
      exportAllButton: "CSV all",
      exportAllLabel: "Export all rows of your current filtered dataset",
      exportPageButton: "CSV page",
      exportPageLabel: "Export rows of the current page",
      from: "From",
      inputPlaceholder: "Search...",
      makeDonut: "Generate",
      nb: "Nb",
      page: "Page",
      paginatorLabel: "Rows per page",
      sizeWarning: "Displaying too many rows at a time can lead to slower performance",
      sum: "Sum",
      to: "To",
      total: "Total",
      totalRows: "Total rows",
      filename: "File name",
      xAxisLabels: "X axis labels"
    },
    useChart: !0
  }, Qe = {
    height: "100%",
    width: null,
    backgroundColor: r,
    digits: {
      color: e,
      skeletonColor: o,
      thickness: 1
    }
  }, Je = {
    responsiveBreakpoint: 400,
    userOptions: i({
      tooltip: !1,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !1,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      animation: !0
    }),
    animation: {
      type: "scroll",
      // 'scroll' | 'marquee'
      use: !0,
      speedMs: 1e3,
      pauseOnHover: !0
    },
    style: {
      backgroundColor: r,
      color: e,
      fontFamily: "inherit"
    },
    border: {
      size: 0,
      color: s
    },
    caption: {
      text: "",
      padding: p([12, 12, 12, 12]),
      style: {
        backgroundColor: r,
        color: e,
        fontSize: `${t._16}px`,
        fontWeight: "bold",
        textAlign: "left"
      }
    },
    scrollbar: {
      showOnlyOnHover: !1,
      hide: !1
    },
    thead: {
      style: {
        verticalAlign: "middle"
      },
      tr: {
        height: 32,
        border: {
          size: 0,
          color: s
        },
        style: {
          backgroundColor: r,
          color: e,
          boxShadow: `0px 6px 12px -6px ${s}50`
        },
        th: {
          border: {
            size: 0,
            color: s
          },
          padding: p([0, 12, 0, 0]),
          style: {
            borderSpacing: 0,
            border: "none",
            textAlign: "right",
            fontVariantNumeric: "tabular-nums"
          }
        }
      }
    },
    tbody: {
      backgroundColor: r,
      tr: {
        visible: 5,
        height: 32,
        border: {
          size: 0,
          color: s
        },
        style: {
          backgroundColor: r,
          color: e
        },
        td: {
          alternateColor: !0,
          alternateOpacity: 0.5,
          border: {
            size: 0,
            color: s
          },
          padding: p([0, 12, 0, 0]),
          style: {
            fontVariantNumeric: "tabular-nums",
            textAlign: "right",
            backgroundColor: o
          }
        }
      }
    }
  }, et = {
    a11y: {
      translations: {
        label: "Progress"
      }
    },
    skeletonConfig: null,
    skeletonDataset: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    type: "battery",
    // battery | gauge
    size: 64,
    stroke: h,
    color: c,
    useGradient: !0,
    gradientColor: "#9db5ed",
    showPercentage: !0,
    textColor: s,
    fontFamily: "inherit",
    formatter: null
  }, tt = {
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    // v3
    theme: "",
    userOptions: i({
      tooltip: !1,
      pdf: !0,
      csv: !1,
      img: !0,
      table: !1,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 96,
        width: 600,
        padding: p([24, 24, 24, 12]),
        animation: {
          show: !0,
          animationFrames: 60
        },
        segments: {
          baseColor: "#9A9A9A",
          // default if dataset.segments colors are not provided
          dataLabels: {
            show: !0,
            color: e,
            fontSize: t._10,
            formatter: null,
            bold: !1,
            prefix: "",
            suffix: "",
            rounding: 0,
            offsetY: 0
          },
          ticks: {
            show: !0,
            divisions: 10,
            stroke: "#8A8A8A"
          }
        },
        target: {
          show: !0,
          // v3
          onTop: !0,
          color: e,
          rounded: !0,
          heightRatio: 0.8,
          stroke: l,
          strokeWidth: 1,
          width: 6
        },
        valueBar: {
          color: "#3A3A3A",
          heightRatio: 0.33,
          stroke: l,
          strokeWidth: 1,
          label: {
            show: !0,
            color: e,
            fontSize: t._14,
            bold: !0,
            offsetY: 0
          }
        },
        title: {
          ...n,
          textAlign: m.LEFT,
          paddingLeft: 12
        },
        legend: {
          ...b,
          roundingValue: 0,
          position: "bottom"
        }
      }
    }
  }, ot = {
    theme: "",
    responsive: !1,
    responsiveProportionalSizing: !0,
    useCssAnimation: !0,
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0,
        roundingPercentage: 0
      },
      columnNames: {
        series: "Step",
        value: "Value",
        percentage: "Percentage"
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      annotator: !0,
      svg: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        width: 600,
        height: 500,
        title: n,
        padding: p([12, 128, 24, 24]),
        barCircleSpacingRatio: 0.05,
        circles: {
          stroke: l,
          strokeWidth: 1,
          dataLabels: {
            formatter: null,
            fontSize: t._16,
            offsetY: 0,
            adaptColorToBackground: !0,
            color: e,
            rounding: 1,
            bold: !0
          }
        },
        circleLinks: {
          show: !0,
          color: o,
          widthRatio: 1
        },
        area: {
          show: !0,
          color: o
        },
        bars: {
          stroke: l,
          defaultColor: c,
          strokeWidth: 1,
          gapRatio: 0.2,
          borderRadius: 3,
          dataLabels: {
            name: {
              fontSize: t._16,
              color: e,
              bold: !0,
              offsetX: 0,
              offsetY: 0
            },
            value: {
              formatter: null,
              fontSize: t._16,
              rounding: 0,
              bold: !1,
              color: e,
              prefix: "",
              suffix: "",
              offsetX: 0,
              offsetY: 0
            }
          }
        }
      }
    }
  }, rt = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    responsiveProportionalSizing: !0,
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0
      },
      columnNames: {
        series: "Series",
        datapoint: "Datapoint",
        x: "x",
        y: "y"
      }
    },
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 500,
        width: 600,
        padding: p([12, 12, 12, 12]),
        grid: {
          xAxis: {
            show: !0,
            stroke: o,
            strokeWidth: 1
          },
          horizontalLines: {
            show: !0,
            stroke: o,
            strokeWidth: 0.6
          },
          yAxis: {
            show: !0,
            stroke: o,
            strokeWidth: 1
          },
          verticalLines: {
            show: !0,
            stroke: o,
            strokeWidth: 0.6
          }
        },
        axes: {
          x: {
            scaleMin: null,
            scaleMax: null,
            ticks: 10,
            labels: {
              show: !0,
              fontSize: t._16,
              color: e,
              bold: !1,
              rounding: 1,
              offsetY: 0,
              rotation: 0,
              autoRotate: {
                // v3
                enable: !0,
                // v3
                angle: -30
                // v3
              },
              formatter: null,
              prefix: "",
              suffix: ""
            },
            name: {
              text: "",
              fontSize: t._16,
              offsetX: 0,
              offsetY: 0,
              bold: !1,
              color: e
            }
          },
          y: {
            scaleMin: null,
            scaleMax: null,
            ticks: 10,
            labels: {
              show: !0,
              fontSize: t._16,
              color: e,
              bold: !1,
              rounding: 1,
              offsetX: 0,
              formatter: null,
              prefix: "",
              suffix: ""
            },
            name: {
              text: "",
              fontSize: t._16,
              offsetX: 0,
              offsetY: 0,
              bold: !1,
              color: e
            }
          }
        },
        plots: {
          radius: 16,
          stroke: l,
          strokeWidth: 1,
          gradient: {
            show: !0,
            intensity: 40
          },
          indexLabels: {
            show: !0,
            startAtZero: !1,
            adaptColorToBackground: !0,
            color: e,
            fontSize: t._16,
            bold: !1,
            offsetY: 0,
            offsetX: 0
          },
          labels: {
            show: !0,
            fontSize: t._10,
            color: e,
            bold: !1,
            offsetY: 0,
            offsetX: 0
          }
        },
        paths: {
          show: !0,
          strokeWidth: 1.6,
          useSerieColor: !0,
          stroke: s
        },
        legend: {
          ...b,
          position: "bottom"
        },
        title: n,
        tooltip: g
      }
    }
  }, at = {
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: d,
      columnNames: {
        datapoint: "Datapoint",
        value: "Value"
      }
    },
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        title: n,
        width: 512,
        height: 316,
        circles: {
          stroke: l,
          strokeWidth: 1,
          selectedShadowColor: s,
          gradient: {
            show: !0,
            intensity: 40
          },
          labels: {
            name: {
              fontSizeRatio: 1,
              show: !0,
              bold: !1,
              offsetY: 0,
              color: "auto"
            },
            value: {
              fontSizeRatio: 1,
              show: !0,
              color: "auto",
              rounding: 0,
              prefix: "",
              suffix: "",
              formatter: null,
              bold: !1,
              offsetY: 0
            }
          }
        },
        tooltip: g
      }
    }
  }, lt = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      annotator: !0,
      svg: !0
    }),
    customPalette: [],
    projection: "globe",
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        padding: p([12, 12, 12, 12]),
        dataLabels: {
          prefix: "",
          suffix: "",
          rounding: 0,
          formatter: null
        },
        dimensions: {
          height: null,
          width: null
        },
        // Specific config for the globe projection
        globe: {
          // Center on China: x:98, y:34
          // Center on USA: x:-101, y:42
          // Center on Australia: x:133, y:-26.5
          // Center on Europe: x:13, y:44
          // etc
          center: {
            x: 0,
            y: 0
          },
          waterColor: c
        },
        territory: {
          stroke: "#666666",
          emptyColor: "#E0E0E0",
          strokeWidth: 0.5,
          strokeWidthSelected: 2,
          colors: {
            min: "#E0E0E0",
            max: null
            // defaults to palette[0] if kept null
          },
          showTaiwanAsPartOfChina: !1
        },
        tooltip: {
          ...g,
          showMinimap: !0
        },
        title: n,
        legend: {
          ...b,
          position: "bottom"
        }
      }
    },
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0
      },
      columnNames: {
        series: "Country",
        value: "Value",
        category: "Category"
      }
    }
  }, st = {
    skeletonDataset: null,
    skeletonConfig: null,
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      // v3
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    // v2 = true
    userOptions: i({
      tooltip: !1,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: {
        ...d,
        roundingValue: 0
      },
      columnNames: {
        series: "Series"
      }
    },
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        padding: p([12, 0, 0, 12]),
        title: n,
        legend: b,
        dialog: {
          show: !0,
          backgroundColor: r,
          color: e,
          header: {
            backgroundColor: o,
            color: e
          },
          xyChart: {
            ...S,
            responsive: !0,
            line: {
              ...S.line,
              showTransition: !1,
              labels: {
                ...S.line.labels,
                show: !0
              }
            },
            chart: {
              ...S.chart,
              tooltip: {
                ...S.chart.tooltip,
                showPercentage: !1
              },
              userOptions: {
                ...S.chart.userOptions,
                buttons: {
                  ...S.chart.userOptions.buttons,
                  pdf: !1,
                  fullscreen: !1
                }
              },
              zoom: {
                ...S.chart.zoom,
                minimap: {
                  ...S.chart.zoom.minimap,
                  show: !0
                }
              }
            }
          }
        },
        areas: {
          height: 60,
          // height > rowHeight results in the classic ridgeline look
          rowHeight: 30,
          strokeWidth: 1,
          stroke: {
            useSerieColor: !1,
            color: l
          },
          smooth: !0,
          // Also applied in dialog chart
          opacity: 0.9,
          useCommonColor: !1,
          useGradient: !0,
          maxPoint: {
            show: !0,
            adaptStrokeToBackground: !0,
            stroke: s,
            strokeWidth: 1,
            strokeDasharray: 4
          }
        },
        selector: {
          show: !0,
          stroke: h,
          strokeWidth: 1,
          strokeDasharray: 2,
          labels: {
            fontSize: 12,
            formatter: null,
            rounding: 0,
            color: e
          },
          dot: {
            radius: 4,
            useDatapointColor: !0,
            fill: h,
            stroke: l,
            strokeWidth: 0.5
          }
        },
        zeroLine: {
          show: !0,
          strokeWidth: 1,
          strokeDasharray: 0,
          useSerieColor: !1,
          stroke: h
        },
        xAxis: {
          labels: {
            prefix: "",
            suffix: "",
            rotation: 0,
            autoRotate: {
              // v3
              enable: !0,
              // v3
              angle: -30
              // v3
            },
            values: [],
            datetimeFormatter: C,
            color: e,
            fontSize: 14,
            bold: !1,
            showOnlyAtModulo: !1,
            modulo: 12,
            showOnlyFirstAndLast: !1,
            offsetY: 0
          }
        },
        yAxis: {
          labels: {
            fontSize: 16,
            bold: !1,
            color: e,
            offsetX: 0
          }
        }
      }
    }
  }, nt = {
    debug: !1,
    // v3
    loading: !1,
    // v3
    responsive: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      datapointEnter: null,
      // v3
      datapointLeave: null,
      // v3
      datapointClick: null
      // v3
    },
    theme: "",
    customPalette: [],
    enableRotation: !0,
    initialRotation: 0,
    useCssAnimation: !1,
    // v2 = true
    userOptions: i({
      tooltip: !1,
      pdf: !0,
      csv: !0,
      img: !0,
      table: !0,
      labels: !1,
      fullscreen: !0,
      stack: !1,
      annotator: !0,
      svg: !0
    }),
    table: {
      ...f,
      useDialog: !1,
      th: u,
      td: d
    },
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        legend: {
          ...b,
          position: "bottom"
        },
        title: n,
        arcs: {
          innerRadiusRatio: 1,
          outerRadiusRatio: 1,
          padAngle: 5,
          stroke: l,
          strokeWidth: 1,
          labels: {
            show: !0,
            fontSize: 14,
            bold: !1,
            curved: !1,
            adaptColorToBackground: !0,
            // Applied if curved
            color: e,
            // If curved, applied if !adaptColorToBackground
            offset: 0,
            showPercentage: !0,
            roundingPercentage: 0
          }
        },
        ribbons: {
          stroke: l,
          strokeWidth: 1,
          underlayerOpacity: 1,
          labels: {
            show: !0,
            formatter: null,
            prefix: "",
            suffix: "",
            rounding: 0,
            fontSize: 14,
            bold: !1,
            useSerieColor: !1,
            color: e,
            offset: 0,
            minSeparationDeg: 3,
            connector: {
              stroke: s,
              strokeWidth: 1
            },
            marker: {
              show: !0,
              radius: 3,
              stroke: l,
              strokeWidth: 1
            }
          }
        }
      }
    }
  }, it = {
    skeletonConfig: null,
    skeletonDataset: null,
    loading: !1,
    debug: !1,
    responsive: !1,
    theme: "",
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    userOptions: i({
      tooltip: !1,
      pdf: !0,
      csv: !1,
      img: !0,
      fullscreen: !0,
      annotator: !0,
      svg: !0,
      zoom: !0
    }),
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        width: null,
        height: null,
        backgroundPattern: {
          show: !1,
          spacingRatio: 3,
          dotRadiusRatio: 5,
          dotColor: o,
          opacity: 1
        },
        layout: {
          rankDirection: "TB",
          rankSeparation: 60,
          nodeSeparation: 50,
          edgeSeparation: 30,
          nodeWidth: 100,
          nodeHeight: 40,
          curvedEdges: !0,
          padding: 48,
          arrowShape: "vee",
          arrowSize: 8,
          align: void 0
        },
        nodes: {
          stroke: h,
          strokeWidth: 1,
          borderRadius: 3,
          backgroundColor: r,
          labels: {
            color: e,
            fontSize: t._12,
            bold: !1
          },
          tooltip: {
            showOnClick: !1,
            backgroundColor: o,
            color: e,
            maxWidth: "300px"
          },
          selected: {
            stroke: null,
            strokeWidth: null,
            backgroundColor: null,
            labelColor: null,
            downstreamEdges: {
              stroke: null,
              animated: null
            },
            upstreamEdges: {
              stroke: null,
              animated: null
            }
          }
        },
        edges: {
          stroke: h,
          strokeWidth: 1,
          animations: {
            dasharray: "2 6",
            animationDurationMs: 1e3
          }
        },
        midpoints: {
          show: !1,
          radius: 4,
          stroke: h,
          fill: l,
          strokeWidth: 1,
          tooltip: {
            maxWidth: "300px",
            backgroundColor: o,
            color: e
          },
          selectedEdge: {
            stroke: null,
            animated: null
          }
        },
        controls: {
          position: "bottom",
          show: !0,
          backgroundColor: o,
          buttonColor: o,
          color: e,
          fontSize: 14,
          border: `1px solid ${k}`,
          padding: "0.5rem",
          borderRadius: "0.25rem"
        },
        zoom: {
          active: !0
        },
        title: n
      }
    }
  }, ut = {
    skeletonDataset: null,
    skeletonConfig: null,
    loading: !1,
    debug: !1,
    responsive: !1,
    projection: "equirectangular",
    theme: "",
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    userOptions: i({
      tooltip: !0,
      pdf: !0,
      csv: !1,
      img: !0,
      svg: !0,
      table: !1,
      labels: !1,
      fullscreen: !0,
      annotator: !0,
      zoom: !0
    }),
    map: {
      geoJson: null,
      center: [0, 0],
      fitPadding: 0
    },
    events: {
      datapointEnter: null,
      datapointLeave: null,
      datapointClick: null,
      territoryEnter: null,
      territoryLeave: null,
      territoryClick: null
    },
    style: {
      fontFamily: "inherit",
      chart: {
        dimensions: {
          width: null,
          height: null
        },
        backgroundColor: r,
        color: e,
        territory: {
          fill: "#F2F3F5",
          stroke: "#D0D4D8",
          strokeWidth: 1,
          hover: {
            enabledWhenEmpty: !1,
            fill: o,
            stroke: h,
            strokeWidth: 1.5
          }
        },
        points: {
          radius: 1,
          stroke: l,
          strokeWidth: 1,
          fill: V,
          hoverRadiusRatio: 1.2,
          labels: {
            show: !0,
            fontSizeRatio: 1,
            color: e,
            offsetY: 0
          }
        },
        controls: {
          position: "bottom",
          show: !0,
          backgroundColor: o,
          buttonColor: o,
          color: e,
          fontSize: 14,
          border: `1px solid ${k}`,
          padding: "0.5rem",
          borderRadius: "0.25rem"
        },
        title: n,
        tooltip: g,
        zoom: {
          active: !0
        }
      }
    }
  }, dt = {
    skeletonConfig: null,
    skeletonDataset: null,
    loading: !1,
    debug: !1,
    responsive: !1,
    theme: "",
    customPalette: [],
    useCssAnimation: !1,
    a11y: {
      ...a,
      translations: {
        ...a.translations,
        keyboardNavigation: "Use the left and right, or up and down arrow keys to move between datapoints"
      }
    },
    events: {
      datapointEnter: null,
      datapointLeave: null,
      datapointClick: null
    },
    userOptions: i({
      tooltip: !1,
      pdf: !0,
      csv: !0,
      img: !0,
      svg: !0,
      table: !0,
      labels: !1,
      fullscreen: !0,
      sort: !1,
      stack: !1,
      animation: !1,
      annotator: !0,
      zoom: !1
    }),
    table: {
      ...f,
      useDialog: !1,
      columnNames: {
        series: "Series",
        period: "Period",
        values: "Values",
        ranking: "Ranking"
      },
      th: u,
      td: {
        ...d,
        roundingValue: 0
      }
    },
    style: {
      fontFamily: "inherit",
      chart: {
        backgroundColor: r,
        color: e,
        height: 500,
        width: 800,
        layout: {
          timeLabels: {
            show: !0,
            values: [],
            datetimeFormatter: C,
            offsetY: 0,
            rotation: 0,
            autoRotate: {
              enable: !0,
              angle: -30
            },
            fontSize: t._14,
            color: e,
            bold: !1,
            showOnlyFirstAndLast: !1,
            showOnlyAtModulo: !1,
            modulo: 12
          },
          lines: {
            smooth: !0,
            strokeWidth: 4,
            coatingColor: l
          },
          plots: {
            stroke: l,
            strokeWidth: 1,
            radius: 12,
            labels: {
              color: "auto",
              show: !0,
              bold: !0,
              displayedValue: "value",
              // 'rank' | 'value'
              fontSize: t._12,
              prefix: "",
              suffix: "",
              rounding: 0,
              formatter: null
            }
          },
          nameLabels: {
            fontSize: t._14,
            color: e,
            useSerieColor: !1,
            bold: !1,
            offsetX: 0,
            leftLabels: {
              show: !0
            },
            rightLabels: {
              show: !0
            }
          }
        },
        padding: p([12, 12, 12, 12]),
        title: n
      }
    }
  };
  return {
    vue_ui_xy: S,
    vue_ui_donut: x,
    vue_ui_treemap: re,
    vue_ui_waffle: ae,
    vue_ui_radar: le,
    vue_ui_quadrant: se,
    vue_ui_gauge: ne,
    vue_ui_wheel: ie,
    vue_ui_tiremarks: ue,
    vue_ui_chestnut: de,
    vue_ui_onion: fe,
    vue_ui_vertical_bar: B,
    // deprecate in v4
    vue_ui_horizontal_bar: ce,
    // v3
    vue_ui_heatmap: he,
    vue_ui_scatter: ge,
    vue_ui_candlestick: pe,
    vue_ui_sparkline: be,
    vue_ui_sparkbar: me,
    vue_ui_sparkstackbar: ke,
    vue_ui_sparkhistogram: we,
    vue_ui_sparkgauge: ye,
    vue_ui_spark_trend: ve,
    vue_ui_quick_chart: Ce,
    vue_ui_age_pyramid: xe,
    vue_ui_relation_circle: Se,
    vue_ui_thermometer: _e,
    vue_ui_rings: ze,
    vue_ui_donut_evolution: Le,
    vue_ui_mood_radar: Pe,
    vue_ui_molecule: Ae,
    vue_ui_nested_donuts: Te,
    vue_ui_galaxy: De,
    vue_ui_strip_plot: Oe,
    vue_ui_dumbbell: Fe,
    vue_ui_3d_bar: Re,
    vue_ui_table_sparkline: We,
    vue_ui_table_heatmap: Ve,
    vue_ui_word_cloud: Ee,
    vue_ui_xy_canvas: Ne,
    vue_ui_flow: Me,
    vue_ui_parallel_coordinate_plot: Ye,
    vue_ui_timer: Be,
    vue_ui_carousel_table: Je,
    vue_ui_gizmo: et,
    vue_ui_stackbar: oe,
    vue_ui_bullet: tt,
    vue_ui_funnel: ot,
    vue_ui_history_plot: rt,
    vue_ui_world: lt,
    vue_ui_ridgeline: st,
    vue_ui_chord: nt,
    vue_ui_stackline: te,
    vue_ui_dag: it,
    vue_ui_geo: ut,
    vue_ui_bump: dt,
    // non chart components
    vue_ui_cursor: He,
    vue_ui_accordion: Ue,
    vue_ui_kpi: Ie,
    vue_ui_mini_loader: Xe,
    vue_ui_smiley: Ge,
    vue_ui_rating: $e,
    vue_ui_annotator: je,
    vue_ui_dashboard: qe,
    vue_ui_skeleton: Ze,
    vue_ui_table: Ke,
    vue_ui_digits: Qe,
    vue_ui_circle_pack: at
  };
}
export {
  ct as u
};
