// vue-data-ui/dist/vue-ui-sparkline-BYPPlWj6.js
import { ref as m, onMounted as xe, onBeforeUnmount as te, computed as s, watch as w, nextTick as M, createElementBlock as c, createCommentVNode as k, openBlock as i, Fragment as J, renderList as ee, createElementVNode as g, defineAsyncComponent as De, toRefs as pt, shallowRef as Be, watchEffect as yt, normalizeStyle as N, renderSlot as z, createBlock as me, normalizeProps as Re, guardReactiveProps as Ee, toDisplayString as Z, unref as n, createVNode as gt, withCtx as bt, createTextVNode as mt } from "vue";
import { c as kt, t as ze, ay as ke, o as Oe, g as Ve, l as Ct, h as Q, n as At, an as wt, X as xt, s as H, A as qe, C as Lt, Q as je, a as Ce, d as Ae, az as Mt } from "./lib-DYekjHyv.js";
import { t as $t, u as _t } from "./useResponsive-DfdjqQps.js";
import { u as Nt, a as we } from "./useNestedProp-CZTW4ubg.js";
import { u as It, B as St } from "./BaseScanner-BMpwQAfz.js";
import { u as Ft } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Tt } from "./useTimeLabels-C-A2uZKi.js";
import { u as Pt } from "./useChartAccessibility-9icAAmYg.js";
import { _ as Dt } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const We = { style: { line: { color: "#2A2929" }, area: { color: "#65698E" }, bar: { color: "#2A2929" } } }, Ge = { style: { backgroundColor: "#1A1A1A", zeroLine: { color: "#5A5A5A" }, line: { color: "#65698E" }, area: { color: "#8D99AE" }, bar: { color: "#65698E" }, plot: { stroke: "#1A1A1A" }, verticalIndicator: { color: "#8A8A8A" }, dataLabel: { color: "#CCCCCC" }, title: { color: "#CCCCCC" }, tooltip: { backgroundColor: "#1A1A1A", backgroundOpacity: 70, color: "#CCCCCC", borderColor: "#3A3A3A" } } }, Ke = { style: { backgroundColor: "#1A1A1A", zeroLine: { color: "#5A5A5A" }, plot: { stroke: "#1A1A1A" }, verticalIndicator: { color: "#8A8A8A" }, dataLabel: { color: "#CCCCCC" }, title: { color: "#CCCCCC" }, tooltip: { color: "#CCCCCC", backgroundColor: "#1A1A1A", borderColor: "#5A5A5A", backgroundOpacity: 70 } } }, Ue = { style: { backgroundColor: "#FFF8E1", line: { color: "#D32F2F" }, bar: { color: "#D32F2F", borderRadius: 0 }, zeroLine: { color: "#5D4037" }, plot: { stroke: "#FFF8E1" }, verticalIndicator: { color: "#FF9800" }, dataLabel: { color: "#424242" }, title: { color: "#424242" }, area: { color: "#E64A19" }, tooltip: { backgroundColor: "#FFECB3", backgroundOpacity: 30, color: "#424242", borderColor: "#FF8A65" } } }, Xe = { style: { backgroundColor: "#1E1E1E", line: { color: "#D32F2F" }, bar: { color: "#D32F2F", borderRadius: 0 }, zeroLine: { color: "#5D4037" }, plot: { stroke: "#1E1E1E" }, verticalIndicator: { color: "#B71C1C" }, dataLabel: { color: "#BDBDBD" }, title: { color: "#BDBDBD" }, area: { color: "#E64A19" }, tooltip: { backgroundColor: "#1E1E1E", backgroundOpacity: 30, color: "#FFF8E1", borderColor: "#FF8A65" } } }, Ye = { style: { backgroundColor: "#1A1A1A", line: { color: "#66CC66" }, bar: { color: "#66CC66", borderRadius: 0 }, zeroLine: { color: "#5F6A5F" }, plot: { stroke: "#1A1A1A" }, verticalIndicator: { color: "#66CC66" }, dataLabel: { color: "#AACCAA" }, title: { color: "#66CC66" }, area: { color: "#66CC66" }, tooltip: { backgroundColor: "#2A2F2A", color: "#AACCAA", borderColor: "#66CC66" } } }, Ze = { style: { backgroundColor: "#fbfafa", line: { color: "#B9B99D", smooth: !0 }, bar: { color: "#B9B99D", borderRadius: 3 }, zeroLine: { strokeWidth: 0 }, plot: { stroke: "#fbfafa" }, verticalIndicator: { color: "#B9B99D" }, dataLabel: { color: "#A0AC94" }, title: { color: "#8A9892" }, area: { color: "#B9B99D" }, tooltip: { backgroundColor: "#fbfafa", color: "#8A9892" } } }, Qe = { style: { backgroundColor: "#f6f6fb", line: { color: "#4A6A75" }, bar: { color: "#4A6A75", borderRadius: 0 }, zeroLine: { color: "#9FA9BE" }, plot: { stroke: "#f6f6fb" }, verticalIndicator: { color: "#4A6A75" }, dataLabel: { color: "#61747E" }, title: { color: "#50606C" }, area: { color: "#4A6A75" }, tooltip: { backgroundColor: "#f6f6fb", color: "#50606C" } } }, He = {
  default: {},
  minimal: We,
  minimalDark: Ge,
  dark: Ke,
  celebration: Ue,
  celebrationNight: Xe,
  hack: Ye,
  zen: Ze,
  concrete: Qe
}, _l = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  celebration: Ue,
  celebrationNight: Xe,
  concrete: Qe,
  dark: Ke,
  default: He,
  hack: Ye,
  minimal: We,
  minimalDark: Ge,
  zen: Ze
}, Symbol.toStringTag, { value: "Module" }));
function Bt() {
  const o = m(!1);
  return xe(() => {
    const b = window.matchMedia("(prefers-reduced-motion: reduce)"), I = () => {
      o.value = b.matches;
    };
    I(), b.addEventListener("change", I), te(() => {
      b.removeEventListener("change", I);
    });
  }), o;
}
const Rt = {
  key: 0,
  style: { "pointer-events": "none" }
}, Et = ["r", "fill", "filter"], zt = ["begin", "dur", "repeatCount", "fill", "calcMode", "keySplines", "keyTimes", "keyPoints"], Ot = ["href"], Vt = ["dur", "repeatCount", "fill", "values"], qt = ["r", "fill"], jt = ["begin", "dur", "repeatCount", "calcMode", "keySplines", "keyTimes", "keyPoints"], Wt = ["href"], Gt = ["dur", "repeatCount"], Kt = {
  __name: "SparklinePulse",
  props: {
    uid: { type: String, required: !0 },
    svgRef: { type: Object, default: null },
    pulsePathId: { type: String, required: !0 },
    pulsePathLength: { type: Number, required: !0 },
    pulseDur: { type: String, required: !0 },
    pulseBegin: { type: String, required: !0 },
    pulseRepeatCount: { type: String, required: !0 },
    pulseFillMode: { type: String, default: void 0 },
    pulseKeyPoints: { type: String, required: !0 },
    pulseMotion: { type: Object, required: !0 },
    pulse: { type: Object, required: !0 },
    pulseTrail: { type: Object, required: !0 },
    pulseTrailLength: { type: Number, required: !0 },
    prefersReducedMotion: { type: Boolean, required: !0 },
    loading: { type: Boolean, required: !0 },
    isBar: { type: Boolean, required: !0 }
  },
  setup(o) {
    const b = o, I = s(() => !!b.pulse?.show && !b.isBar && !b.prefersReducedMotion && !b.loading && Number(b.pulsePathLength) > 0);
    function T() {
      const v = b.svgRef?.value;
      v && (typeof v.pauseAnimations == "function" && v.pauseAnimations(), typeof v.setCurrentTime == "function" && v.setCurrentTime(0), typeof v.unpauseAnimations == "function" && v.unpauseAnimations());
    }
    function le(v, d) {
      if (typeof v != "string") return v;
      const x = v.trim().match(/^([\d.]+)\s*(ms|s)$/);
      if (!x) return v;
      const f = Number(x[1]), O = x[2];
      return Number.isNaN(f) ? v : `${f + d}${O}`;
    }
    function ae(v) {
      return (b.pulseTrailLength - v) / b.pulseTrailLength * b.pulse.radius;
    }
    function oe(v) {
      const d = b.pulse.trail.opacity, x = (b.pulseTrailLength - v) / b.pulseTrailLength, f = v === 0 ? 1 : x * d;
      return `0;${f};${f};0`;
    }
    return w(
      () => b.loading,
      async (v) => {
        if (v) {
          T();
          return;
        }
        await M(), T();
      }
    ), w(
      () => b.pulsePathId,
      async () => {
        await M(), T();
      }
    ), te(() => {
      T();
    }), (v, d) => I.value ? (i(), c("g", Rt, [
      (i(!0), c(J, null, ee(o.pulseTrailLength, (x, f) => (i(), c("circle", {
        key: `sparkline_dot_${f}_${o.pulsePathId}`,
        r: ae(f),
        fill: o.pulse.color,
        filter: `url(#sparkline_pulse_glow_${o.uid})`,
        opacity: "0"
      }, [
        g("animateMotion", {
          begin: le(o.pulseBegin, f * 10),
          dur: o.pulseDur,
          repeatCount: o.pulseRepeatCount,
          fill: o.pulseFillMode,
          calcMode: o.pulseMotion.calcMode,
          keySplines: o.pulseMotion.keySplines || void 0,
          keyTimes: o.pulseMotion.keyTimes || void 0,
          keyPoints: o.pulseKeyPoints,
          rotate: "auto"
        }, [
          g("mpath", {
            href: `#${o.pulsePathId}`
          }, null, 8, Ot)
        ], 8, zt),
        g("animate", {
          attributeName: "opacity",
          dur: o.pulseDur,
          repeatCount: o.pulseRepeatCount,
          fill: o.pulseFillMode,
          values: oe(f),
          keyTimes: "0;0.1;0.9;1"
        }, null, 8, Vt)
      ], 8, Et))), 128)),
      (i(), c("circle", {
        key: `sparkline_halo_${o.pulsePathId}`,
        r: Math.max(o.pulse.radius * 1.3),
        fill: o.pulse.color,
        opacity: "0"
      }, [
        g("animateMotion", {
          begin: o.pulseBegin,
          dur: o.pulseDur,
          repeatCount: o.pulseRepeatCount,
          calcMode: o.pulseMotion.calcMode,
          keySplines: o.pulseMotion.keySplines || void 0,
          keyTimes: o.pulseMotion.keyTimes || void 0,
          keyPoints: o.pulseKeyPoints,
          rotate: "auto"
        }, [
          g("mpath", {
            href: `#${o.pulsePathId}`
          }, null, 8, Wt)
        ], 8, jt),
        g("animate", {
          attributeName: "opacity",
          values: "0;0.35;0.35;0",
          keyTimes: "0;0.15;0.85;1",
          dur: o.pulseDur,
          repeatCount: o.pulseRepeatCount
        }, null, 8, Gt)
      ], 8, qt))
    ])) : k("", !0);
  }
}, Ut = ["id"], Xt = ["xmlns", "viewBox"], Yt = ["width", "height"], Zt = ["id"], Qt = ["stop-color"], Ht = ["stop-color"], Jt = ["id"], el = ["stop-color"], tl = ["stop-color"], ll = ["id"], al = ["stop-color"], ol = ["stop-color"], ul = ["id"], nl = { key: 1 }, rl = ["d", "fill"], sl = ["d", "fill"], il = ["id", "d", "stroke", "stroke-width"], dl = ["id", "d", "stroke", "stroke-width"], cl = ["x", "y", "width", "height", "fill", "rx"], vl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], fl = ["x1", "x2", "y1", "y2", "stroke", "stroke-dasharray", "stroke-width"], hl = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], pl = ["x", "y", "font-size", "font-weight", "fill"], yl = ["x", "y", "height", "width", "onMouseenter", "onMouseleave", "onClick"], gl = {
  __name: "vue-ui-sparkline",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    showInfo: {
      type: Boolean,
      default: !0
    },
    selectedIndex: {
      type: Number,
      default: void 0
    },
    heightRatio: {
      type: Number,
      default: 1
    },
    forcedPadding: {
      type: Number,
      default: 30
    }
  },
  emits: ["hoverIndex", "selectDatapoint"],
  setup(o, { emit: b }) {
    const I = De(() => import("./PackageVersion-DEUaLSdf.js")), T = De(() => import("./SparkTooltip-Cvo31nyB.js")), { vue_ui_sparkline: le } = Nt(), { isThemeValid: ae, warnInvalidTheme: oe } = Ft(), v = Bt(), d = o, x = s(() => Array.isArray(d.dataset) && d.dataset.length > 0), f = m(kt()), O = m(null), ue = m(null), Le = m(null), e = m(re());
    function Je(t) {
      return d.config?.skeletonDataset && Array.isArray(d.config.skeletonDataset) ? d.config.skeletonDataset.map((l) => ({ period: "-", value: l })) : Mt(t).map((l) => ({ period: "-", value: l }));
    }
    const et = s(() => ze({
      defaultConfig: {
        style: {
          backgroundColor: "#99999930",
          scaleMin: 0,
          scaleMax: null,
          animation: { show: !1 },
          line: { color: "#AAAAAA", pulse: { show: !1 } },
          bar: { color: "#AAAAAA" },
          area: { color: "#CACACA" },
          zeroLine: { color: "#6A6A6A" },
          dataLabel: { show: !1 },
          tooltip: { show: !1 }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: C, FINAL_DATASET: V, manualLoading: ne } = It({
      ...pt(d),
      FINAL_CONFIG: e,
      prepareConfig: re,
      callback: () => {
        Promise.resolve().then(async () => {
          await M(), fe();
        });
      },
      skeletonDataset: Je(12),
      skeletonConfig: ze({
        defaultConfig: e.value,
        userConfig: et.value
      })
    }), { svgRef: U } = Pt({ config: e.value.style.title });
    function re() {
      const t = we({
        userConfig: d.config,
        defaultConfig: le
      });
      let l = {};
      const a = t.theme;
      if (a)
        if (!ae.value(t))
          oe(t), l = t;
        else {
          const u = we({
            userConfig: He[a] || d.config,
            defaultConfig: t
          });
          l = {
            ...we({
              userConfig: d.config,
              defaultConfig: u
            })
          };
        }
      else
        l = t;
      return l;
    }
    const A = s(() => e.value?.style?.line.pulse || {}), tt = s(() => `${Math.max(200, Number(A.value.durationMs) || 4e3) / 1e3}s`), X = m(0), lt = s(() => A.value?.begin || "0ms"), at = m("0;1"), ot = s(() => A.value?.loop === !1 ? "1" : "indefinite"), ut = s(() => A.value?.loop === !1 ? "freeze" : void 0), nt = s(() => A.value.trail.show && Me.value.lengthPx || 1), se = s(() => !!A.value?.show && !S.value && !v.value && !C.value && (h.value?.length || 0) > 1);
    function ie() {
      if (!se.value) {
        X.value = 0;
        return;
      }
      const t = U.value;
      if (!t) return;
      const l = `#${W.value}`, a = t.querySelector?.(l);
      if (a && typeof a.getTotalLength == "function") {
        const u = a.getTotalLength();
        Number.isFinite(u) && u > 0 && (X.value = u);
        return;
      }
      requestAnimationFrame(() => {
        const u = t.querySelector?.(l);
        if (u && typeof u.getTotalLength == "function") {
          const F = u.getTotalLength();
          Number.isFinite(F) && F > 0 && (X.value = F);
        }
      });
    }
    const rt = s(() => {
      const t = A.value?.easing || "ease-in-out", l = {
        ease: [0.25, 0.1, 0.25, 1],
        "ease-in": [0.42, 0, 1, 1],
        "ease-out": [0, 0, 0.58, 1],
        "ease-in-out": [0.42, 0, 0.58, 1]
      };
      return t === "linear" ? {
        calcMode: "linear",
        keySplines: null,
        keyTimes: "0;1"
      } : t === "steps" ? {
        calcMode: "discrete",
        keySplines: null,
        keyTimes: "0;1"
      } : {
        calcMode: "spline",
        keySplines: (t === "cubic-bezier" ? Array.isArray(A.value?.cubicBezier) && A.value.cubicBezier.length === 4 ? A.value.cubicBezier : [0.4, 0, 0.2, 1] : l[t] || l["ease-in-out"]).join(" "),
        keyTimes: "0;1"
      };
    }), Me = s(() => {
      const t = A.value?.trail || {}, l = e.value.style.line.strokeWidth || 1;
      return {
        show: t.show !== !1,
        lengthPx: t.length,
        width: Math.max(1, Number(t.strokeWidth) || l * 2.2),
        opacity: Math.min(1, Math.max(0, Number(t.opacity) ?? 0.6)),
        fadeIn: 0.5,
        fadeOut: 0.2
      };
    }), P = s(() => ke({
      data: V.value,
      threshold: e.value.downsample.threshold
    }));
    w(() => d.config, (t) => {
      C.value || (e.value = re()), Ie(), r.value.chartWidth = e.value.style.chartWidth;
    }, { deep: !0 }), w(() => d.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (ne.value = !1), L.value = ke({
        data: V.value.map((l) => ({
          ...l,
          value: [void 0].includes(l.value) ? null : l.value
        })),
        threshold: e.value.downsample.threshold
      });
    }, { deep: !0 });
    const L = m(st());
    function st() {
      return ke({
        data: V.value.map((t) => e.value.style.animation.show && V.value.length > 1 ? {
          ...t,
          value: null
        } : {
          ...t,
          value: [void 0].includes(t.value) ? null : t.value
        }),
        threshold: e.value.downsample.threshold
      });
    }
    const $ = Be(null), D = Be(null), q = m(!1), j = m(0), de = m([]), ce = m(""), ve = s(() => {
      const l = (P.value || []).map((u) => `${u.period}::${Number.isFinite(u.value) ? u.value : 0}`).join("|"), a = e.value?.style?.animation || {};
      return `${l}#${!!a.show}#${a.animationFrames || 0}`;
    });
    function Y() {
      j.value && (cancelAnimationFrame(j.value), j.value = 0), de.value.forEach((t) => clearTimeout(t)), de.value = [], q.value = !1;
    }
    function fe() {
      const t = e.value?.style?.animation || {}, l = P.value || [], a = ve.value;
      if (a && a === ce.value && (q.value || L.value.length === l.length))
        return;
      if (Y(), !t.show || C.value || l.length <= 1) {
        L.value = l, ce.value = a;
        return;
      }
      q.value = !0, ce.value = a, L.value = [];
      const u = Math.max(1, Number(t.animationFrames) || 1), F = Math.max(1, Math.floor(u / l.length));
      let be = 0;
      const Pe = () => {
        if (a !== ve.value) {
          Y();
          return;
        }
        if (be < l.length) {
          L.value.push(l[be]);
          const ht = setTimeout(() => {
            j.value = requestAnimationFrame(Pe);
          }, F);
          de.value.push(ht), be += 1;
        } else
          L.value = l, Y();
      };
      j.value = requestAnimationFrame(Pe);
    }
    w(ve, () => {
      fe();
    }), xe(() => {
      Ie(), fe();
    }), te(() => {
      Y();
    });
    const $e = m(0);
    w(
      () => C.value,
      async (t) => {
        t || (await M(), $e.value += 1);
      }
    ), w(
      () => q.value,
      async (t) => {
        t || C.value || (await M(), $e.value += 1);
      }
    );
    const W = s(() => `sparkline_line_path_${f.value}`), he = m(!0);
    async function _e() {
      he.value = !1, await M(), he.value = !0, await M(), ie();
    }
    const Ne = s(() => !!e.value.debug);
    function Ie() {
      if (Oe(d.dataset) ? (Ve({
        componentName: "VueUiSparkline",
        type: "dataset",
        debug: Ne.value
      }), ne.value = !0) : d.dataset.forEach((t, l) => {
        Ct({
          datasetObject: t,
          requiredAttributes: ["period", "value"]
        }).forEach((a) => {
          Ve({
            componentName: "VueUiSparkline",
            type: "datasetSerieAttribute",
            property: a,
            index: l,
            debug: Ne.value
          });
        });
      }), Oe(d.dataset) || (ne.value = e.value.loading), e.value.responsive) {
        const t = $t(() => {
          const { width: l, height: a } = _t({
            chart: O.value,
            title: e.value.style.title.show && d.showInfo ? ue.value : null,
            source: Le.value
          });
          requestAnimationFrame(() => {
            r.value.width = l, r.value.height = a, r.value.chartWidth = e.value.style.chartWidth / 500 * l, r.value.padding = d.forcedPadding / 500 * l;
          });
        });
        $.value && (D.value && $.value.unobserve(D.value), $.value.disconnect()), $.value = new ResizeObserver(t), D.value = O.value.parentNode, $.value.observe(D.value);
      }
    }
    te(() => {
      $.value && (D.value && $.value.unobserve(D.value), $.value.disconnect());
    });
    const r = m({
      height: 80 * d.heightRatio,
      width: 500,
      chartWidth: e.value.style.chartWidth,
      padding: d.forcedPadding
    }), pe = b, p = s(() => {
      const { top: t, right: l, bottom: a, left: u } = e.value.style.padding;
      return {
        top: t,
        left: u,
        right: r.value.width - l,
        bottom: r.value.height - a,
        start: d.showInfo && e.value.style.dataLabel.show && e.value.style.dataLabel.position === "left" ? r.value.width - r.value.chartWidth + u : r.value.padding + u,
        width: d.showInfo && e.value.style.dataLabel.show ? r.value.chartWidth - u - l : r.value.width - r.value.padding - u - l,
        height: r.value.height - t - a
      };
    }), ye = s(() => [null, void 0].includes(e.value.style.scaleMin) ? Math.min(...L.value.map((t) => isNaN(t.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(t.value) ? 0 : t.value || 0)) : e.value.style.scaleMin), it = s(() => [null, void 0].includes(e.value.style.scaleMax) ? Math.max(...L.value.map((t) => isNaN(t.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(t.value) ? 0 : t.value || 0)) : e.value.style.scaleMax), G = s(() => {
      const t = ye.value >= 0 ? 0 : ye.value;
      return Math.abs(t);
    }), Se = s(() => it.value + G.value), B = s(() => p.value.bottom - p.value.height * ge(G.value));
    function ge(t) {
      return isNaN(t / Se.value) ? 0 : t / Se.value;
    }
    const R = s(() => P.value.length - 1 || 1), K = m([]);
    let Fe = 0;
    yt(() => {
      const t = ++Fe;
      (async () => {
        const l = await Tt({
          values: P.value.map((a) => a.period),
          maxDatapoints: P.value.length,
          formatter: e.value.style.dataLabel.datetimeFormatter,
          start: 0,
          end: P.value.length
        });
        t === Fe && (K.value = l);
      })();
    });
    const h = s(() => L.value.map((t, l) => {
      const a = isNaN(t.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(t.value) ? 0 : t.value || 0, u = p.value.width / R.value;
      return {
        absoluteValue: a,
        period: K.value && K.value[l] && K.value[l].text ? K.value[l].text : t.period,
        plotValue: a + G.value,
        toMax: ge(a + G.value),
        x: p.value.start + l * u,
        y: p.value.bottom - p.value.height * ge(a + G.value),
        id: `plot_${f.value}_${l}`,
        color: S.value ? e.value.style.bar.color : e.value.style.area.useGradient ? Q(e.value.style.line.color, 0.05 * (1 - l / R.value)) : e.value.style.line.color,
        width: u
      };
    })), dt = s(() => {
      const t = { x: h.value[0].x || 0, y: (r.value.height || 0) - 6 }, l = { x: h.value[h.value.length - 1].x || 0, y: (r.value.height || 0) - 6 }, a = [];
      return h.value.forEach((u) => {
        a.push(`${u.x || 0},${u.y || 0} `);
      }), [t.x, t.y, ...a, l.x, l.y].toString();
    }), y = m(void 0), E = m(void 0);
    function ct(t, l) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: l }), y.value = t, E.value || (E.value = t), pe("hoverIndex", { index: l });
    }
    function vt(t, l) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: l }), E.value = y.value, y.value = void 0, pe("hoverIndex", { index: void 0 });
    }
    const _ = s(() => {
      if (x.value) {
        const t = h.value.map((a) => a.absoluteValue), l = t.reduce((a, u) => a + u, 0);
        return {
          latest: h.value[h.value.length - 1] ? h.value[h.value.length - 1].absoluteValue : 0,
          sum: l,
          average: l / h.value.length,
          median: wt(t),
          trend: At(h.value.map(({ x: a, y: u, absoluteValue: F }) => ({
            x: a,
            y: u,
            value: F
          }))).trend
        };
      } else
        return {
          latest: null,
          sum: null,
          average: null,
          median: null,
          trend: null
        };
    }), Te = s(() => x.value ? e.value.style.dataLabel.valueType === "latest" ? _.value.latest : e.value.style.dataLabel.valueType === "sum" ? _.value.sum : e.value.style.dataLabel.valueType === "average" ? _.value.average : 0 : 0), S = s(() => e.value.type && e.value.type === "bar");
    function ft(t, l) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: l }), pe("selectDatapoint", { datapoint: t, index: l });
    }
    return w(
      () => [
        se.value,
        W.value,
        h.value.length,
        r.value.width,
        r.value.height,
        e.value?.style?.line?.smooth
      ],
      async () => {
        await M(), ie();
      },
      { immediate: !0 }
    ), xe(async () => {
      await M(), ie();
    }), w(
      () => C.value,
      async (t) => {
        t || await _e();
      }
    ), w(
      () => q.value,
      async (t) => {
        t || C.value || await _e();
      }
    ), (t, l) => (i(), c("div", {
      ref_key: "sparklineChart",
      ref: O,
      class: "vue-data-ui-component vue-ui-sparkline",
      id: f.value,
      style: N(`width:100%;font-family:${e.value.style.fontFamily};`)
    }, [
      z(t.$slots, "before", Re(Ee({
        selected: y.value,
        latest: _.value.latest,
        sum: _.value.sum,
        average: _.value.average,
        median: _.value.median,
        trend: _.value.trend
      })), void 0, !0),
      e.value.style.title.show && o.showInfo ? (i(), c("div", {
        key: 0,
        ref_key: "chartTitle",
        ref: ue,
        class: "vue-ui-sparkline-title",
        style: N(`display:flex;align-items:center;width:100%;color:${e.value.style.title.color};background:${e.value.style.backgroundColor};justify-content:${e.value.style.title.textAlign === "left" ? "flex-start" : e.value.style.title.textAlign === "right" ? "flex-end" : "center"};height:${e.value.style.title.fontSize * 2}px;font-size:${e.value.style.title.fontSize}px;font-weight:${e.value.style.title.bold ? "bold" : "normal"};`)
      }, [
        g("span", {
          style: N(`padding:${e.value.style.title.textAlign === "left" ? "0 0 0 12px" : e.value.style.title.textAlign === "right" ? "0 12px 0 0" : "0"}`)
        }, Z(y.value ? y.value.period : e.value.style.title.text), 5)
      ], 4)) : k("", !0),
      (i(), c("svg", {
        ref_key: "svgRef",
        ref: U,
        xmlns: n(xt),
        viewBox: `0 0 ${r.value.width} ${r.value.height}`,
        style: N(`background:${e.value.style.backgroundColor};overflow:visible`),
        onMouseleave: l[0] || (l[0] = (a) => E.value = void 0)
      }, [
        gt(n(I)),
        t.$slots["chart-background"] ? (i(), c("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: r.value.width <= 0 ? 10 : r.value.width,
          height: r.value.height <= 0 ? 10 : r.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          z(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Yt)) : k("", !0),
        g("defs", null, [
          g("linearGradient", {
            x1: "0%",
            y1: "0%",
            x2: "100%",
            y2: "0%",
            id: `sparkline_gradient_${f.value}`
          }, [
            g("stop", {
              offset: "0%",
              "stop-color": n(H)(n(Q)(e.value.style.area.color, 0.05), e.value.style.area.opacity)
            }, null, 8, Qt),
            g("stop", {
              offset: "100%",
              "stop-color": n(H)(e.value.style.area.color, e.value.style.area.opacity)
            }, null, 8, Ht)
          ], 8, Zt),
          g("linearGradient", {
            x2: "0%",
            y2: "100%",
            id: `sparkline_bar_gradient_pos_${f.value}`
          }, [
            g("stop", {
              offset: "0%",
              "stop-color": e.value.style.bar.color
            }, null, 8, el),
            g("stop", {
              offset: "100%",
              "stop-color": n(Q)(e.value.style.bar.color, 0.05)
            }, null, 8, tl)
          ], 8, Jt),
          g("linearGradient", {
            x2: "0%",
            y2: "100%",
            id: `sparkline_bar_gradient_neg_${f.value}`
          }, [
            g("stop", {
              offset: "0%",
              "stop-color": n(Q)(e.value.style.bar.color, 0.05)
            }, null, 8, al),
            g("stop", {
              offset: "100%",
              "stop-color": e.value.style.bar.color
            }, null, 8, ol)
          ], 8, ll),
          g("filter", {
            id: `sparkline_pulse_glow_${f.value}`,
            filterUnits: "userSpaceOnUse",
            x: "-50",
            y: "-50",
            width: "100",
            height: "100"
          }, [...l[1] || (l[1] = [
            g("feGaussianBlur", {
              in: "SourceGraphic",
              stdDeviation: "3",
              result: "blur"
            }, null, -1),
            g("feMerge", null, [
              g("feMergeNode", { in: "blur" }),
              g("feMergeNode", { in: "SourceGraphic" })
            ], -1)
          ])], 8, ul)
        ]),
        e.value.style.area.show && !S.value && h.value[0] && h.value.length > 1 ? (i(), c("g", nl, [
          e.value.style.line.smooth ? (i(), c("path", {
            key: 0,
            class: "vue-ui-sparkline-area",
            d: `M ${h.value[0].x},${p.value.bottom} ${n(qe)(h.value)} L ${h.value.at(-1).x},${p.value.bottom} Z`,
            fill: e.value.style.area.useGradient ? `url(#sparkline_gradient_${f.value})` : n(H)(e.value.style.area.color, e.value.style.area.opacity),
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            style: N({
              transition: n(C) ? void 0 : "all 0.2s"
            })
          }, null, 12, rl)) : (i(), c("path", {
            key: 1,
            class: "vue-ui-sparkline-area",
            d: `M${dt.value}Z`,
            fill: e.value.style.area.useGradient ? `url(#sparkline_gradient_${f.value})` : n(H)(e.value.style.area.color, e.value.style.area.opacity),
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            style: N({
              transition: n(C) ? void 0 : "all 0.2s"
            })
          }, null, 12, sl))
        ])) : k("", !0),
        e.value.style.line.smooth && !S.value ? (i(), c("path", {
          key: 2,
          class: "vue-ui-sparkline-path",
          id: W.value,
          d: `M ${n(qe)(h.value) || "0,0"}`,
          stroke: e.value.style.line.color,
          fill: "none",
          "stroke-width": e.value.style.line.strokeWidth,
          "stroke-linecap": "round",
          "stroke-linejoin": "round",
          style: N({
            transition: n(C) ? void 0 : "all 0.2s"
          })
        }, null, 12, il)) : k("", !0),
        !e.value.style.line.smooth && !S.value ? (i(), c("path", {
          key: 3,
          class: "vue-ui-sparkline-path",
          id: W.value,
          d: `M ${n(Lt)(h.value) || "0,0"}`,
          stroke: e.value.style.line.color,
          fill: "none",
          "stroke-width": e.value.style.line.strokeWidth,
          "stroke-linecap": "round",
          "stroke-linejoin": "round",
          style: N({
            transition: n(C) ? void 0 : "all 0.2s"
          })
        }, null, 12, dl)) : k("", !0),
        he.value && se.value ? (i(), me(Kt, {
          key: 4,
          uid: f.value,
          svgRef: n(U),
          pulsePathId: W.value,
          pulsePathLength: X.value,
          pulseDur: tt.value,
          pulseBegin: lt.value,
          pulseRepeatCount: ot.value,
          pulseFillMode: ut.value,
          pulseKeyPoints: at.value,
          pulseMotion: rt.value,
          pulse: A.value,
          pulseTrail: Me.value,
          pulseTrailLength: nt.value,
          prefersReducedMotion: n(v),
          loading: n(C),
          isBar: S.value
        }, null, 8, ["uid", "svgRef", "pulsePathId", "pulsePathLength", "pulseDur", "pulseBegin", "pulseRepeatCount", "pulseFillMode", "pulseKeyPoints", "pulseMotion", "pulse", "pulseTrail", "pulseTrailLength", "prefersReducedMotion", "loading", "isBar"])) : k("", !0),
        (i(!0), c(J, null, ee(h.value, (a, u) => (i(), c("g", null, [
          S.value ? (i(), c("rect", {
            key: 0,
            x: a.x - a.width / 2,
            y: isNaN(a.absoluteValue > 0 ? a.y : B.value) ? 0 : a.absoluteValue > 0 ? a.y : B.value,
            width: a.width,
            height: isNaN(Math.abs(a.y - B.value)) ? 0 : Math.abs(a.y - B.value),
            fill: a.absoluteValue > 0 ? `url(#sparkline_bar_gradient_pos_${f.value})` : `url(#sparkline_bar_gradient_neg_${f.value})`,
            rx: e.value.style.bar.borderRadius
          }, null, 8, cl)) : k("", !0),
          e.value.style.verticalIndicator.show && (y.value && a.id === y.value.id || o.selectedIndex === u) ? (i(), c("line", {
            key: 1,
            x1: a.x,
            x2: a.x,
            y1: p.value.top - 6,
            y2: p.value.bottom,
            stroke: e.value.style.verticalIndicator.color || a.color,
            "stroke-width": e.value.style.verticalIndicator.strokeWidth,
            "stroke-linecap": "round",
            "stroke-dasharray": e.value.style.verticalIndicator.strokeDasharray || 0
          }, null, 8, vl)) : k("", !0)
        ]))), 256)),
        ye.value < 0 ? (i(), c("line", {
          key: 5,
          x1: p.value.start,
          x2: p.value.start + p.value.width,
          y1: n(je)(B.value, p.value.bottom),
          y2: n(je)(B.value, p.value.bottom),
          stroke: e.value.style.zeroLine.color,
          "stroke-dasharray": e.value.style.zeroLine.strokeWidth * 2,
          "stroke-width": e.value.style.zeroLine.strokeWidth,
          "stroke-linecap": "round"
        }, null, 8, fl)) : k("", !0),
        e.value.style.plot.show ? (i(!0), c(J, { key: 6 }, ee(h.value, (a, u) => (i(), c("g", null, [
          y.value && a.id === y.value.id || o.selectedIndex === u || n(V).length === 1 ? (i(), c("circle", {
            key: 0,
            cx: a.x,
            cy: a.y,
            r: e.value.style.plot.radius,
            fill: a.color,
            stroke: e.value.style.plot.stroke,
            "stroke-width": e.value.style.plot.strokeWidth
          }, null, 8, hl)) : k("", !0)
        ]))), 256)) : k("", !0),
        o.showInfo && e.value.style.dataLabel.show ? (i(), c("text", {
          key: 7,
          x: e.value.style.dataLabel.position === "left" ? 12 + e.value.style.dataLabel.offsetX : p.value.width + 12 + e.value.style.dataLabel.offsetX,
          y: r.value.height / 2 + e.value.style.dataLabel.fontSize / 2.5 + e.value.style.dataLabel.offsetY,
          "font-size": e.value.style.dataLabel.fontSize,
          "font-weight": e.value.style.dataLabel.bold ? "bold" : "normal",
          fill: e.value.style.dataLabel.color
        }, Z(y.value ? n(Ce)(
          e.value.style.dataLabel.formatter,
          y.value.absoluteValue,
          n(Ae)({ p: e.value.style.dataLabel.prefix, v: y.value.absoluteValue, s: e.value.style.dataLabel.suffix, r: e.value.style.dataLabel.roundingValue }),
          { datapoint: y.value }
        ) : n(Ce)(
          e.value.style.dataLabel.formatter,
          Te.value,
          n(Ae)({ p: e.value.style.dataLabel.prefix, v: Te.value, s: e.value.style.dataLabel.suffix, r: e.value.style.dataLabel.roundingValue })
        )), 9, pl)) : k("", !0),
        (i(!0), c(J, null, ee(h.value, (a, u) => (i(), c("rect", {
          x: a.x - (p.value.width / (R.value + 1) > r.value.padding ? r.value.padding : p.value.width / (R.value + 1)) / 2,
          y: p.value.top - 6,
          height: p.value.height + 6,
          width: p.value.width / (R.value + 1) > r.value.padding ? r.value.padding : p.value.width / (R.value + 1),
          fill: "transparent",
          onMouseenter: () => ct(a, u),
          onMouseleave: () => vt(a, u),
          onClick: () => ft(a, u)
        }, null, 40, yl))), 256)),
        z(t.$slots, "svg", { svg: r.value }, void 0, !0)
      ], 44, Xt)),
      y.value && e.value.style.tooltip.show ? (i(), me(n(T), {
        key: 1,
        x: y.value.x,
        y: y.value.y,
        prevX: E.value.x,
        prevY: E.value.y,
        offsetY: e.value.style.plot.radius * 3 + e.value.style.tooltip.offsetY,
        svgRef: n(U),
        background: e.value.style.tooltip.backgroundColor,
        color: e.value.style.tooltip.color,
        fontSize: e.value.style.tooltip.fontSize,
        borderWidth: e.value.style.tooltip.borderWidth,
        borderColor: e.value.style.tooltip.borderColor,
        borderRadius: e.value.style.tooltip.borderRadius,
        backgroundOpacity: e.value.style.tooltip.backgroundOpacity
      }, {
        default: bt(() => [
          z(t.$slots, "tooltip", Re(Ee({ ...y.value })), () => [
            mt(Z(y.value.period) + ": " + Z(n(Ce)(
              e.value.style.dataLabel.formatter,
              y.value.absoluteValue,
              n(Ae)({
                p: e.value.style.dataLabel.prefix,
                v: y.value.absoluteValue,
                s: e.value.style.dataLabel.suffix,
                r: e.value.style.dataLabel.roundingValue
              }),
              { datapoint: y.value }
            )), 1)
          ], !0)
        ]),
        _: 3
      }, 8, ["x", "y", "prevX", "prevY", "offsetY", "svgRef", "background", "color", "fontSize", "borderWidth", "borderColor", "borderRadius", "backgroundOpacity"])) : k("", !0),
      t.$slots.source ? (i(), c("div", {
        key: 2,
        ref_key: "source",
        ref: Le,
        dir: "auto"
      }, [
        z(t.$slots, "source", {}, void 0, !0)
      ], 512)) : k("", !0),
      z(t.$slots, "skeleton", {}, () => [
        n(C) ? (i(), me(St, { key: 0 })) : k("", !0)
      ], !0)
    ], 12, Ut));
  }
}, bl = /* @__PURE__ */ Dt(gl, [["__scopeId", "data-v-fe99dd90"]]), Nl = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: bl
}, Symbol.toStringTag, { value: "Module" }));
export {
  _l as a,
  Nl as b,
  bl as v
};
