// vue-data-ui/dist/vue-ui-sparkline-D-rQ9PsC.js
import { computed as m, watch as z, nextTick as J, onBeforeUnmount as nt, openBlock as f, createElementBlock as k, Fragment as we, renderList as Pe, createElementVNode as A, createCommentVNode as L, ref as C, defineAsyncComponent as ft, toRefs as Bt, shallowRef as pt, onMounted as gt, watchEffect as Dt, normalizeStyle as ne, toDisplayString as Ce, createBlock as Le, renderSlot as ve, normalizeProps as Xe, guardReactiveProps as Ke, unref as y, createVNode as Et, withCtx as Ft, createTextVNode as Vt } from "vue";
import { j as qt, t as yt, aD as Ue, o as mt, q as bt, r as Gt, v as Ve, Y as Wt, ap as Ot, c as He, b as kt, s as zt, X as jt, u as qe, V as Yt, G as wt, k as Ze, l as Je, aE as Xt } from "./lib-DBptbyyI.js";
import { t as Kt, u as Ut } from "./useResponsive-DfdjqQps.js";
import { u as Ht } from "./useConfig-DO94Aj5s.js";
import { u as Zt, B as Jt } from "./BaseScanner-BhhkKVoj.js";
import { u as Qe } from "./useNestedProp-Bf5XFbs8.js";
import { u as Qt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ea } from "./useTimeLabels-C-A2uZKi.js";
import { u as ta } from "./useChartAccessibility-9icAAmYg.js";
import { u as aa } from "./usePrefersMotion-CUKm8_t8.js";
import la from "./vue_ui_sparkline-CQ8DJVJx.js";
import { A as na } from "./A11yDataTable-KcVAX2cP.js";
import { _ as ra } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ua = {
  key: 0,
  style: { "pointer-events": "none" }
}, oa = ["r", "fill", "filter"], sa = ["begin", "dur", "repeatCount", "fill", "calcMode", "keySplines", "keyTimes", "keyPoints"], ia = ["href"], da = ["dur", "repeatCount", "fill", "values"], ca = ["r", "fill"], va = ["begin", "dur", "repeatCount", "calcMode", "keySplines", "keyTimes", "keyPoints"], ha = ["href"], fa = ["dur", "repeatCount"], pa = {
  __name: "SparklinePulse",
  props: {
    uid: { type: String, required: !0 },
    svgRef: { type: Object, default: null },
    pulsePathId: { type: String, required: !0 },
    pulsePathLength: { type: Number, required: !0 },
    pulseDur: { type: String, required: !0 },
    pulseBegin: { type: String, required: !0 },
    pulseRepeatCount: { type: String, required: !0 },
    pulseFillMode: { type: String, default: void 0 },
    pulseKeyPoints: { type: String, required: !0 },
    pulseMotion: { type: Object, required: !0 },
    pulse: { type: Object, required: !0 },
    pulseTrail: { type: Object, required: !0 },
    pulseTrailLength: { type: Number, required: !0 },
    prefersReducedMotion: { type: Boolean, required: !0 },
    loading: { type: Boolean, required: !0 },
    isBar: { type: Boolean, required: !0 }
  },
  setup(e) {
    const a = e, n = m(() => Math.min(a.pulseTrailLength, 45)), s = m(() => !!a.pulse?.show && !a.isBar && !a.prefersReducedMotion && !a.loading && Number(a.pulsePathLength) > 0);
    function o() {
      const i = a.svgRef?.value;
      i && (typeof i.pauseAnimations == "function" && i.pauseAnimations(), typeof i.setCurrentTime == "function" && i.setCurrentTime(0), typeof i.unpauseAnimations == "function" && i.unpauseAnimations());
    }
    function d(i, M) {
      if (typeof i != "string") return i;
      const h = i.trim().match(/^([\d.]+)\s*(ms|s)$/);
      if (!h) return i;
      const w = Number(h[1]), P = h[2];
      return Number.isNaN(w) ? i : `${w + M}${P}`;
    }
    function v(i) {
      return (n.value - i) / n.value * a.pulse.radius;
    }
    function p(i) {
      const M = a.pulse.trail.opacity, h = (n.value - i) / n.value, w = i === 0 ? 1 : h * M;
      return `0;${w};${w};0`;
    }
    return z(
      () => a.loading,
      async (i) => {
        if (i) {
          o();
          return;
        }
        await J(), o();
      }
    ), z(
      () => a.pulsePathId,
      async () => {
        await J(), o();
      }
    ), nt(() => {
      o();
    }), (i, M) => s.value ? (f(), k("g", ua, [
      (f(!0), k(we, null, Pe(n.value, (h, w) => (f(), k(we, null, [
        w % 3 === 0 ? (f(), k("circle", {
          key: `sparkline_dot_${w}_${e.pulsePathId}`,
          r: v(w),
          fill: e.pulse.color,
          filter: `url(#sparkline_pulse_glow_${e.uid})`,
          opacity: "0"
        }, [
          A("animateMotion", {
            begin: d(e.pulseBegin, w * 10),
            dur: e.pulseDur,
            repeatCount: e.pulseRepeatCount,
            fill: e.pulseFillMode,
            calcMode: e.pulseMotion.calcMode,
            keySplines: e.pulseMotion.keySplines || void 0,
            keyTimes: e.pulseMotion.keyTimes || void 0,
            keyPoints: e.pulseKeyPoints,
            rotate: "auto"
          }, [
            A("mpath", {
              href: `#${e.pulsePathId}`
            }, null, 8, ia)
          ], 8, sa),
          A("animate", {
            attributeName: "opacity",
            dur: e.pulseDur,
            repeatCount: e.pulseRepeatCount,
            fill: e.pulseFillMode,
            values: p(w),
            keyTimes: "0;0.1;0.9;1"
          }, null, 8, da)
        ], 8, oa)) : L("", !0)
      ], 64))), 256)),
      (f(), k("circle", {
        key: `sparkline_halo_${e.pulsePathId}`,
        r: Math.max(e.pulse.radius * 1.3),
        fill: e.pulse.color,
        opacity: "0"
      }, [
        A("animateMotion", {
          begin: e.pulseBegin,
          dur: e.pulseDur,
          repeatCount: e.pulseRepeatCount,
          calcMode: e.pulseMotion.calcMode,
          keySplines: e.pulseMotion.keySplines || void 0,
          keyTimes: e.pulseMotion.keyTimes || void 0,
          keyPoints: e.pulseKeyPoints,
          rotate: "auto"
        }, [
          A("mpath", {
            href: `#${e.pulsePathId}`
          }, null, 8, ha)
        ], 8, va),
        A("animate", {
          attributeName: "opacity",
          values: "0;0.35;0.35;0",
          keyTimes: "0;0.15;0.85;1",
          dur: e.pulseDur,
          repeatCount: e.pulseRepeatCount
        }, null, 8, fa)
      ], 8, ca))
    ])) : L("", !0);
  }
}, Ge = {
  canvas: null,
  ctx: null
};
function ga() {
  if (typeof document > "u")
    throw new Error("color-utils: document is not available (browser-only).");
  if (!Ge.canvas) {
    const e = document.createElement("canvas");
    e.width = 1, e.height = 1;
    const a = e.getContext("2d", { willReadFrequently: !0 });
    if (!a)
      throw new Error("color-utils: unable to get 2D canvas context.");
    Ge.canvas = e, Ge.ctx = a;
  }
  return Ge.ctx;
}
function ya(e) {
  return e < 0 ? 0 : e > 255 ? 255 : Math.round(e);
}
function ut(e) {
  return e < 0 ? 0 : e > 1 ? 1 : e;
}
function ma(e, a = null) {
  if (typeof e != "string" || e.trim().length === 0)
    throw new Error("colorToRgba: inputColor must be a non-empty string.");
  const n = e.trim();
  if (n.toLowerCase() === "transparent")
    return { red: 0, green: 0, blue: 0, alpha: 0 };
  const s = ba(n, a), o = ga();
  o.clearRect(0, 0, 1, 1), o.fillStyle = "#000";
  const d = o.fillStyle;
  if (o.fillStyle = s, o.fillStyle === d && !ka(s))
    throw new Error(`colorToRgba: unsupported or invalid color "${e}".`);
  o.fillRect(0, 0, 1, 1);
  const p = o.getImageData(0, 0, 1, 1).data;
  return {
    red: p[0],
    green: p[1],
    blue: p[2],
    alpha: p[3] / 255
  };
}
function xt(e, a = null) {
  if (!Array.isArray(e) || e.length === 0)
    throw new Error("colorsToRgba: colors must be a non-empty array.");
  const n = new Array(e.length);
  for (let s = 0; s < e.length; s += 1)
    n[s] = ma(e[s], a);
  return n;
}
function ba(e, a = null) {
  return e !== "currentColor" || typeof window > "u" || !a ? e : window.getComputedStyle(a).color || e;
}
function ka(e) {
  const a = e.trim().toLowerCase();
  return a === "black" || a === "#000" || a === "#000000" || a === "rgb(0, 0, 0)" || a === "rgba(0, 0, 0, 1)" || a === "rgba(0, 0, 0, 0)";
}
function be(e) {
  const a = ut(e / 255);
  return a <= 0.04045 ? a / 12.92 : Math.pow((a + 0.055) / 1.055, 2.4);
}
function et(e) {
  const a = ut(e);
  let n;
  return a <= 31308e-7 ? n = a * 12.92 : n = 1.055 * Math.pow(a, 1 / 2.4) - 0.055, ya(n * 255);
}
function $t(e, a, n, s) {
  const o = ut(n);
  if (s === "linearRGB") {
    const d = be(e.red), v = be(e.green), p = be(e.blue), i = be(a.red), M = be(a.green), h = be(a.blue), w = d + (i - d) * o, P = v + (M - v) * o, N = p + (h - p) * o;
    return {
      red: et(w),
      green: et(P),
      blue: et(N),
      alpha: e.alpha + (a.alpha - e.alpha) * o
    };
  }
  return {
    red: e.red + (a.red - e.red) * o,
    green: e.green + (a.green - e.green) * o,
    blue: e.blue + (a.blue - e.blue) * o,
    alpha: e.alpha + (a.alpha - e.alpha) * o
  };
}
function O(e, a, n) {
  let s = e;
  return s < a && (s = a), s > n && (s = n), s;
}
function Q(e, a) {
  const n = Math.pow(10, a);
  return Math.round(e * n) / n;
}
function re(e) {
  return Number.isFinite(e) ? e.toString() : "0";
}
function tt(e) {
  const n = O(e, 0, 255).toString(16);
  return n.length === 1 ? "0" + n : n;
}
function wa(e) {
  if (typeof e != "string") return;
  const a = Number.parseFloat(e);
  return Number.isFinite(a) ? a : void 0;
}
function xa(e, a, n, s = !0) {
  const o = n - a;
  if (o === 0) return 0;
  const d = (e - a) / o;
  return s ? O(d, 0, 1) : d;
}
const Mt = "http://www.w3.org/2000/svg";
function $a(e) {
  if (typeof document > "u")
    throw new Error("SvgPathGradientAsync: document is not available (browser-only implementation).");
  const a = document.createElementNS(Mt, "svg");
  a.setAttribute("width", "0"), a.setAttribute("height", "0"), a.setAttribute("viewBox", "0 0 0 0"), a.style.position = "absolute", a.style.left = "-10000px", a.style.top = "-10000px", a.style.visibility = "hidden", a.style.pointerEvents = "none";
  const n = document.createElementNS(Mt, "path");
  return e === "M " && (e = "M 0,0"), n.setAttribute("d", e), a.appendChild(n), document.body.appendChild(a), { svgElement: a, pathElement: n };
}
function at(e) {
  if (e)
    try {
      e.svgElement?.remove?.();
    } catch {
    }
}
function Ma(e, a, n) {
  if (typeof n == "number" && n > 0)
    return n;
  let s = a * 0.75;
  s < 0.5 && (s = 0.5), s > 8 && (s = 8);
  const o = e / 2;
  return s > o && o > 0 && (s = o), s <= 0 && (s = 1), s;
}
function Aa(e, a, n) {
  if (typeof a == "number" && a >= 1)
    return Math.floor(a);
  const s = typeof n == "number" && n > 0 ? n : 2, o = Math.ceil(e / s);
  return o < 1 ? 1 : o;
}
function Sa(e, a, n, s, o, d) {
  const v = n - a;
  let p = Math.ceil(v / s) + 1;
  p < 2 && (p = 2), p > d && (p = d);
  const i = p > 1 ? v / (p - 1) : 0, M = e.getPointAtLength(a);
  let h = `M ${re(Q(M.x, o))} ${re(
    Q(M.y, o)
  )}`;
  for (let w = 1; w < p; w += 1) {
    const P = a + w * i, N = e.getPointAtLength(P);
    h += ` L ${re(Q(N.x, o))} ${re(
      Q(N.y, o)
    )}`;
  }
  return h;
}
function Ca(e, a, n, s, o, d) {
  const v = n - a;
  let p = Math.ceil(v / s) + 1;
  p < 2 && (p = 2), p > d && (p = d);
  const i = p > 1 ? v / (p - 1) : 0, M = e.getPointAtLength(a);
  let h = M.x, w = M.x, P = M.y, N = M.y, q = `M ${re(Q(M.x, o))} ${re(
    Q(M.y, o)
  )}`;
  for (let t = 1; t < p; t += 1) {
    const ue = a + t * i, B = e.getPointAtLength(ue);
    B.x < h && (h = B.x), B.x > w && (w = B.x), B.y < P && (P = B.y), B.y > N && (N = B.y), q += ` L ${re(Q(B.x, o))} ${re(
      Q(B.y, o)
    )}`;
  }
  return {
    pathData: q,
    bounds: { minX: h, maxX: w, minY: P, maxY: N }
  };
}
function La(e) {
  if (e.length === 1) return [{ position: 0, rgba: e[0] }];
  const a = new Array(e.length), n = e.length - 1;
  for (let s = 0; s < e.length; s += 1)
    a[s] = { position: s / n, rgba: e[s] };
  return a;
}
function Pa(e, a, n, s) {
  const o = O(a, 0, 1);
  let d = s;
  for (d < 0 && (d = 0), d > e.length - 2 && (d = e.length - 2); d > 0 && o < e[d].position; ) d -= 1;
  for (; d < e.length - 2 && o > e[d + 1].position; ) d += 1;
  const v = e[d], p = e[d + 1], i = p.position - v.position, M = i > 0 ? (o - v.position) / i : 0, h = _a(v.rgba, p.rgba, M, n);
  return rt(h);
}
function _a(e, a, n, s) {
  const o = O(n, 0, 1);
  if (s === "linearRGB") {
    const d = ke(e.red), v = ke(e.green), p = ke(e.blue), i = ke(a.red), M = ke(a.green), h = ke(a.blue), w = d + (i - d) * o, P = v + (M - v) * o, N = p + (h - p) * o;
    return {
      red: lt(w),
      green: lt(P),
      blue: lt(N),
      alpha: e.alpha + (a.alpha - e.alpha) * o
    };
  }
  return {
    red: e.red + (a.red - e.red) * o,
    green: e.green + (a.green - e.green) * o,
    blue: e.blue + (a.blue - e.blue) * o,
    alpha: e.alpha + (a.alpha - e.alpha) * o
  };
}
function ke(e) {
  const a = O(e / 255, 0, 1);
  return a <= 0.04045 ? a / 12.92 : Math.pow((a + 0.055) / 1.055, 2.4);
}
function lt(e) {
  const a = O(e, 0, 1);
  let n;
  a <= 31308e-7 ? n = a * 12.92 : n = 1.055 * Math.pow(a, 1 / 2.4) - 0.055;
  const s = Math.round(n * 255);
  return O(s, 0, 255);
}
function rt(e) {
  const a = O(e.alpha, 0, 1), n = O(Math.round(e.red), 0, 255), s = O(Math.round(e.green), 0, 255), o = O(Math.round(e.blue), 0, 255);
  if (a >= 1)
    return `#${tt(n)}${tt(s)}${tt(o)}`;
  const d = Q(a, 4).toString();
  return `rgba(${n}, ${s}, ${o}, ${d})`;
}
function Na(e) {
  const a = e.fill ?? "none";
  let n = `fill="${ee(a)}"`;
  const s = Object.keys(e);
  for (let o = 0; o < s.length; o += 1) {
    const d = s[o];
    if (d === "stroke" || d === "d") continue;
    const v = e[d];
    n += ` ${d}="${ee(v)}"`;
  }
  return n;
}
function At(e, a, n) {
  return `<path d="${ee(e)}" stroke="${ee(a)}" ${n} />`;
}
function Ta(e) {
  const a = [], n = e.segmentAttributeMap.fill ?? "none";
  a.push(`fill="${ee(n)}"`), a.push(`stroke="${ee(e.stroke)}"`);
  for (const s of Object.keys(e.segmentAttributeMap)) {
    if (s === "stroke" || s === "d") continue;
    const o = e.segmentAttributeMap[s];
    a.push(`${s}="${ee(o)}"`);
  }
  return `<path d="${ee(e.pathData)}" ${a.join(" ")} />`;
}
function St(e, a) {
  const n = [];
  for (const o of Object.keys(a))
    n.push(`${o}="${ee(a[o])}"`);
  return `<g data-svg-path-gradient="true"${n.length ? " " + n.join(" ") : ""}>${e.join("")}</g>`;
}
function ee(e) {
  return String(e).replaceAll("&", "&amp;").replaceAll('"', "&quot;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
}
function Ia(e, a = null, n = {}) {
  return new Promise((s, o) => {
    let d = null;
    try {
      if (!e || typeof e != "string")
        throw new Error("SvgPathGradientAsync: pathData must be a non-empty string.");
      const v = n.temperatureMode ?? null;
      if (v === null && (!Array.isArray(a) || a.length === 0))
        throw new Error("SvgPathGradientAsync: colors must be a non-empty array when temperatureMode is not enabled.");
      if ((n.returnMode ?? "string") !== "string")
        throw new Error('SvgPathGradientAsync: only returnMode="string" is supported for the async version.');
      const i = typeof n.decimalPlaces == "number" ? n.decimalPlaces : 3, M = typeof n.flattenTolerance == "number" && n.flattenTolerance > 0 ? n.flattenTolerance : 0.25, h = n.attrs ? { ...n.attrs } : {}, w = n.groupAttrs ? { ...n.groupAttrs } : {}, P = wa(h["stroke-width"]), N = typeof n.strokeWidth == "number" && n.strokeWidth > 0 ? n.strokeWidth : typeof P == "number" && P > 0 ? P : 1;
      d = $a(e);
      const q = d.pathElement, t = q.getTotalLength(), ue = v === null ? xt(a, n.colorReferenceElement) : [];
      if (v === null && (!(t > 0) || ue.length === 1)) {
        const G = rt(ue[0]), W = St(
          [
            Ta({
              pathData: e,
              stroke: G,
              segmentAttributeMap: h
            })
          ],
          w
        );
        at(d), s(W);
        return;
      }
      const B = n.colorSpace ?? "linearRGB";
      if (v !== null) {
        const G = n.temperatureColors;
        if (!Array.isArray(G) || G.length !== 2)
          throw new Error("SvgPathGradientAsync: temperatureColors must be a tuple of exactly 2 colors when temperatureMode is enabled.");
      }
      const D = v !== null ? xt(n.temperatureColors, n.colorReferenceElement) : null, U = v === null ? (() => {
        const G = La(ue), W = new Array(G.length);
        for (let I = 0; I < G.length; I += 1)
          W[I] = { position: G[I].position, rgba: G[I].rgba };
        return W;
      })() : [], xe = Ma(t, N, n.maxSegmentLength), j = Aa(t, n.segments, xe), oe = t / j;
      let T = typeof n.overlap == "number" && n.overlap >= 0 ? n.overlap : N * 0.5;
      const _e = oe * 0.45;
      T > _e && (T = _e);
      const se = typeof n.samplePointLimitPerSegment == "number" && n.samplePointLimitPerSegment > 10 ? n.samplePointLimitPerSegment : 250, Ne = Na(h), he = [], Te = v !== null ? [] : null, ie = v !== null ? [] : null;
      let fe = 0;
      const $e = v === null ? U.length - 2 : 0;
      let Y = 0;
      const Ie = () => {
        const G = performance.now(), W = typeof n.frameBudgetMs == "number" ? n.frameBudgetMs : 8;
        for (; Y < j && performance.now() - G < W; ) {
          const te = Y * oe, V = (Y + 1) * oe;
          let R = te, E = V;
          if (Y !== 0 && (R = te - T), Y !== j - 1 && (E = V + T), R = O(R, 0, t), E = O(E, 0, t), E > R)
            if (v !== null) {
              const _ = Ca(
                q,
                R,
                E,
                M,
                i,
                se
              );
              Te.push(_.pathData), ie.push(_.bounds);
            } else {
              const F = (R + E) * 0.5 / t;
              for (; fe < $e && F > U[fe + 1].position; )
                fe += 1;
              const X = Pa(
                U,
                F,
                B,
                fe
              ), K = Sa(
                q,
                R,
                E,
                M,
                i,
                se
              );
              he.push(
                At(K, X, Ne)
              );
            }
          Y += 1;
        }
        if (Y < j) {
          requestAnimationFrame(Ie);
          return;
        }
        if (v !== null) {
          let te = Number.POSITIVE_INFINITY, V = Number.NEGATIVE_INFINITY;
          for (let _ = 0; _ < ie.length; _ += 1) {
            const F = ie[_], X = v === "vertical" ? F.minY : F.minX, K = v === "vertical" ? F.maxY : F.maxX;
            X < te && (te = X), K > V && (V = K);
          }
          const R = D[0], E = D[1];
          he.length = 0;
          for (let _ = 0; _ < ie.length; _ += 1) {
            const F = ie[_], X = v === "vertical" ? F.minY : F.minX, K = v === "vertical" ? F.maxY : F.maxX, pe = (X + K) * 0.5, ge = xa(pe, te, V, !0), Re = v === "vertical" ? $t(R, E, ge, B) : $t(E, R, ge, B), de = rt(Re), Me = Te[_];
            he.push(
              At(Me, de, Ne)
            );
          }
        }
        const I = St(he, w);
        at(d), s(I);
      };
      requestAnimationFrame(Ie);
    } catch (v) {
      at(d), o(v);
    }
  });
}
const Ra = ["innerHTML"], Ba = {
  __name: "SparklineGradientPath",
  props: {
    svgPathData: { type: String, required: !0 },
    enabled: { type: Boolean, required: !0 },
    strokeWidth: { type: Number, required: !0 },
    highColor: { type: String, required: !0 },
    lowColor: { type: String, required: !0 },
    segments: { type: Number, required: !0 }
  },
  setup(e) {
    const a = e, n = C("");
    let s = 0;
    return z(
      () => [
        a.enabled,
        a.svgPathData,
        a.strokeWidth,
        a.highColor,
        a.lowColor,
        a.segments
      ],
      async ([o]) => {
        const d = ++s;
        if (!o) {
          n.value = "";
          return;
        }
        const v = await Ia(a.svgPathData, null, {
          segments: a.segments,
          temperatureMode: "vertical",
          temperatureColors: [a.highColor, a.lowColor],
          attrs: {
            "stroke-width": a.strokeWidth,
            "stroke-linecap": "round",
            "stroke-linejoin": "round"
          },
          groupAttrs: { class: "vue-ui-sparkline-gradient" }
        });
        d === s && (n.value = v);
      },
      { immediate: !0 }
    ), (o, d) => (f(), k("g", { innerHTML: n.value }, null, 8, Ra));
  }
}, Da = ["id"], Ea = ["id"], Fa = { style: { position: "relative" } }, Va = ["xmlns", "viewBox", "aria-describedby"], qa = ["width", "height"], Ga = ["id"], Wa = ["stop-color"], Oa = ["stop-color"], za = ["id"], ja = ["stop-color"], Ya = ["stop-color"], Xa = ["id"], Ka = ["stop-color"], Ua = ["stop-color"], Ha = ["id"], Za = ["id"], Ja = ["stop-color", "offset"], Qa = { key: 1 }, el = ["d", "fill"], tl = ["d", "fill"], al = ["id", "d", "stroke", "stroke-width"], ll = ["id", "d", "stroke", "stroke-width"], nl = ["x", "y", "width", "height", "fill", "rx"], rl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], ul = ["x1", "x2", "y1", "y2", "stroke", "stroke-dasharray", "stroke-width"], ol = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], sl = ["x", "y", "font-size", "font-weight", "fill"], il = ["x", "y", "height", "width", "onMouseenter", "onMouseleave", "onClick"], dl = {
  key: 0,
  style: { position: "absolute", top: "100%", left: "0", width: "100%" },
  "data-dom-to-png-ignore": "",
  "aria-hidden": "true"
}, cl = {
  __name: "vue-ui-sparkline",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    showInfo: {
      type: Boolean,
      default: !0
    },
    selectedIndex: {
      type: Number,
      default: void 0
    },
    heightRatio: {
      type: Number,
      default: 1
    },
    forcedPadding: {
      type: Number,
      default: 30
    }
  },
  emits: ["hoverIndex", "selectDatapoint"],
  setup(e, { emit: a }) {
    const n = ft(() => import("./PackageVersion-DPO-OOVI.js")), s = ft(() => import("./SparkTooltip-BorBpltj.js")), { vue_ui_sparkline: o } = Ht(), { isThemeValid: d, warnInvalidTheme: v } = Qt(), p = aa(), i = e, M = m(() => Array.isArray(i.dataset) && i.dataset.length > 0), h = C(qt()), w = C(null), P = C(null), N = C(null), q = C(null), t = C(oe());
    function ue(l) {
      return i.config?.skeletonDataset && Array.isArray(i.config.skeletonDataset) ? i.config.skeletonDataset.map((r) => ({ period: "-", value: r })) : Xt(l).map((r) => ({ period: "-", value: r }));
    }
    const B = m(() => yt({
      defaultConfig: {
        gradientPath: { show: !1 },
        temperatureColors: { show: !1 },
        style: {
          backgroundColor: "#99999930",
          scaleMin: 0,
          scaleMax: null,
          animation: { show: !1 },
          line: { color: "#AAAAAA", pulse: { show: !1 } },
          bar: { color: "#AAAAAA" },
          area: { color: "#CACACA" },
          zeroLine: { color: "#6A6A6A" },
          dataLabel: { show: !1 },
          tooltip: { show: !1 }
        }
      },
      userConfig: t.value.skeletonConfig ?? {}
    })), { loading: D, FINAL_DATASET: U, manualLoading: xe } = Zt({
      ...Bt(i),
      FINAL_CONFIG: t,
      prepareConfig: oe,
      callback: () => {
        Promise.resolve().then(async () => {
          await J(), ge();
        });
      },
      skeletonDataset: ue(12),
      skeletonConfig: yt({
        defaultConfig: t.value,
        userConfig: B.value
      })
    }), { svgRef: j } = ta({ config: t.value.style.title });
    function oe() {
      const l = Qe({
        userConfig: i.config,
        defaultConfig: o
      });
      let r = {};
      const u = l.theme;
      if (u)
        if (!d.value(l))
          v(l), r = l;
        else {
          const c = Qe({
            userConfig: la[u] || i.config,
            defaultConfig: l
          });
          r = {
            ...Qe({
              userConfig: i.config,
              defaultConfig: c
            })
          };
        }
      else
        r = l;
      return r;
    }
    const T = m(() => t.value?.style?.line.pulse || {}), _e = m(() => `${Math.max(200, Number(T.value.durationMs) || 4e3) / 1e3}s`), se = C(0), Ne = m(() => T.value?.begin || "0ms"), he = C("0;1"), Te = m(() => T.value?.loop === !1 ? "1" : "indefinite"), ie = m(() => T.value?.loop === !1 ? "freeze" : void 0), fe = m(() => T.value.trail.show && G.value.lengthPx || 1), $e = m(() => !!T.value?.show && !H.value && !p.value && !D.value && (g.value?.length || 0) > 1);
    function Y() {
      if (!$e.value) {
        se.value = 0;
        return;
      }
      const l = j.value;
      if (!l) return;
      const r = `#${de.value}`, u = l.querySelector?.(r);
      if (u && typeof u.getTotalLength == "function") {
        const c = u.getTotalLength();
        Number.isFinite(c) && c > 0 && (se.value = c);
        return;
      }
      requestAnimationFrame(() => {
        const c = l.querySelector?.(r);
        if (c && typeof c.getTotalLength == "function") {
          const b = c.getTotalLength();
          Number.isFinite(b) && b > 0 && (se.value = b);
        }
      });
    }
    const Ie = m(() => {
      const l = T.value?.easing || "ease-in-out", r = {
        ease: [0.25, 0.1, 0.25, 1],
        "ease-in": [0.42, 0, 1, 1],
        "ease-out": [0, 0, 0.58, 1],
        "ease-in-out": [0.42, 0, 0.58, 1]
      };
      return l === "linear" ? {
        calcMode: "linear",
        keySplines: null,
        keyTimes: "0;1"
      } : l === "steps" ? {
        calcMode: "discrete",
        keySplines: null,
        keyTimes: "0;1"
      } : {
        calcMode: "spline",
        keySplines: (l === "cubic-bezier" ? Array.isArray(T.value?.cubicBezier) && T.value.cubicBezier.length === 4 ? T.value.cubicBezier : [0.4, 0, 0.2, 1] : r[l] || r["ease-in-out"]).join(" "),
        keyTimes: "0;1"
      };
    }), G = m(() => {
      const l = T.value?.trail || {}, r = t.value.style.line.strokeWidth || 1;
      return {
        show: l.show !== !1,
        lengthPx: l.length,
        width: Math.max(1, Number(l.strokeWidth) || r * 2.2),
        opacity: Math.min(1, Math.max(0, Number(l.opacity) ?? 0.6)),
        fadeIn: 0.5,
        fadeOut: 0.2
      };
    }), W = m(() => Ue({
      data: U.value,
      threshold: t.value.downsample.threshold
    }));
    z(() => i.config, (l) => {
      D.value || (t.value = oe()), it(), x.value.chartWidth = t.value.style.chartWidth;
    }, { deep: !0 }), z(() => i.dataset, (l) => {
      Array.isArray(l) && l.length > 0 && (xe.value = !1), I.value = Ue({
        data: U.value.map((r) => ({
          ...r,
          value: [void 0].includes(r.value) ? null : r.value
        })),
        threshold: t.value.downsample.threshold
      });
    }, { deep: !0 });
    const I = C(te());
    function te() {
      return Ue({
        data: U.value.map((l) => t.value.style.animation.show && U.value.length > 1 ? {
          ...l,
          value: null
        } : {
          ...l,
          value: [void 0].includes(l.value) ? null : l.value
        }),
        threshold: t.value.downsample.threshold
      });
    }
    const V = pt(null), R = pt(null), E = C(!1), _ = C(0), F = C([]), X = C(""), K = m(() => {
      const r = (W.value || []).map((b) => `${b.period}::${Number.isFinite(b.value) ? b.value : 0}`).join("|"), u = t.value?.style?.animation || {}, c = !!t.value?.gradientPath?.show && !t.value.temperatureColors.show;
      return `${r}#${!!u.show}#${u.animationFrames || 0}#${c}`;
    });
    function pe() {
      _.value && (cancelAnimationFrame(_.value), _.value = 0), F.value.forEach((l) => clearTimeout(l)), F.value = [], E.value = !1;
    }
    function ge() {
      const l = t.value?.style?.animation || {}, r = W.value || [], u = K.value, c = !!t.value.gradientPath.show && !t.value.temperatureColors.show;
      if (u && u === X.value && (E.value || I.value.length === r.length))
        return;
      if (pe(), c || !l.show || D.value || r.length <= 1) {
        I.value = r, X.value = u;
        return;
      }
      E.value = !0, X.value = u, I.value = [];
      const b = Math.max(1, Number(l.animationFrames) || 1), Ye = Math.max(1, Math.floor(b / r.length));
      let Z = 0;
      const Ee = () => {
        if (u !== K.value) {
          pe();
          return;
        }
        if (Z < r.length) {
          I.value.push(r[Z]);
          const Fe = setTimeout(() => {
            _.value = requestAnimationFrame(Ee);
          }, Ye);
          F.value.push(Fe), Z += 1;
        } else
          I.value = r, pe();
      };
      _.value = requestAnimationFrame(Ee);
    }
    z(K, () => {
      ge();
    }), gt(() => {
      it(), ge();
    }), nt(() => {
      pe();
    });
    const Re = C(0);
    z(
      () => D.value,
      async (l) => {
        l || (await J(), Re.value += 1);
      }
    ), z(
      () => E.value,
      async (l) => {
        l || D.value || (await J(), Re.value += 1);
      }
    );
    const de = m(() => `sparkline_line_path_${h.value}`), Me = C(!0);
    async function ot() {
      Me.value = !1, await J(), Me.value = !0, await J(), Y();
    }
    const st = m(() => !!t.value.debug);
    function it() {
      if (mt(i.dataset) ? (bt({
        componentName: "VueUiSparkline",
        type: "dataset",
        debug: st.value
      }), xe.value = !0) : i.dataset.forEach((l, r) => {
        Gt({
          datasetObject: l,
          requiredAttributes: ["period", "value"]
        }).forEach((u) => {
          bt({
            componentName: "VueUiSparkline",
            type: "datasetSerieAttribute",
            property: u,
            index: r,
            debug: st.value
          });
        });
      }), mt(i.dataset) || (xe.value = t.value.loading), t.value.responsive) {
        const l = Kt(() => {
          const { width: r, height: u } = Ut({
            chart: w.value,
            title: t.value.style.title.show && i.showInfo ? P.value : null,
            source: N.value
          });
          requestAnimationFrame(() => {
            x.value.width = r, x.value.height = u, x.value.chartWidth = t.value.style.chartWidth / 500 * r, x.value.padding = i.forcedPadding / 500 * r;
          });
        });
        V.value && (R.value && V.value.unobserve(R.value), V.value.disconnect()), V.value = new ResizeObserver(l), R.value = w.value.parentNode, V.value.observe(R.value);
      }
    }
    nt(() => {
      V.value && (R.value && V.value.unobserve(R.value), V.value.disconnect());
    });
    const x = C({
      height: 80 * i.heightRatio,
      width: 500,
      chartWidth: t.value.style.chartWidth,
      padding: i.forcedPadding
    }), Be = a, S = m(() => {
      const { top: l, right: r, bottom: u, left: c } = t.value.style.padding;
      return {
        top: l,
        left: c,
        right: x.value.width - r,
        bottom: x.value.height - u,
        start: i.showInfo && t.value.style.dataLabel.show && t.value.style.dataLabel.position === "left" ? x.value.width - x.value.chartWidth + c : x.value.padding + c,
        width: i.showInfo && t.value.style.dataLabel.show ? x.value.chartWidth - c - r : x.value.width - x.value.padding - c - r,
        height: x.value.height - l - u
      };
    }), We = m(() => [null, void 0].includes(t.value.style.scaleMin) ? Math.min(...I.value.map((l) => isNaN(l.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(l.value) ? 0 : l.value || 0)) : t.value.style.scaleMin), Ct = m(() => [null, void 0].includes(t.value.style.scaleMax) ? Math.max(...I.value.map((l) => isNaN(l.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(l.value) ? 0 : l.value || 0)) : t.value.style.scaleMax), Ae = m(() => {
      const l = We.value >= 0 ? 0 : We.value;
      return Math.abs(l);
    }), dt = m(() => Ct.value + Ae.value), ye = m(() => S.value.bottom - S.value.height * Oe(Ae.value));
    function Oe(l) {
      return isNaN(l / dt.value) ? 0 : l / dt.value;
    }
    const me = m(() => W.value.length - 1 || 1), Se = C([]);
    let ct = 0;
    Dt(() => {
      const l = ++ct;
      (async () => {
        const r = await ea({
          values: W.value.map((u) => u.period),
          maxDatapoints: W.value.length,
          formatter: t.value.style.dataLabel.datetimeFormatter,
          start: 0,
          end: W.value.length
        });
        l === ct && (Se.value = r);
      })();
    });
    const g = m(() => I.value.map((l, r) => {
      const u = isNaN(l.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(l.value) ? 0 : l.value || 0, c = S.value.width / me.value;
      return {
        absoluteValue: u,
        period: Se.value && Se.value[r] && Se.value[r].text ? Se.value[r].text : l.period,
        plotValue: u + Ae.value,
        toMax: Oe(u + Ae.value),
        x: S.value.start + r * c,
        y: S.value.bottom - S.value.height * Oe(u + Ae.value),
        id: `plot_${h.value}_${r}`,
        color: H.value ? t.value.style.bar.color : t.value.style.area.useGradient ? Ve(t.value.style.line.color, 0.05 * (1 - r / me.value)) : t.value.style.line.color,
        width: c
      };
    })), Lt = m(() => {
      const l = { x: g.value[0].x || 0, y: (x.value.height || 0) - 6 }, r = { x: g.value[g.value.length - 1].x || 0, y: (x.value.height || 0) - 6 }, u = [];
      return g.value.forEach((c) => {
        u.push(`${c.x || 0},${c.y || 0} `);
      }), [l.x, l.y, ...u, r.x, r.y].toString();
    }), $ = C(void 0), ae = C(void 0);
    function De(l, r) {
      t.value.events.datapointEnter && t.value.events.datapointEnter({ datapoint: l, seriesIndex: r }), q.value = r, $.value = l, ae.value || (ae.value = l), Be("hoverIndex", { index: r });
    }
    function Pt(l, r) {
      t.value.events.datapointLeave && t.value.events.datapointLeave({ datapoint: l, seriesIndex: r }), ae.value = $.value, $.value = void 0, q.value = null, Be("hoverIndex", { index: void 0 });
    }
    const le = m(() => {
      if (M.value) {
        const l = g.value.map((u) => u.absoluteValue), r = l.reduce((u, c) => u + c, 0);
        return {
          latest: g.value[g.value.length - 1] ? g.value[g.value.length - 1].absoluteValue : 0,
          sum: r,
          average: r / g.value.length,
          median: Ot(l),
          trend: Wt(g.value.map(({ x: u, y: c, absoluteValue: b }) => ({
            x: u,
            y: c,
            value: b
          }))).trend
        };
      } else
        return {
          latest: null,
          sum: null,
          average: null,
          median: null,
          trend: null
        };
    }), vt = m(() => M.value ? t.value.style.dataLabel.valueType === "latest" ? le.value.latest : t.value.style.dataLabel.valueType === "sum" ? le.value.sum : t.value.style.dataLabel.valueType === "average" ? le.value.average : 0 : 0), H = m(() => t.value.type && t.value.type === "bar");
    function _t(l, r) {
      t.value.events.datapointClick && t.value.events.datapointClick({ datapoint: l, seriesIndex: r }), Be("selectDatapoint", { datapoint: l, index: r });
    }
    const ht = m(() => H.value || !t.value.gradientPath.show || t.value.temperatureColors.show ? "" : `M ${(t.value.style.line.smooth ? He(g.value) : kt(g.value)) || "0,0"}`), ce = m(() => !t.value.temperatureColors.show || t.value.temperatureColors.colors.length === 0 ? null : t.value.temperatureColors.colors.map((l) => zt(l)));
    z(
      () => [
        $e.value,
        de.value,
        g.value.length,
        x.value.width,
        x.value.height,
        t.value?.style?.line?.smooth
      ],
      async () => {
        await J(), Y();
      },
      { immediate: !0 }
    ), gt(async () => {
      await J(), Y();
    }), z(
      () => D.value,
      async (l) => {
        l || await ot();
      }
    ), z(
      () => E.value,
      async (l) => {
        l || D.value || await ot();
      }
    ), z(
      () => i.selectedIndex,
      (l) => {
        if (l == null || l < 0 || l >= g.value.length) return;
        const r = g.value[l];
        r && (q.value = l, De(r, l));
      }
    );
    const ze = C(!1);
    function Nt() {
      q.value = null, $.value || De(g.value.at(-1), g.value.length - 1), ze.value = !0;
    }
    function Tt() {
      q.value = null, ae.value = $.value, $.value = void 0, Be("hoverIndex", { index: void 0 }), ze.value = !1;
    }
    function It(l) {
      if (!j.value || document.activeElement !== j.value) return;
      ae.value = $.value;
      const r = l.key === "ArrowLeft", u = l.key === "ArrowRight";
      if (!r && !u) return;
      const c = g.value.length;
      if (!c) return;
      l.preventDefault(), l.stopPropagation();
      let b = q.value;
      const Ye = b !== null && b >= 0 && b < c, Z = $.value ? g.value.findIndex((Rt) => Rt.id === $.value.id) : -1, Ee = Z !== null && Z >= 0 && Z < c;
      Ye ? u ? (b += 1, b >= c && (b = 0)) : r && (b -= 1, b < 0 && (b = c - 1)) : Ee ? (b = u ? Z + 1 : Z - 1, b >= c && (b = 0), b < 0 && (b = c - 1)) : u ? b = 0 : b = c - 1;
      const Fe = g.value[b];
      Fe && (q.value = b, De(Fe, b));
    }
    const je = m(() => {
      const l = [
        t.value.translations.period,
        t.value.translations.value
      ], r = g.value.map((u) => [u.period, u.absoluteValue]);
      return { headers: l, rows: r };
    });
    return (l, r) => (f(), k("div", {
      ref_key: "sparklineChart",
      ref: w,
      class: "vue-data-ui-component vue-ui-sparkline",
      id: h.value,
      style: ne(`width:100%;font-family:${t.value.style.fontFamily};`)
    }, [
      A("p", {
        id: `chart-instructions-${h.value}`,
        class: "sr-only"
      }, Ce(t.value.a11y.translations.keyboardNavigation), 9, Ea),
      je.value?.rows?.length ? (f(), Le(na, {
        key: 0,
        uid: h.value,
        head: je.value.headers,
        body: je.value.rows,
        notice: t.value.a11y.translations.tableAvailable,
        caption: t.value.a11y.translations.tableCaption
      }, null, 8, ["uid", "head", "body", "notice", "caption"])) : L("", !0),
      ve(l.$slots, "before", Xe(Ke({
        selected: $.value,
        latest: le.value.latest,
        sum: le.value.sum,
        average: le.value.average,
        median: le.value.median,
        trend: le.value.trend
      })), void 0, !0),
      t.value.style.title.show && e.showInfo ? (f(), k("div", {
        key: 1,
        ref_key: "chartTitle",
        ref: P,
        class: "vue-ui-sparkline-title",
        style: ne(`display:flex;align-items:center;width:100%;color:${t.value.style.title.color};background:${t.value.style.backgroundColor};justify-content:${t.value.style.title.textAlign === "left" ? "flex-start" : t.value.style.title.textAlign === "right" ? "flex-end" : "center"};height:${t.value.style.title.fontSize * 2}px;font-size:${t.value.style.title.fontSize}px;font-weight:${t.value.style.title.bold ? "bold" : "normal"};`)
      }, [
        A("span", {
          style: ne(`padding:${t.value.style.title.textAlign === "left" ? "0 0 0 12px" : t.value.style.title.textAlign === "right" ? "0 12px 0 0" : "0"}`)
        }, Ce($.value ? $.value.period : t.value.style.title.text), 5)
      ], 4)) : L("", !0),
      A("div", Fa, [
        (f(), k("svg", {
          ref_key: "svgRef",
          ref: j,
          xmlns: y(jt),
          viewBox: `0 0 ${x.value.width} ${x.value.height}`,
          style: ne(`background:${t.value.style.backgroundColor};overflow:visible;direction:ltr`),
          tabindex: "0",
          "aria-describedby": `chart-instructions-${h.value}`,
          onMouseleave: r[0] || (r[0] = (u) => ae.value = void 0),
          onFocus: Nt,
          onBlur: Tt,
          onKeydown: It
        }, [
          Et(y(n)),
          l.$slots["chart-background"] ? (f(), k("foreignObject", {
            key: 0,
            x: 0,
            y: 0,
            width: x.value.width <= 0 ? 10 : x.value.width,
            height: x.value.height <= 0 ? 10 : x.value.height,
            style: {
              pointerEvents: "none"
            }
          }, [
            ve(l.$slots, "chart-background", {}, void 0, !0)
          ], 8, qa)) : L("", !0),
          A("defs", null, [
            A("linearGradient", {
              x1: "0%",
              y1: "0%",
              x2: "100%",
              y2: "0%",
              id: `sparkline_gradient_${h.value}`
            }, [
              A("stop", {
                offset: "0%",
                "stop-color": y(qe)(y(Ve)(t.value.style.area.color, 0.05), t.value.style.area.opacity)
              }, null, 8, Wa),
              A("stop", {
                offset: "100%",
                "stop-color": y(qe)(t.value.style.area.color, t.value.style.area.opacity)
              }, null, 8, Oa)
            ], 8, Ga),
            A("linearGradient", {
              x2: "0%",
              y2: "100%",
              id: `sparkline_bar_gradient_pos_${h.value}`
            }, [
              A("stop", {
                offset: "0%",
                "stop-color": t.value.style.bar.color
              }, null, 8, ja),
              A("stop", {
                offset: "100%",
                "stop-color": y(Ve)(t.value.style.bar.color, 0.05)
              }, null, 8, Ya)
            ], 8, za),
            A("linearGradient", {
              x2: "0%",
              y2: "100%",
              id: `sparkline_bar_gradient_neg_${h.value}`
            }, [
              A("stop", {
                offset: "0%",
                "stop-color": y(Ve)(t.value.style.bar.color, 0.05)
              }, null, 8, Ka),
              A("stop", {
                offset: "100%",
                "stop-color": t.value.style.bar.color
              }, null, 8, Ua)
            ], 8, Xa),
            A("filter", {
              id: `sparkline_pulse_glow_${h.value}`,
              filterUnits: "userSpaceOnUse",
              x: "-50",
              y: "-50",
              width: "100",
              height: "100"
            }, [...r[1] || (r[1] = [
              A("feGaussianBlur", {
                in: "SourceGraphic",
                stdDeviation: "3",
                result: "blur"
              }, null, -1),
              A("feMerge", null, [
                A("feMergeNode", { in: "blur" }),
                A("feMergeNode", { in: "SourceGraphic" })
              ], -1)
            ])], 8, Ha),
            t.value.temperatureColors.show && ce.value ? (f(), k("linearGradient", {
              key: 0,
              id: `temperature_grad_sparkline_${h.value}`,
              gradientTransform: "rotate(90)"
            }, [
              (f(!0), k(we, null, Pe(ce.value, (u, c) => (f(), k("stop", {
                key: `temperature_grad_stop_${c}_${h.value}`,
                "stop-color": u,
                offset: y(Yt)(c, ce.value.length)
              }, null, 8, Ja))), 128))
            ], 8, Za)) : L("", !0)
          ]),
          t.value.style.area.show && !H.value && g.value[0] && g.value.length > 1 ? (f(), k("g", Qa, [
            t.value.style.line.smooth ? (f(), k("path", {
              key: 0,
              class: "vue-ui-sparkline-area",
              d: `M ${g.value[0].x},${S.value.bottom} ${y(He)(g.value)} L ${g.value.at(-1).x},${S.value.bottom} Z`,
              fill: t.value.style.area.useGradient ? `url(#sparkline_gradient_${h.value})` : y(qe)(t.value.style.area.color, t.value.style.area.opacity),
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              style: ne({
                transition: y(D) ? void 0 : "all 0.2s"
              })
            }, null, 12, el)) : (f(), k("path", {
              key: 1,
              class: "vue-ui-sparkline-area",
              d: `M${Lt.value}Z`,
              fill: t.value.style.area.useGradient ? `url(#sparkline_gradient_${h.value})` : y(qe)(t.value.style.area.color, t.value.style.area.opacity),
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              style: ne({
                transition: y(D) ? void 0 : "all 0.2s"
              })
            }, null, 12, tl))
          ])) : L("", !0),
          t.value.style.line.smooth && !H.value ? (f(), k("path", {
            key: 2,
            class: "vue-ui-sparkline-path",
            id: de.value,
            d: `M ${y(He)(g.value) || "0,0"}`,
            stroke: ce.value ? `url(#temperature_grad_sparkline_${h.value})` : t.value.style.line.color,
            fill: "none",
            "stroke-width": t.value.style.line.strokeWidth,
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            style: ne({
              transition: y(D) ? void 0 : "all 0.2s"
            })
          }, null, 12, al)) : L("", !0),
          !t.value.style.line.smooth && !H.value ? (f(), k("path", {
            key: 3,
            class: "vue-ui-sparkline-path",
            id: de.value,
            d: `M ${y(kt)(g.value) || "0,0"}`,
            stroke: ce.value ? `url(#temperature_grad_sparkline_${h.value})` : t.value.style.line.color,
            fill: "none",
            "stroke-width": t.value.style.line.strokeWidth,
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            style: ne({
              transition: y(D) ? void 0 : "all 0.2s"
            })
          }, null, 12, ll)) : L("", !0),
          ht.value && !ce.value ? (f(), Le(Ba, {
            key: 4,
            svgPathData: ht.value,
            enabled: t.value.gradientPath.show && !H.value && !ce.value,
            strokeWidth: t.value.style.line.strokeWidth,
            highColor: t.value.gradientPath.colors.high,
            lowColor: t.value.gradientPath.colors.low,
            segments: t.value.gradientPath.segments
          }, null, 8, ["svgPathData", "enabled", "strokeWidth", "highColor", "lowColor", "segments"])) : L("", !0),
          Me.value && $e.value ? (f(), Le(pa, {
            key: 5,
            uid: h.value,
            svgRef: y(j),
            pulsePathId: de.value,
            pulsePathLength: se.value,
            pulseDur: _e.value,
            pulseBegin: Ne.value,
            pulseRepeatCount: Te.value,
            pulseFillMode: ie.value,
            pulseKeyPoints: he.value,
            pulseMotion: Ie.value,
            pulse: T.value,
            pulseTrail: G.value,
            pulseTrailLength: fe.value,
            prefersReducedMotion: y(p),
            loading: y(D),
            isBar: H.value
          }, null, 8, ["uid", "svgRef", "pulsePathId", "pulsePathLength", "pulseDur", "pulseBegin", "pulseRepeatCount", "pulseFillMode", "pulseKeyPoints", "pulseMotion", "pulse", "pulseTrail", "pulseTrailLength", "prefersReducedMotion", "loading", "isBar"])) : L("", !0),
          (f(!0), k(we, null, Pe(g.value, (u, c) => (f(), k("g", null, [
            H.value ? (f(), k("rect", {
              key: 0,
              x: u.x - u.width / 2,
              y: isNaN(u.absoluteValue > 0 ? u.y : ye.value) ? 0 : u.absoluteValue > 0 ? u.y : ye.value,
              width: u.width,
              height: isNaN(Math.abs(u.y - ye.value)) ? 0 : Math.abs(u.y - ye.value),
              fill: u.absoluteValue > 0 ? `url(#sparkline_bar_gradient_pos_${h.value})` : `url(#sparkline_bar_gradient_neg_${h.value})`,
              rx: t.value.style.bar.borderRadius
            }, null, 8, nl)) : L("", !0),
            t.value.style.verticalIndicator.show && ($.value && u.id === $.value.id || e.selectedIndex === c) ? (f(), k("line", {
              key: 1,
              x1: u.x,
              x2: u.x,
              y1: S.value.top - 6,
              y2: S.value.bottom,
              stroke: t.value.style.verticalIndicator.color || u.color,
              "stroke-width": t.value.style.verticalIndicator.strokeWidth,
              "stroke-linecap": "round",
              "stroke-dasharray": t.value.style.verticalIndicator.strokeDasharray || 0
            }, null, 8, rl)) : L("", !0)
          ]))), 256)),
          We.value < 0 ? (f(), k("line", {
            key: 6,
            x1: S.value.start,
            x2: S.value.start + S.value.width,
            y1: y(wt)(ye.value, S.value.bottom),
            y2: y(wt)(ye.value, S.value.bottom),
            stroke: t.value.style.zeroLine.color,
            "stroke-dasharray": t.value.style.zeroLine.strokeWidth * 2,
            "stroke-width": t.value.style.zeroLine.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, ul)) : L("", !0),
          t.value.style.plot.show ? (f(!0), k(we, { key: 7 }, Pe(g.value, (u, c) => (f(), k("g", null, [
            $.value && u.id === $.value.id || e.selectedIndex === c || y(U).length === 1 ? (f(), k("circle", {
              key: 0,
              cx: u.x,
              cy: u.y,
              r: t.value.style.plot.radius,
              fill: u.color,
              stroke: t.value.style.plot.stroke,
              "stroke-width": t.value.style.plot.strokeWidth
            }, null, 8, ol)) : L("", !0)
          ]))), 256)) : L("", !0),
          e.showInfo && t.value.style.dataLabel.show ? (f(), k("text", {
            key: 8,
            x: t.value.style.dataLabel.position === "left" ? 12 + t.value.style.dataLabel.offsetX : S.value.width + 12 + t.value.style.dataLabel.offsetX,
            y: x.value.height / 2 + t.value.style.dataLabel.fontSize / 2.5 + t.value.style.dataLabel.offsetY,
            "font-size": t.value.style.dataLabel.fontSize,
            "font-weight": t.value.style.dataLabel.bold ? "bold" : "normal",
            fill: t.value.style.dataLabel.color
          }, Ce($.value ? y(Ze)(
            t.value.style.dataLabel.formatter,
            $.value.absoluteValue,
            y(Je)({ p: t.value.style.dataLabel.prefix, v: $.value.absoluteValue, s: t.value.style.dataLabel.suffix, r: t.value.style.dataLabel.roundingValue }),
            { datapoint: $.value }
          ) : y(Ze)(
            t.value.style.dataLabel.formatter,
            vt.value,
            y(Je)({ p: t.value.style.dataLabel.prefix, v: vt.value, s: t.value.style.dataLabel.suffix, r: t.value.style.dataLabel.roundingValue })
          )), 9, sl)) : L("", !0),
          (f(!0), k(we, null, Pe(g.value, (u, c) => (f(), k("rect", {
            x: u.x - (S.value.width / (me.value + 1) > x.value.padding ? x.value.padding : S.value.width / (me.value + 1)) / 2,
            y: S.value.top - 6,
            height: S.value.height + 6,
            width: S.value.width / (me.value + 1) > x.value.padding ? x.value.padding : S.value.width / (me.value + 1),
            fill: "transparent",
            onMouseenter: () => De(u, c),
            onMouseleave: () => Pt(u, c),
            onClick: () => _t(u, c)
          }, null, 40, il))), 256)),
          ve(l.$slots, "svg", { svg: x.value }, void 0, !0)
        ], 44, Va)),
        l.$slots.hint ? (f(), k("div", dl, [
          ve(l.$slots, "hint", Xe(Ke({ hint: t.value.a11y.translations.keyboardNavigation, isVisible: ze.value })), void 0, !0)
        ])) : L("", !0)
      ]),
      $.value && t.value.style.tooltip.show ? (f(), Le(y(s), {
        key: 2,
        x: $.value.x,
        y: $.value.y,
        prevX: ae.value.x,
        prevY: ae.value.y,
        offsetY: t.value.style.plot.radius * 3 + t.value.style.tooltip.offsetY,
        svgRef: y(j),
        background: t.value.style.tooltip.backgroundColor,
        color: t.value.style.tooltip.color,
        fontSize: t.value.style.tooltip.fontSize,
        borderWidth: t.value.style.tooltip.borderWidth,
        borderColor: t.value.style.tooltip.borderColor,
        borderRadius: t.value.style.tooltip.borderRadius,
        backgroundOpacity: t.value.style.tooltip.backgroundOpacity
      }, {
        default: Ft(() => [
          ve(l.$slots, "tooltip", Xe(Ke({ ...$.value })), () => [
            Vt(Ce($.value.period) + ": " + Ce(y(Ze)(
              t.value.style.dataLabel.formatter,
              $.value.absoluteValue,
              y(Je)({
                p: t.value.style.dataLabel.prefix,
                v: $.value.absoluteValue,
                s: t.value.style.dataLabel.suffix,
                r: t.value.style.dataLabel.roundingValue
              }),
              { datapoint: $.value }
            )), 1)
          ], !0)
        ]),
        _: 3
      }, 8, ["x", "y", "prevX", "prevY", "offsetY", "svgRef", "background", "color", "fontSize", "borderWidth", "borderColor", "borderRadius", "backgroundOpacity"])) : L("", !0),
      l.$slots.source ? (f(), k("div", {
        key: 3,
        ref_key: "source",
        ref: N,
        dir: "auto"
      }, [
        ve(l.$slots, "source", {}, void 0, !0)
      ], 512)) : L("", !0),
      ve(l.$slots, "skeleton", {}, () => [
        y(D) ? (f(), Le(Jt, { key: 0 })) : L("", !0)
      ], !0)
    ], 12, Da));
  }
}, Al = /* @__PURE__ */ ra(cl, [["__scopeId", "data-v-5d38ffad"]]);
export {
  Al as default
};
