// @atproto/oauth-client/dist/util.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomEventTarget = exports.CustomEvent = exports.ifString = void 0;
exports.contentMime = contentMime;
exports.combineSignals = combineSignals;
const ifString = (v) => (typeof v === 'string' ? v : undefined);
exports.ifString = ifString;
function contentMime(headers) {
    return headers.get('content-type')?.split(';')[0].trim();
}
/**
 * Ponyfill for `CustomEvent` constructor.
 */
exports.CustomEvent = globalThis.CustomEvent ??
    (() => {
        var _CustomEvent_detail;
        class CustomEvent extends Event {
            constructor(type, options) {
                if (!arguments.length)
                    throw new TypeError('type argument is required');
                super(type, options);
                _CustomEvent_detail.set(this, void 0);
                __classPrivateFieldSet(this, _CustomEvent_detail, options?.detail ?? null, "f");
            }
            get detail() {
                return __classPrivateFieldGet(this, _CustomEvent_detail, "f");
            }
        }
        _CustomEvent_detail = new WeakMap();
        Object.defineProperties(CustomEvent.prototype, {
            [Symbol.toStringTag]: {
                writable: false,
                enumerable: false,
                configurable: true,
                value: 'CustomEvent',
            },
            detail: {
                enumerable: true,
            },
        });
        return CustomEvent;
    })();
class CustomEventTarget {
    constructor() {
        Object.defineProperty(this, "eventTarget", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new EventTarget()
        });
    }
    addEventListener(type, callback, options) {
        this.eventTarget.addEventListener(type, callback, options);
    }
    removeEventListener(type, callback, options) {
        this.eventTarget.removeEventListener(type, callback, options);
    }
    dispatchCustomEvent(type, detail, init) {
        return this.eventTarget.dispatchEvent(new exports.CustomEvent(type, { ...init, detail }));
    }
}
exports.CustomEventTarget = CustomEventTarget;
function combineSignals(signals) {
    const controller = new DisposableAbortController();
    const onAbort = function (_event) {
        const reason = new Error('This operation was aborted', {
            cause: this.reason,
        });
        controller.abort(reason);
    };
    try {
        for (const sig of signals) {
            if (sig) {
                sig.throwIfAborted();
                sig.addEventListener('abort', onAbort, { signal: controller.signal });
            }
        }
        return controller;
    }
    catch (err) {
        controller.abort(err);
        throw err;
    }
}
/**
 * Allows using {@link AbortController} with the `using` keyword, in order to
 * automatically abort them once the execution block ends.
 */
class DisposableAbortController extends AbortController {
    [Symbol.dispose]() {
        this.abort(new Error('AbortController was disposed'));
    }
}
