// @atproto/lex-schema/dist/core/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.success = success;
exports.failure = failure;
exports.failureReason = failureReason;
exports.successValue = successValue;
exports.catchall = catchall;
exports.createCatcher = createCatcher;
/*@__NO_SIDE_EFFECTS__*/
function success(value) {
    return { success: true, value };
}
/*@__NO_SIDE_EFFECTS__*/
function failure(reason) {
    return { success: false, reason };
}
/*@__NO_SIDE_EFFECTS__*/
function failureReason(result) {
    return result.reason;
}
/*@__NO_SIDE_EFFECTS__*/
function successValue(result) {
    return result.value;
}
/**
 * Catches any error and wraps it in a {@link ResultFailure<Error>}.
 *
 * @param err - The error to catch.
 * @returns A {@link ResultFailure<Error>} containing the caught error.
 * @example
 *
 * ```ts
 * declare function someFunction(): Promise<string>
 *
 * const result = await someFunction().then(success, catchall)
 * if (result.success) {
 *   console.log(result.value) // string
 * } else {
 *   console.error(result.error instanceof Error) // true
 *   console.error(result.error.message) // string
 * }
 * ```
 */
/*@__NO_SIDE_EFFECTS__*/
function catchall(err) {
    if (err instanceof Error)
        return failure(err);
    return failure(new Error('Unknown error', { cause: err }));
}
/**
 * Creates a catcher function for the given constructor that wraps caught errors
 * in a {@link ResultFailure}.
 *
 * @example
 *
 * ```ts
 * class FooError extends Error {}
 * class BarError extends Error {}
 *
 * declare function someFunction(): Promise<string>
 *
 * const result = await someFunction()
 *   .then(success)
 *   .catch(createCatcher(FooError))
 *   .catch(createCatcher(BarError))
 *
 * if (result.success) {
 *   console.log(result.value) // string
 * } else {
 *   console.error(result.error) // FooError | BarError
 * }
 */
/*@__NO_SIDE_EFFECTS__*/
function createCatcher(Ctor) {
    return (err) => {
        if (err instanceof Ctor)
            return failure(err);
        throw err;
    };
}
