// @atproto/lex-schema/dist/core/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schema = void 0;
const validator_js_1 = require("./validator.js");
class Schema {
    /**
     * @note use {@link check}() instead of {@link assert}() if you encounter a
     * `ts(2775)` error and you are not able to fully type the validator. This
     * will typically arise in generic contexts, where the narrowed type is not
     * needed.
     */
    assert(input) {
        const result = validator_js_1.ValidationContext.validate(input, this);
        if (!result.success)
            throw result.reason;
    }
    /**
     * Alias for {@link assert}(). Most useful in generic contexts where the
     * validator is not exactly typed, allowing to avoid "_Assertions require
     * every name in the call target to be declared with an explicit type
     * annotation. ts(2775)_" errors.
     */
    check(input) {
        this.assert(input);
    }
    /**
     * Casts the input (by validating it) to the output type if it matches the
     * schema, otherwise throws. This is the same as calling {@link parse}() with
     * `mode: "validate"`.
     */
    cast(input) {
        const result = validator_js_1.ValidationContext.validate(input, this);
        if (result.success)
            return result.value;
        throw result.reason;
    }
    matches(input) {
        const result = validator_js_1.ValidationContext.validate(input, this);
        return result.success;
    }
    ifMatches(input) {
        return this.matches(input) ? input : undefined;
    }
    parse(input, options) {
        const result = this.safeParse(input, options);
        if (result.success)
            return result.value;
        throw result.reason;
    }
    safeParse(input, options) {
        return validator_js_1.ValidationContext.validate(input, this, {
            ...options,
            mode: 'parse',
        });
    }
    validate(input, options) {
        const result = this.safeValidate(input, options);
        if (result.success)
            return result.value;
        throw result.reason;
    }
    safeValidate(input, options) {
        return validator_js_1.ValidationContext.validate(input, this, {
            ...options,
            mode: 'validate',
        });
    }
    // @NOTE The built lexicons namespaces will export utility functions that
    // allow accessing the schema's methods without the need to specify ".main."
    // as part of the namespace. This way, a utility for a particular record type
    // can be called like "app.bsky.feed.post.<utility>()" instead of
    // "app.bsky.feed.post.main.<utility>()". Because those utilities could
    // conflict with other schemas (e.g. if there is a lexicon definition at
    // "#<utility>"), those exported utilities will be prefixed with "$". In order
    // to be able to consistently call the utilities, when using the "main" and
    // non "main" definitions, we also expose the same methods with a "$" prefix.
    // Thanks to this, both of the following call will be possible:
    //
    // - "app.bsky.feed.post.$parse(...)" // calls a utility function created by "lex build"
    // - "app.bsky.feed.defs.postView.$parse(...)" // uses the alias defined below on the schema instance
    $assert(input) {
        return this.assert(input);
    }
    $check(input) {
        return this.check(input);
    }
    $cast(input) {
        return this.cast(input);
    }
    $matches(input) {
        return this.matches(input);
    }
    $ifMatches(input) {
        return this.ifMatches(input);
    }
    $parse(input, options) {
        return this.parse(input, options);
    }
    $safeParse(input, options) {
        return this.safeParse(input, options);
    }
    $validate(input, options) {
        return this.validate(input, options);
    }
    $safeValidate(input, options) {
        return this.safeValidate(input, options);
    }
}
exports.Schema = Schema;
