// @atproto/lex-schema/dist/core/string-format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STRING_FORMATS = exports.isUriString = exports.isTidString = exports.isRecordKeyString = exports.isNsidString = exports.isLanguageString = exports.isHandleString = exports.isDidString = exports.isDatetimeString = exports.isCidString = exports.isAtUriString = exports.isAtIdentifierString = void 0;
exports.isStringFormat = isStringFormat;
exports.assertStringFormat = assertStringFormat;
exports.asStringFormat = asStringFormat;
exports.ifStringFormat = ifStringFormat;
const lex_data_1 = require("@atproto/lex-data");
const syntax_1 = require("@atproto/syntax");
exports.isAtIdentifierString = syntax_1.isValidAtIdentifier;
exports.isAtUriString = syntax_1.isValidAtUri;
exports.isCidString = ((v) => (0, lex_data_1.validateCidString)(v));
exports.isDatetimeString = syntax_1.isValidDatetime;
exports.isDidString = syntax_1.isValidDid;
exports.isHandleString = syntax_1.isValidHandle;
exports.isLanguageString = syntax_1.isValidLanguage;
exports.isNsidString = syntax_1.isValidNsid;
exports.isRecordKeyString = syntax_1.isValidRecordKey;
exports.isTidString = syntax_1.isValidTid;
exports.isUriString = syntax_1.isValidUri;
const stringFormatVerifiers = /*#__PURE__*/ Object.freeze({
    __proto__: null,
    'at-identifier': exports.isAtIdentifierString,
    'at-uri': exports.isAtUriString,
    cid: exports.isCidString,
    datetime: exports.isDatetimeString,
    did: exports.isDidString,
    handle: exports.isHandleString,
    language: exports.isLanguageString,
    nsid: exports.isNsidString,
    'record-key': exports.isRecordKeyString,
    tid: exports.isTidString,
    uri: exports.isUriString,
});
/*@__NO_SIDE_EFFECTS__*/
function isStringFormat(input, format) {
    const formatVerifier = stringFormatVerifiers[format];
    // Fool-proof
    if (!formatVerifier)
        throw new TypeError(`Unknown string format: ${format}`);
    return formatVerifier(input);
}
/*@__NO_SIDE_EFFECTS__*/
function assertStringFormat(input, format) {
    if (!isStringFormat(input, format)) {
        throw new TypeError(`Invalid string format (${format}): ${input}`);
    }
}
/*@__NO_SIDE_EFFECTS__*/
function asStringFormat(input, format) {
    assertStringFormat(input, format);
    return input;
}
/*@__NO_SIDE_EFFECTS__*/
function ifStringFormat(input, format) {
    return isStringFormat(input, format) ? input : undefined;
}
exports.STRING_FORMATS = Object.freeze(
/*#__PURE__*/ Object.keys(stringFormatVerifiers));
