// @atproto/lex-schema/dist/core/validation-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = void 0;
const lex_data_1 = require("@atproto/lex-data");
const array_agg_js_1 = require("../util/array-agg.js");
const result_js_1 = require("./result.js");
const validation_issue_js_1 = require("./validation-issue.js");
class ValidationError extends lex_data_1.LexError {
    name = 'ValidationError';
    issues;
    constructor(issues, options) {
        const issuesAgg = aggregateIssues(issues);
        super('InvalidRequest', issuesAgg.join(', '), options);
        this.issues = issuesAgg;
    }
    toJSON() {
        return {
            ...super.toJSON(),
            issues: this.issues.map((issue) => issue.toJSON()),
        };
    }
    static fromFailures(failures) {
        if (failures.length === 1)
            return (0, result_js_1.failureReason)(failures[0]);
        const issues = failures.flatMap(extractFailureIssues);
        return new ValidationError(issues, {
            // Keep the original errors as the cause chain
            cause: failures.map(result_js_1.failureReason),
        });
    }
}
exports.ValidationError = ValidationError;
function extractFailureIssues(result) {
    return result.reason.issues;
}
function aggregateIssues(issues) {
    // Quick path for common cases
    if (issues.length <= 1)
        return issues;
    if (issues.length === 2 && issues[0].code !== issues[1].code)
        return issues;
    return [
        // Aggregate invalid_type with identical paths
        ...(0, array_agg_js_1.arrayAgg)(issues.filter((issue) => issue instanceof validation_issue_js_1.IssueInvalidType), (a, b) => comparePropertyPaths(a.path, b.path), (issues) => new validation_issue_js_1.IssueInvalidType(issues[0].path, issues[0].input, Array.from(new Set(issues.flatMap((iss) => iss.expected))))),
        // Aggregate invalid_value with identical paths
        ...(0, array_agg_js_1.arrayAgg)(issues.filter((issue) => issue instanceof validation_issue_js_1.IssueInvalidValue), (a, b) => comparePropertyPaths(a.path, b.path), (issues) => new validation_issue_js_1.IssueInvalidValue(issues[0].path, issues[0].input, Array.from(new Set(issues.flatMap((iss) => iss.values))))),
        // Pass through other issues
        ...issues.filter((issue) => !(issue instanceof validation_issue_js_1.IssueInvalidType) &&
            !(issue instanceof validation_issue_js_1.IssueInvalidValue)),
    ];
}
/*@__NO_SIDE_EFFECTS__*/
function comparePropertyPaths(a, b) {
    if (a.length !== b.length)
        return false;
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i])
            return false;
    }
    return true;
}
