// @atproto/lex-schema/dist/core/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationContext = void 0;
const result_js_1 = require("./result.js");
const validation_error_js_1 = require("./validation-error.js");
const validation_issue_js_1 = require("./validation-issue.js");
class ValidationContext {
    options;
    static validate(input, validator, options) {
        const context = new ValidationContext({
            path: options?.path ?? [],
            mode: options?.mode ?? 'validate',
        });
        return context.validate(input, validator);
    }
    currentPath;
    issues = [];
    constructor(options) {
        this.options = options;
        // Create a copy because we will be mutating the array during validation.
        this.currentPath = Array.from(options.path);
    }
    get path() {
        return Array.from(this.currentPath);
    }
    concatPath(path) {
        if (path == null)
            return this.path;
        return this.currentPath.concat(path);
    }
    /**
     * This is basically the entry point for validation within a context. Use this
     * method instead of using {@link Validator.validateInContext} directly,
     * because this method ensures the proper use of {@link ValidationOptions}.
     */
    validate(input, validator) {
        // This is the only place where validateInContext should be called.
        const result = validator.validateInContext(input, this);
        if (result.success) {
            if (this.issues.length > 0) {
                // Validator returned a success but issues were added via the context.
                // This means the overall validation failed.
                return (0, result_js_1.failure)(new validation_error_js_1.ValidationError(Array.from(this.issues)));
            }
            if (this.options.mode !== 'parse' && !Object.is(result.value, input)) {
                // If the value changed, it means that a default (or some other
                // transformation) was applied, meaning that the original value did
                // *not* match the (output) schema. When not in "parse" mode, we
                // consider this a failure.
                // This check is the reason why Validator.validateInContext should not
                // be used directly, and ValidatorContext.validate should be used
                // instead, even when delegating validation from one validator to
                // another.
                // This if block comes before the next one because 'this.issues' will
                // end-up being appended to the returned ValidationError (see the
                // "failure" method below), resulting in a more complete error report.
                return this.issueInvalidValue(input, [result.value]);
            }
        }
        return result;
    }
    validateChild(input, key, validator) {
        // Instead of creating a new context, we just push/pop the path segment.
        this.currentPath.push(key);
        try {
            return this.validate(input[key], validator);
        }
        finally {
            this.currentPath.length--;
        }
    }
    addIssue(issue) {
        this.issues.push(issue);
    }
    success(value) {
        return (0, result_js_1.success)(value);
    }
    failure(reason) {
        return (0, result_js_1.failure)(reason);
    }
    issue(issue) {
        return this.failure(new validation_error_js_1.ValidationError([...this.issues, issue]));
    }
    issueInvalidValue(input, values) {
        return this.issue(new validation_issue_js_1.IssueInvalidValue(this.path, input, values));
    }
    issueInvalidType(input, expected) {
        return this.issue(new validation_issue_js_1.IssueInvalidType(this.path, input, [expected]));
    }
    issueRequiredKey(input, key) {
        return this.issue(new validation_issue_js_1.IssueRequiredKey(this.path, input, key));
    }
    issueInvalidFormat(input, format, msg) {
        return this.issue(new validation_issue_js_1.IssueInvalidFormat(this.path, input, format, msg));
    }
    issueTooBig(input, type, max, actual) {
        return this.issue(new validation_issue_js_1.IssueTooBig(this.path, input, max, type, actual));
    }
    issueTooSmall(input, type, min, actual) {
        return this.issue(new validation_issue_js_1.IssueTooSmall(this.path, input, min, type, actual));
    }
    issueInvalidPropertyValue(input, property, values) {
        const value = input[property];
        const path = this.concatPath(property);
        return this.issue(new validation_issue_js_1.IssueInvalidValue(path, value, values));
    }
    issueInvalidPropertyType(input, property, expected) {
        const value = input[property];
        const path = this.concatPath(property);
        return this.issue(new validation_issue_js_1.IssueInvalidType(path, value, [expected]));
    }
}
exports.ValidationContext = ValidationContext;
